<!DOCTYPE html>


<html lang="zh-CN">


<head>
  <meta charset="utf-8" />
    
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />
  <title>
    基于芯片数据的亲子鉴定分析代码 |  VincereZhou&#39;s blog
  </title>
  <meta name="generator" content="hexo-theme-ayer">
  
  <link rel="shortcut icon" href="/images/mojie.jpg" />
  
  
<link rel="stylesheet" href="/dist/main.css">

  <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Shen-Yu/cdn/css/remixicon.min.css">
  
<link rel="stylesheet" href="/css/custom.css">

  
  <script src="https://cdn.jsdelivr.net/npm/pace-js@1.0.2/pace.min.js"></script>
  
  

  

<link rel="alternate" href="/atom.xml" title="VincereZhou's blog" type="application/atom+xml">
</head>

</html>

<body>
  <div id="app">
    
      
    <main class="content on">
      <section class="outer">
  <article
  id="post-基于芯片数据的亲子鉴定分析代码"
  class="article article-type-post"
  itemscope
  itemprop="blogPost"
  data-scroll-reveal
>
  <div class="article-inner">
    
    <header class="article-header">
       
<h1 class="article-title sea-center" style="border-left:0" itemprop="name">
  基于芯片数据的亲子鉴定分析代码
</h1>
 

    </header>
     
    <div class="article-meta">
      <a href="/posts/42752926/" class="article-date">
  <time datetime="2021-08-03T06:25:45.000Z" itemprop="datePublished">2021-08-03</time>
</a> 
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/">数据分析</a> / <a class="article-category-link" href="/categories/%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/python/">python</a>
  </div>
  
<div class="word_count">
    <span class="post-time">
        <span class="post-meta-item-icon">
            <i class="ri-quill-pen-line"></i>
            <span class="post-meta-item-text"> 字数统计:</span>
            <span class="post-count">3.3k</span>
        </span>
    </span>

    <span class="post-time">
        &nbsp; | &nbsp;
        <span class="post-meta-item-icon">
            <i class="ri-book-open-line"></i>
            <span class="post-meta-item-text"> 阅读时长≈</span>
            <span class="post-count">15 分钟</span>
        </span>
    </span>
</div>
 
    </div>
      
    <div class="tocbot"></div>




  
    <div class="article-entry" itemprop="articleBody">
       
  <p>本软件通过利用SNP信息根据孟德尔错误位点比例，对于系谱中均具有基因型的亲子对进行亲子鉴定，判断系谱是否正确。这里的孟德尔错误位点指个体和亲本之间具有相反的纯和子的位点（例如个体的基因型为AA，亲本的基因型为GG），依据孟德尔定律亲子间不应存在这种情况，但由于突变和分型错误等原因，真实的亲子对之间仍存在少量的孟德尔错误位点。当使用的位点数目较多时，通过设定一个合理的阈值可以明确判定系谱中正确的亲子关系和错误的亲子关系。</p>
<span id="more"></span>
<h1>使用软件</h1>
<h1>软件</h1>
<p>python 3.0 以上版本，事先安装 numpy 模块</p>
<h1>输入文件格式</h1>
<ul>
<li>
<p>raw 文件：通过 plink 软件对<strong>原始芯片数据</strong>（未填充）采用 <code>--recodeA</code> 选项生成的文件，后缀为 <code>.raw</code></p>
</li>
<li>
<p>映射文件： 两列，所有基因型个体的芯片号与个体号，空格或 tab分隔，无标题。</p>
</li>
<li>
<p>系谱文件：三列，个体号-父号-母号，空格或 tab分隔，无标题。</p>
</li>
</ul>
<h1>输出文件说明</h1>
<ul>
<li><code>check_pedigree.csv</code>: 五列，个体 - 原系谱中的父本 - 原系谱中的母本 -  父本判定信息 - 母本判定信息（ 三种判定信息，par_nogeno : 亲本没有基因型信息，Match : 亲本符合阈值要求， No-Match : 亲本不符合阈值要求 ）</li>
<li><code>check_paternity_prob.csv</code>: 七列，亲子对类型 - 个体 - 亲本 - 错误位点数 -  匹配的位点总数（剔除缺失位点）- 错误位点概率 - 判定信息</li>
</ul>
<h1>运行代码</h1>
<p>代码文件见：<a target="_blank" rel="noopener" href="https://github.com/VincereZhou/code_backup/blob/master/paternity_check_1.0.py">paternity_check_1.0.py</a></p>
<p>将输入文件和本程序放在同一文件夹下，运行命令示范为</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">python paternity_check_1.0.py --raw test.raw --chipid test_id.txt --pedigree pedigree.txt --check-prob 0.005</span><br></pre></td></tr></table></figure>
<p><strong>可用参数说明：</strong></p>
<p><code>--raw</code> : 基因型文件名，必选。</p>
<p><code>--chipid</code> : 基因型个体映射文件名，必选。</p>
<p><code>--pedigree</code> : 系谱文件名，必选</p>
<p><code>--list</code> : 需要进行亲子鉴定的个体文件名，可选。如果提供该参数，则只对这个文件中的个体进行亲子鉴定；如果不提供该参数，则默认对所有基因型个体进行亲子鉴定。</p>
<p>-<code>-check-prob</code> : 亲子鉴定孟德尔错误率的阈值，可选，默认为0.005</p>
<p>-<code>-miss-snp</code> : SNP位点的基因分型缺失率阈值，可选，默认为0.2</p>
<p>-<code>-maf</code> : SNP 位点的最小等位基因频率阈值，可选，默认为0.01</p>
<p>-<code>-miss-sample</code> : 样本的基因分型缺失率阈值，可选，默认为0.5</p>
<h1>代码说明</h1>
<h2 id="读取文件">读取文件</h2>
<p>这里首先打印了一下 raw 文件中的样本数目，实质采用 linux 命令 <code>wc -l</code></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#打印raw文件行数</span></span><br><span class="line">raw_lines = os.popen(<span class="string">&quot;wc -l &#123;&#125;&quot;</span>.<span class="built_in">format</span>(args.raw)).read().split()[<span class="number">0</span>]</span><br><span class="line">true_raw_lines = <span class="built_in">int</span>(raw_lines) -<span class="number">1</span> </span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Number of records in raw file: &#123;&#125;\n&quot;</span>.<span class="built_in">format</span>(true_raw_lines))</span><br><span class="line"></span><br><span class="line"><span class="comment">#读取文件</span></span><br><span class="line">raw_file = <span class="built_in">open</span>(args.raw,<span class="string">&#x27;r&#x27;</span>)</span><br><span class="line">chipid_file = <span class="built_in">open</span>(args.chipid,<span class="string">&#x27;r&#x27;</span>)</span><br><span class="line">pedigree_file = <span class="built_in">open</span>(args.pedigree,<span class="string">&#x27;r&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#new_raw = open(&quot;newid.raw&quot;,&#x27;w+&#x27;)</span></span><br><span class="line"></span><br><span class="line">check_prob = <span class="built_in">open</span>(<span class="string">&quot;chek_paternity_prob.txt&quot;</span>,<span class="string">&#x27;w&#x27;</span>)</span><br><span class="line">check_pedigree = <span class="built_in">open</span>(<span class="string">&quot;check_pedigree.txt&quot;</span>,<span class="string">&#x27;w&#x27;</span>)</span><br></pre></td></tr></table></figure>
<h2 id="映射与raw文件处理">映射与raw文件处理</h2>
<p>这里先对映射文件进行处理，过滤条件有两个：</p>
<ol>
<li>每一行的列数必须大于等于2</li>
<li>重复个体号会保留第一个</li>
</ol>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">set1=<span class="built_in">set</span>();dick=&#123;&#125;</span><br><span class="line">j=<span class="number">0</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> chipid_file:</span><br><span class="line">    f = i.split()</span><br><span class="line">    j+=<span class="number">1</span></span><br><span class="line">    <span class="keyword">if</span> <span class="built_in">len</span>(f) &gt;= <span class="number">2</span>:</span><br><span class="line">        <span class="keyword">if</span> f[<span class="number">1</span>] <span class="keyword">not</span> <span class="keyword">in</span> set1: <span class="comment">#去除重复个体号</span></span><br><span class="line">            set1.add(f[<span class="number">1</span>])</span><br><span class="line">            dick[f[<span class="number">0</span>]] = f[<span class="number">1</span>]</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&quot;Warning: &#123;&#125; row in chipid file has less than 2 coloum\n&quot;</span>.<span class="built_in">format</span>(j))</span><br></pre></td></tr></table></figure>
<p>再对 raw 文件进行处理，提取在映射文件中的个体的基因型，生成基因型的二维数组（array1）。这里将缺失值NA替换为3。<code>gene_id_array</code> 是相应的个体号一维数组。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br></pre></td><td class="code"><pre><span class="line">gene_list=[]</span><br><span class="line">raw_list=[]</span><br><span class="line">raw_file.readline() <span class="comment">#剔除标题</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> raw_file:</span><br><span class="line">    f = i.split()</span><br><span class="line">    <span class="keyword">if</span> f[<span class="number">1</span>] <span class="keyword">in</span> dick:</span><br><span class="line">        gene_list.append(dick[f[<span class="number">1</span>]])</span><br><span class="line">        raw_list.append(<span class="string">&#x27;\t&#x27;</span>.join(f[<span class="number">6</span>:]))</span><br><span class="line">        <span class="comment">#new_raw.write(&#x27;\t&#x27;.join(f[6:])+&#x27;\n&#x27;)</span></span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&quot;Warning: &#123;&#125; in raw file not in chpid file or is duplicated samples\n&quot;</span>.<span class="built_in">format</span>(f[<span class="number">1</span>]))</span><br><span class="line"></span><br><span class="line">gene_id_array = np.array(gene_list)</span><br><span class="line"></span><br><span class="line"><span class="comment">#new_raw.seek(0)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># snp 缺失值过滤，结果文件是 array2</span></span><br><span class="line"><span class="comment">#numpy 0/1 结果会显示为 nan</span></span><br><span class="line">array1 = np.genfromtxt(raw_list, dtype=np.int8, missing_values=<span class="string">&quot;NA&quot;</span>, filling_values=<span class="number">3</span>)</span><br><span class="line"><span class="comment">#array2 = array1.copy()</span></span><br><span class="line"></span><br><span class="line">sample_num = array1.shape[<span class="number">0</span>]</span><br><span class="line">snp_num = array1.shape[<span class="number">1</span>]</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Number of SNPs in raw file: &#123;&#125;\n&quot;</span>.<span class="built_in">format</span>(snp_num))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Number of records present in chipid file: &#123;&#125;\n&quot;</span>.<span class="built_in">format</span>(sample_num))</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h2 id="位点和样本质控">位点和样本质控</h2>
<h3 id="位点缺失率">位点缺失率</h3>
<p>这里统计每一列（SNP）中等于3（表示缺失）的比例，剔除缺失率高于阈值的SNP。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line">bool_array1 = array1==<span class="number">3</span></span><br><span class="line"></span><br><span class="line">total_miss = bool_array1.<span class="built_in">sum</span>()/(bool_array1.shape[<span class="number">0</span>] * bool_array1.shape[<span class="number">1</span>])</span><br><span class="line">toatl_callrate = <span class="number">1</span> - total_miss</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Total genotyping rate is &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(toatl_callrate))</span><br><span class="line"></span><br><span class="line">snp_missing_count = bool_array1.<span class="built_in">sum</span>(axis=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line">snp_missing_prob = snp_missing_count/sample_num</span><br><span class="line"></span><br><span class="line">snp_missing_min = snp_missing_prob.<span class="built_in">min</span>()</span><br><span class="line">snp_missing_mean = snp_missing_prob.mean()</span><br><span class="line">snp_missing_max = snp_missing_prob.<span class="built_in">max</span>()</span><br><span class="line"></span><br><span class="line"><span class="comment">#打印snp缺失值情况</span></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;SNP missing\n&quot;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  Min missing rate: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(snp_missing_min))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  AVG missing rate: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(snp_missing_mean))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  Max missing rate: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(snp_missing_max))</span><br><span class="line"><span class="comment">#np.savetxt(&quot;snp_missing.txt&quot;,snp_missing_prob, fmt=&#x27;%.4f&#x27;)</span></span><br><span class="line"></span><br><span class="line">snp_missing_pick = snp_missing_prob &lt; <span class="built_in">float</span>(args.miss_snp)</span><br><span class="line">array2 = array1[:,snp_missing_pick]</span><br><span class="line">snp_miss_remove = array1.shape[<span class="number">1</span>] - array2.shape[<span class="number">1</span>]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;&#123;&#125; snps removed due to missing genotype data\n&quot;</span>.<span class="built_in">format</span>(snp_miss_remove))</span><br><span class="line"><span class="keyword">del</span> array1</span><br><span class="line"><span class="keyword">del</span> bool_array1</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="MAF-过滤">MAF 过滤</h3>
<p>统计每个SNP的最小等位基因频率（MAF），剔除低于阈值的位点。</p>
<p>这里的做法是先将缺失值替换为0，统计每一列的和作为分子，分母为该列非缺失样本数的两倍。这么做的原理是 plink 软件 在生成 raw 文件时是按照最小等位基因计数的（raw 文件中的 0 1 2 表示最小等位基因的数目），因此将缺失值替换为0，再求每列的和，便得到最小等位基因的总数，除以所有等位基因总数，便得到了最小等位基因频率。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># maf 过滤,结果文件是 array3</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#计算array2每列的缺失值</span></span><br><span class="line">bool_array2 = array2==<span class="number">3</span></span><br><span class="line">snp_missing_count2 = bool_array2.<span class="built_in">sum</span>(axis=<span class="number">0</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#将array2 中的3替换为0，提取备份为 array2_copy</span></span><br><span class="line">array2_copy = array2.copy()</span><br><span class="line">array2[array2==<span class="number">3</span>] = <span class="number">0</span></span><br><span class="line"></span><br><span class="line">maf_sum = array2.<span class="built_in">sum</span>(axis=<span class="number">0</span>)</span><br><span class="line"><span class="comment">#print(&quot;maf_sum[1:5] are &#123;&#125;&quot;.format(maf_sum[1:5]))</span></span><br><span class="line">maf_denominator = <span class="number">2</span>*(sample_num-snp_missing_count2)</span><br><span class="line"><span class="comment">#print(&quot;sample_num are &#123;&#125;&quot;.format(sample_num))</span></span><br><span class="line"><span class="comment">#print(&quot;maf_denominator[1:5] are &#123;&#125;&quot;.format(maf_denominator[1:5]))</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line">maf_prob = maf_sum/maf_denominator</span><br><span class="line"><span class="comment">#print(&quot;maf_prob[1:5] are &#123;&#125;&quot;.format(maf_prob[1:5]))</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#maf_out = np.column_stack((maf_sum,maf_denominator,maf_prob))</span></span><br><span class="line"><span class="comment">#np.savetxt(&quot;maf_all.txt&quot;,maf_out)</span></span><br><span class="line"><span class="comment">#np.savetxt(&quot;maf.txt&quot;,maf_prob, fmt=&#x27;%.4f&#x27;)</span></span><br><span class="line"></span><br><span class="line">maf_min = maf_prob.<span class="built_in">min</span>()</span><br><span class="line">maf_mean = maf_prob.mean()</span><br><span class="line">maf_max = maf_prob.<span class="built_in">max</span>()</span><br><span class="line"></span><br><span class="line"><span class="comment">#打印MAF情况</span></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Minor alleles frequencies\n&quot;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  Min alleles frequencies: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(maf_min))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  AVG alleles frequencies: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(maf_mean))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  Max alleles frequencies: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(maf_max))</span><br><span class="line"></span><br><span class="line">snp_maf_pick = maf_prob &gt; <span class="built_in">float</span>(args.maf)</span><br><span class="line">array3 = array2_copy[:,snp_maf_pick]</span><br><span class="line"></span><br><span class="line">snp_maf_remove = array2_copy.shape[<span class="number">1</span>] - array3.shape[<span class="number">1</span>]</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;&#123;&#125; snps removed due to minor allele threshold\n&quot;</span>.<span class="built_in">format</span>(snp_maf_remove))</span><br><span class="line"><span class="keyword">del</span> array2</span><br><span class="line"><span class="keyword">del</span> array2_copy</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="样本缺失率过滤">样本缺失率过滤</h3>
<p>先过滤SNP，再过滤样本，这和 plink 软件的逻辑一致，也符合实际需求。毕竟每个样本都是花了钱的，能不剔除就不剔除。</p>
<p>原理同SNP过滤，只不过列换成了行（每一列表示一个SNP，每一行表示一个样本）。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># sample_miss 过滤，结果为 array4</span></span><br><span class="line"></span><br><span class="line">sample_num3 = array3.shape[<span class="number">0</span>]</span><br><span class="line">snp_num3 = array3.shape[<span class="number">1</span>]</span><br><span class="line"></span><br><span class="line">bool_array3 = array3==<span class="number">3</span></span><br><span class="line"></span><br><span class="line">sample_missing_count = bool_array3.<span class="built_in">sum</span>(axis=<span class="number">1</span>)</span><br><span class="line">sample_missing_prob = sample_missing_count/snp_num3</span><br><span class="line"></span><br><span class="line">sample_missing_min = sample_missing_prob.<span class="built_in">min</span>()</span><br><span class="line">sample_missing_mean = sample_missing_prob.mean()</span><br><span class="line">sample_missing_max = sample_missing_prob.<span class="built_in">max</span>()</span><br><span class="line"></span><br><span class="line"><span class="comment">#打印sample缺失值情况</span></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Sample missing\n&quot;</span>)</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  Min missing rate: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(sample_missing_min))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  AVG missing rate: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(sample_missing_mean))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  Max missing rate: &#123;:.3f&#125;\n&quot;</span>.<span class="built_in">format</span>(sample_missing_max))</span><br><span class="line"></span><br><span class="line"><span class="comment">#np.savetxt(&quot;sample_missing.txt&quot;,sample_missing_prob, fmt=&#x27;%.4f&#x27;)</span></span><br><span class="line"></span><br><span class="line">sample_missing_pick = sample_missing_prob &lt; <span class="built_in">float</span>(args.miss_sample)</span><br><span class="line">array4 = array3[sample_missing_pick,:]</span><br><span class="line"></span><br><span class="line">gene_id_array_new = gene_id_array[sample_missing_pick] <span class="comment">#样本过滤后的基因个体号</span></span><br><span class="line"></span><br><span class="line">sample_miss_remove = array3.shape[<span class="number">0</span>] - array4.shape[<span class="number">0</span>]</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;&#123;&#125; samples removed due to missing genotype data\n&quot;</span>.<span class="built_in">format</span>(sample_miss_remove))</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;&#123;&#125; samples and &#123;&#125; snps pass QC finally\n&quot;</span>.<span class="built_in">format</span>(array4.shape[<span class="number">0</span>],array4.shape[<span class="number">1</span>]))</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>如果剔除了样本，将样本和缺失率写入一个文件中</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#如果删除了样本，将删除的样本id和缺失率写入一个新文件</span></span><br><span class="line"><span class="keyword">if</span> sample_miss_remove &gt; <span class="number">0</span> :</span><br><span class="line">    filter_sample_file = <span class="built_in">open</span>(<span class="string">&#x27;sample_filted.txt&#x27;</span>,<span class="string">&#x27;w&#x27;</span>)</span><br><span class="line">    sample_missing_fail = sample_missing_prob &gt;= <span class="built_in">float</span>(args.miss_sample)</span><br><span class="line">    gene_id_fail_array = gene_id_array[sample_missing_fail]</span><br><span class="line">    sample_missing_prob_array = sample_missing_prob[sample_missing_fail]</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> <span class="built_in">range</span>(gene_id_fail_array.shape[<span class="number">0</span>]):</span><br><span class="line">        filter_sample_file.write(<span class="built_in">str</span>(gene_id_fail_array[i])+<span class="string">&#x27;\t&#x27;</span>+<span class="built_in">str</span>(<span class="string">&quot;&#123;:.4f&#125;&quot;</span>.<span class="built_in">format</span>(sample_missing_prob_array[i]))+<span class="string">&#x27;\n&#x27;</span>)</span><br><span class="line">    filter_sample_file.close()</span><br><span class="line">    </span><br><span class="line"><span class="keyword">del</span> array3</span><br><span class="line"><span class="keyword">del</span> bool_array3</span><br></pre></td></tr></table></figure>
<h2 id="亲子鉴定">亲子鉴定</h2>
<p>首先将基因型数组改为 1 2 3 0 ，对应之前的 0 1 2 3 。改完以后，0 就表示缺失，1 2 3 表示三种基因型。这样改方便之处在于：</p>
<ol>
<li>缺失值是0，0乘任何数还是0。</li>
<li>1,2,3 都是<strong>质数</strong>，这三个数之间任选两个数进行相乘得到的积是唯一的，比如如果乘积为2，那么只能是1和2相乘，3只能是1和3相乘……</li>
</ol>
<p>因此，这样我们在比对两个样本的基因型时，只需要将两个个体的基因型进行<strong>相乘</strong>，如果某个位点乘积为0，表示存在缺失（至少一个个体的基因型为缺失），无法比较；如果某个位点乘积为 3，表示在这个位点上存在孟德尔错误（两个个体为相反纯合子）。这样，我们统计两个个体的缺失位点和孟德尔错误位点非常方便。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#开始进行亲子鉴定</span><br><span class="line">#array4 改为 1 2 3 0(缺失)</span><br><span class="line">#这样两个array 的乘积的array为3的是错误位点，是0的是缺失位点。</span><br><span class="line"></span><br><span class="line">array4 +&#x3D; 1</span><br><span class="line">array4[array4&#x3D;&#x3D;4] &#x3D; 0</span><br><span class="line"></span><br><span class="line">snp_num4 &#x3D; array4.shape[1]</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="清洗系谱">清洗系谱</h3>
<p>这里剔除了第一列个体没有基因型的行，因为个体没有基因型就不可能进行亲子鉴定 （如果提过了 --list 选项，那么只会提取需要亲子鉴定的个体的系谱）。</p>
<p>这里得到个体和亲本均有基因型的亲子对列表：<code>offspring_sire_list</code>  和  <code>offspring_dam_list</code></p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#清洗系谱，剔除第一列个体没有基因型的行</span></span><br><span class="line"></span><br><span class="line">offspring_sire_list = []</span><br><span class="line">offspring_dam_list = []</span><br><span class="line">pedigree_list = []</span><br><span class="line"></span><br><span class="line"><span class="comment">#判断 args.list 是否存在</span></span><br><span class="line">id_set = <span class="built_in">set</span>()</span><br><span class="line"><span class="keyword">if</span> args.<span class="built_in">list</span> <span class="keyword">is</span> <span class="literal">None</span>:</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> pedigree_file:</span><br><span class="line">        f = i.split()</span><br><span class="line">        id_set.add(f[<span class="number">0</span>])</span><br><span class="line"><span class="keyword">else</span>:</span><br><span class="line">    list_file = <span class="built_in">open</span>(args.<span class="built_in">list</span>,<span class="string">&#x27;r&#x27;</span>)</span><br><span class="line">    <span class="keyword">for</span> i <span class="keyword">in</span> list_file:</span><br><span class="line">        f = i.split()</span><br><span class="line">        id_set.add(f[<span class="number">0</span>])</span><br><span class="line"></span><br><span class="line">test_records=<span class="number">0</span></span><br><span class="line">sire_pairs=<span class="number">0</span></span><br><span class="line">dam_pairs=<span class="number">0</span></span><br><span class="line">pedigree_file.seek(<span class="number">0</span>)</span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> pedigree_file:</span><br><span class="line">    f = i.split()</span><br><span class="line">    <span class="keyword">if</span> (f[<span class="number">0</span>] <span class="keyword">in</span> gene_id_array_new) <span class="keyword">and</span> (f[<span class="number">0</span>] <span class="keyword">in</span> id_set):</span><br><span class="line">        pedigree_list.append(f[<span class="number">0</span>:<span class="number">3</span>])</span><br><span class="line">        test_records+=<span class="number">1</span></span><br><span class="line">        <span class="keyword">if</span> f[<span class="number">1</span>] <span class="keyword">in</span> gene_id_array_new:</span><br><span class="line">            sire_pairs+=<span class="number">1</span></span><br><span class="line">            offspring_sire_list.append([f[<span class="number">0</span>],f[<span class="number">1</span>]])</span><br><span class="line">        <span class="keyword">if</span> f[<span class="number">2</span>] <span class="keyword">in</span> gene_id_array_new:</span><br><span class="line">            dam_pairs+=<span class="number">1</span></span><br><span class="line">            offspring_dam_list.append([f[<span class="number">0</span>],f[<span class="number">2</span>]])</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="亲子鉴定-2">亲子鉴定</h3>
<p>这里的思路很简单，就是上面提到了，计算两个个体基因型的乘积，统计乘积等于3的位点数目，除以位点总数（剔除缺失个体后），便得到了孟德尔错误率，再与给定的阈值进行比对。如果小于阈值，则认为是正确的亲子对，反之则认为错误。</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br></pre></td><td class="code"><pre><span class="line">check_dick = &#123;&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment">#np.savetxt(&quot;gene_id_array_new.txt&quot;,gene_id_array_new, fmt=&quot;%s&quot;)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#np.where 结果是一个元组，元组里是一个索引构成的数组，如(array([1, 2]),)</span></span><br><span class="line">conflict_sire_pairs=<span class="number">0</span></span><br><span class="line">conflict_dam_pairs=<span class="number">0</span></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> offspring_sire_list:</span><br><span class="line">    first_index = np.where(gene_id_array_new == i[<span class="number">0</span>])</span><br><span class="line">    second_index = np.where(gene_id_array_new == i[<span class="number">1</span>])</span><br><span class="line">    first_array = array4[first_index]</span><br><span class="line">    second_array = array4[second_index]</span><br><span class="line">    product_array = first_array * second_array</span><br><span class="line">    error_snp_nums = (product_array==<span class="number">3</span>).<span class="built_in">sum</span>()</span><br><span class="line">    miss_snp_nums =  (product_array==<span class="number">0</span>).<span class="built_in">sum</span>()</span><br><span class="line">    total_snp_nums = snp_num4 - miss_snp_nums</span><br><span class="line">    error_snp_prob = error_snp_nums/total_snp_nums</span><br><span class="line">    <span class="keyword">if</span> error_snp_prob &lt; <span class="built_in">float</span>(args.check_prob):</span><br><span class="line">        match_status = <span class="string">&quot;Match&quot;</span></span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        match_status = <span class="string">&quot;No-Match&quot;</span></span><br><span class="line">        conflict_sire_pairs+=<span class="number">1</span></span><br><span class="line">    check_dick[<span class="built_in">tuple</span>(i)] = match_status</span><br><span class="line">    check_prob.write(<span class="string">&#x27;Offspring-Sire&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+i[<span class="number">0</span>]+<span class="string">&#x27;\t&#x27;</span>+i[<span class="number">1</span>]+<span class="string">&#x27;\t&#x27;</span>+<span class="built_in">str</span>(error_snp_nums)+<span class="string">&#x27;\t&#x27;</span>+<span class="built_in">str</span>(total_snp_nums)+<span class="string">&#x27;\t&#x27;</span>+<span class="built_in">str</span>(error_snp_prob)+<span class="string">&#x27;\t&#x27;</span>+match_status+<span class="string">&#x27;\n&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> offspring_dam_list:</span><br><span class="line">    first_index = np.where(gene_id_array_new == i[<span class="number">0</span>])</span><br><span class="line">    second_index = np.where(gene_id_array_new == i[<span class="number">1</span>])</span><br><span class="line">    first_array = array4[first_index]</span><br><span class="line">    second_array = array4[second_index]</span><br><span class="line">    product_array = first_array * second_array</span><br><span class="line">    error_snp_nums = (product_array==<span class="number">3</span>).<span class="built_in">sum</span>()</span><br><span class="line">    miss_snp_nums =  (product_array==<span class="number">0</span>).<span class="built_in">sum</span>()</span><br><span class="line">    total_snp_nums = snp_num4 - miss_snp_nums</span><br><span class="line">    error_snp_prob = error_snp_nums/total_snp_nums</span><br><span class="line">    <span class="keyword">if</span> error_snp_prob &lt; <span class="built_in">float</span>(args.check_prob):</span><br><span class="line">        match_status = <span class="string">&quot;Match&quot;</span></span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        match_status = <span class="string">&quot;No-Match&quot;</span></span><br><span class="line">        conflict_dam_pairs+=<span class="number">1</span></span><br><span class="line">    check_dick[<span class="built_in">tuple</span>(i)] = match_status</span><br><span class="line">    check_prob.write(<span class="string">&#x27;Offspring-Dam&#x27;</span>+<span class="string">&#x27;\t&#x27;</span>+i[<span class="number">0</span>]+<span class="string">&#x27;\t&#x27;</span>+i[<span class="number">1</span>]+<span class="string">&#x27;\t&#x27;</span>+<span class="built_in">str</span>(error_snp_nums)+<span class="string">&#x27;\t&#x27;</span>+<span class="built_in">str</span>(total_snp_nums)+<span class="string">&#x27;\t&#x27;</span>+<span class="built_in">str</span>(<span class="string">&quot;&#123;:.4f&#125;&quot;</span>.<span class="built_in">format</span>(error_snp_prob))+<span class="string">&#x27;\t&#x27;</span>+match_status+<span class="string">&#x27;\n&#x27;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> pedigree_list:</span><br><span class="line">    offspring_sire = (i[<span class="number">0</span>],i[<span class="number">1</span>])</span><br><span class="line">    offspring_dam = (i[<span class="number">0</span>],i[<span class="number">2</span>])</span><br><span class="line">    <span class="keyword">if</span> offspring_sire <span class="keyword">in</span> check_dick:</span><br><span class="line">        sire_info = check_dick[offspring_sire]</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        sire_info = <span class="string">&quot;par_nogeno&quot;</span></span><br><span class="line">    <span class="keyword">if</span> offspring_dam <span class="keyword">in</span> check_dick:</span><br><span class="line">        dam_info = check_dick[offspring_dam]</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        dam_info = <span class="string">&quot;par_nogeno&quot;</span></span><br><span class="line">    check_pedigree.write(<span class="string">&#x27;\t&#x27;</span>.join(i)+<span class="string">&#x27;\t&#x27;</span>+sire_info+<span class="string">&#x27;\t&#x27;</span>+dam_info+<span class="string">&#x27;\n&#x27;</span>)</span><br><span class="line">    </span><br><span class="line"></span><br><span class="line">raw_file.close()</span><br><span class="line">chipid_file.close()</span><br><span class="line">pedigree_file.close()</span><br><span class="line"><span class="comment">#new_raw.close()</span></span><br><span class="line">check_prob.close()</span><br><span class="line">check_pedigree.close()</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Records tested: &#123;&#125;\n&quot;</span>.<span class="built_in">format</span>(test_records))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  Pair parent/progeny tested: &#123;&#125;\n&quot;</span>.<span class="built_in">format</span>(sire_pairs+dam_pairs))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;  Pair with conflicts: &#123;&#125;\n\n&quot;</span>.<span class="built_in">format</span>(conflict_sire_pairs+conflict_dam_pairs))</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Sire-progeny tested: &#123;&#125;\n&quot;</span>.<span class="built_in">format</span>(sire_pairs))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Sire-progeny with conflicts: &#123;&#125;\n\n&quot;</span>.<span class="built_in">format</span>(conflict_sire_pairs))</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Dam-progeny tested: &#123;&#125;\n&quot;</span>.<span class="built_in">format</span>(dam_pairs))</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;Dam-progeny with conflicts: &#123;&#125;\n\n&quot;</span>.<span class="built_in">format</span>(conflict_dam_pairs))</span><br><span class="line"></span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;End paternity check\n&quot;</span>)</span><br><span class="line"></span><br><span class="line">end_time = time.strftime(<span class="string">&quot;%Y-%m-%d %H:%M:%S&quot;</span>, time.localtime())</span><br><span class="line"><span class="built_in">print</span>(<span class="string">&quot;\nFinish time: &#123;&#125;\n&quot;</span>.<span class="built_in">format</span>(end_time)) </span><br></pre></td></tr></table></figure>
<h1>不足</h1>
<ol>
<li>
<p>质控过程缺少了重复样本检验（基因型相同的样本）</p>
</li>
<li>
<p>亲子鉴定阈值需要通过经验给定，如果阈值设定不好，可能无法准确地将正确的亲子对和错误的亲子对分开，可能会造成错误鉴定。</p>
</li>
<li>
<p>后续可以对存在错误的个体查找真实亲本，方法就是对该个体遍历所有的候选亲本。</p>
</li>
<li>
<p>因为实际生产中可能存在采样错误（采的样本/基因型与标记的个体号不一致），因此本流程找到的错误系谱也有可能是由于采样错误导致的（例如这头个体本身系谱可能没错，但是可能采样采错了，不是采的这个个体，那么亲子鉴定肯定显示为错误）。也就是说，亲子鉴定找的错误实际由采样错误+系谱错误两部分组成。</p>
</li>
</ol>
 
      <!-- reward -->
      
    </div>
    

    <!-- copyright -->
    
    <div class="declare">
      <ul class="post-copyright">
        <li>
          <i class="ri-copyright-line"></i>
          <strong>版权声明： </strong>
          
          本博客所有文章除特别声明外，著作权归作者所有。转载请注明出处！
          
        </li>
      </ul>
    </div>
    
    <footer class="article-footer">
       
<div class="share-btn">
      <span class="share-sns share-outer">
        <i class="ri-share-forward-line"></i>
        分享
      </span>
      <div class="share-wrap">
        <i class="arrow"></i>
        <div class="share-icons">
          
          <a class="weibo share-sns" href="javascript:;" data-type="weibo">
            <i class="ri-weibo-fill"></i>
          </a>
          <a class="weixin share-sns wxFab" href="javascript:;" data-type="weixin">
            <i class="ri-wechat-fill"></i>
          </a>
          <a class="qq share-sns" href="javascript:;" data-type="qq">
            <i class="ri-qq-fill"></i>
          </a>
          <a class="douban share-sns" href="javascript:;" data-type="douban">
            <i class="ri-douban-line"></i>
          </a>
          <!-- <a class="qzone share-sns" href="javascript:;" data-type="qzone">
            <i class="icon icon-qzone"></i>
          </a> -->
          
          <a class="facebook share-sns" href="javascript:;" data-type="facebook">
            <i class="ri-facebook-circle-fill"></i>
          </a>
          <a class="twitter share-sns" href="javascript:;" data-type="twitter">
            <i class="ri-twitter-fill"></i>
          </a>
          <a class="google share-sns" href="javascript:;" data-type="google">
            <i class="ri-google-fill"></i>
          </a>
        </div>
      </div>
</div>

<div class="wx-share-modal">
    <a class="modal-close" href="javascript:;"><i class="ri-close-circle-line"></i></a>
    <p>扫一扫，分享到微信</p>
    <div class="wx-qrcode">
      <img src="//api.qrserver.com/v1/create-qr-code/?size=150x150&data=http://yoursite.com/posts/42752926/" alt="微信分享二维码">
    </div>
</div>

<div id="share-mask"></div>  
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/" rel="tag">数据分析</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF/" rel="tag">生物信息</a></li></ul>

    </footer>
  </div>

   
  <nav class="article-nav">
    
      <a href="/posts/428cb22d/" class="article-nav-link">
        <strong class="article-nav-caption">上一篇</strong>
        <div class="article-nav-title">
          
            Numpy和Pandas模块学习
          
        </div>
      </a>
    
    
      <a href="/posts/e5954c2f/" class="article-nav-link">
        <strong class="article-nav-caption">下一篇</strong>
        <div class="article-nav-title">shell编程</div>
      </a>
    
  </nav>

   
<!-- valine评论 -->
<div id="vcomments-box">
  <div id="vcomments"></div>
</div>
<script src="//cdn1.lncld.net/static/js/3.0.4/av-min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/valine@1.4.14/dist/Valine.min.js"></script>
<script>
  new Valine({
    el: "#vcomments",
    app_id: "yHN3kf7fHt5wvleM2DVoHLdY-gzGzoHsz",
    app_key: "RPIwmdftljIzOtAULwc7JCAp",
    path: window.location.pathname,
    avatar: "monsterid",
    placeholder: "靓仔，看完留个评论再走哇！\n只需要填入昵称和邮箱就可以了",
    recordIP: true,
  });
  const infoEle = document.querySelector("#vcomments .info");
  if (infoEle && infoEle.childNodes && infoEle.childNodes.length > 0) {
    infoEle.childNodes.forEach(function (item) {
      item.parentNode.removeChild(item);
    });
  }
</script>
<style>
  #vcomments-box {
    padding: 5px 30px;
  }

  @media screen and (max-width: 800px) {
    #vcomments-box {
      padding: 5px 0px;
    }
  }

  #vcomments-box #vcomments {
    background-color: #fff;
  }

  .v .vlist .vcard .vh {
    padding-right: 20px;
  }

  .v .vlist .vcard {
    padding-left: 10px;
  }
</style>

 
   
     
</article>

</section>
      <footer class="footer">
  <div class="outer">
    <ul>
      <li>
        Copyrights &copy;
        2019-2021
        <i class="ri-heart-fill heart_icon"></i> Vincere Zhou
      </li>
    </ul>
    <ul>
      <li>
        
        
        <span>
  <span><i class="ri-user-3-fill"></i>访问人数:<span id="busuanzi_value_site_uv"></span></s>
  <span class="division">|</span>
  <span><i class="ri-eye-fill"></i>浏览次数:<span id="busuanzi_value_page_pv"></span></span>
</span>
        
      </li>
    </ul>
    <ul>
      
    </ul>
    <ul>
      
    </ul>
    <ul>
      <li>
        <!-- cnzz统计 -->
        
      </li>
    </ul>

    <!-- 与只只在一起天数 -->
	<ul>
		<li><span id="lovetime_span"></span></li>
	</ul>
    <script type="text/javascript">			
        function show_runtime() {
            window.setTimeout("show_runtime()", 1000);
            X = new Date("03/04/2021 22:11:00");
            Y = new Date();
            T = (Y.getTime() - X.getTime());
            M = 24 * 60 * 60 * 1000;
            a = T / M;
            A = Math.floor(a);
            b = (a - A) * 24;
            B = Math.floor(b);
            c = (b - B) * 60;
            C = Math.floor((b - B) * 60);
            D = Math.floor((c - C) * 60);
            lovetime_span.innerHTML = "只只和男朋友在一起了 " + A + "天" + B + "小时" + C + "分" + D + "秒"
        }
        show_runtime();
    </script>

  </div>
</footer>
      <div class="float_btns">
        <div class="totop" id="totop">
  <i class="ri-arrow-up-line"></i>
</div>

<div class="todark" id="todark">
  <i class="ri-moon-line"></i>
</div>

      </div>
    </main>
    <aside class="sidebar on">
      <button class="navbar-toggle"></button>
<nav class="navbar">
  
  <div class="logo">
    <a href="/"><img src="/images/mojie.jpg" alt="VincereZhou&#39;s blog"></a>
  </div>
  
  <ul class="nav nav-main">
    
    <li class="nav-item">
      <a class="nav-item-link" href="/">主页</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/archives">归档</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/categories">分类</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/tags">标签</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/friends">友链</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/posts/ac7827ff">我的只只</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/about">关于我</a>
    </li>
    
  </ul>
</nav>
<nav class="navbar navbar-bottom">
  <ul class="nav">
    <li class="nav-item">
      
      <a class="nav-item-link nav-item-search"  title="搜索">
        <i class="ri-search-line"></i>
      </a>
      
      
      <a class="nav-item-link" target="_blank" href="/atom.xml" title="RSS Feed">
        <i class="ri-rss-line"></i>
      </a>
      
    </li>
  </ul>
</nav>
<div class="search-form-wrap">
  <div class="local-search local-search-plugin">
  <input type="search" id="local-search-input" class="local-search-input" placeholder="Search...">
  <div id="local-search-result" class="local-search-result"></div>
</div>
</div>
    </aside>
    <script>
      if (window.matchMedia("(max-width: 768px)").matches) {
        document.querySelector('.content').classList.remove('on');
        document.querySelector('.sidebar').classList.remove('on');
      }
    </script>
    <div id="mask"></div>

<!-- #reward -->
<div id="reward">
  <span class="close"><i class="ri-close-line"></i></span>
  <p class="reward-p"><i class="ri-cup-line"></i>请我喝杯咖啡吧~</p>
  <div class="reward-box">
    
    <div class="reward-item">
      <img class="reward-img" src="/images/alipay.jpg">
      <span class="reward-type">支付宝</span>
    </div>
    
    
    <div class="reward-item">
      <img class="reward-img" src="/images/weixinpay.jpg">
      <span class="reward-type">微信</span>
    </div>
    
  </div>
</div>
    
<script src="/js/jquery-2.0.3.min.js"></script>


<script src="/js/lazyload.min.js"></script>

<!-- Tocbot -->


<script src="/js/tocbot.min.js"></script>

<script>
  tocbot.init({
    tocSelector: '.tocbot',
    contentSelector: '.article-entry',
    headingSelector: 'h1, h2, h3, h4, h5, h6',
    hasInnerContainers: true,
    scrollSmooth: true,
    scrollContainer: 'main',
    positionFixedSelector: '.tocbot',
    positionFixedClass: 'is-position-fixed',
    fixedSidebarOffset: 'auto'
  });
</script>

<script src="https://cdn.jsdelivr.net/npm/jquery-modal@0.9.2/jquery.modal.min.js"></script>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/jquery-modal@0.9.2/jquery.modal.min.css">
<script src="https://cdn.jsdelivr.net/npm/justifiedGallery@3.7.0/dist/js/jquery.justifiedGallery.min.js"></script>

<script src="/dist/main.js"></script>

<!-- ImageViewer -->

<!-- Root element of PhotoSwipe. Must have class pswp. -->
<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

    <!-- Background of PhotoSwipe. 
         It's a separate element as animating opacity is faster than rgba(). -->
    <div class="pswp__bg"></div>

    <!-- Slides wrapper with overflow:hidden. -->
    <div class="pswp__scroll-wrap">

        <!-- Container that holds slides. 
            PhotoSwipe keeps only 3 of them in the DOM to save memory.
            Don't modify these 3 pswp__item elements, data is added later on. -->
        <div class="pswp__container">
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
        </div>

        <!-- Default (PhotoSwipeUI_Default) interface on top of sliding area. Can be changed. -->
        <div class="pswp__ui pswp__ui--hidden">

            <div class="pswp__top-bar">

                <!--  Controls are self-explanatory. Order can be changed. -->

                <div class="pswp__counter"></div>

                <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>

                <button class="pswp__button pswp__button--share" style="display:none" title="Share"></button>

                <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>

                <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>

                <!-- Preloader demo http://codepen.io/dimsemenov/pen/yyBWoR -->
                <!-- element will get class pswp__preloader--active when preloader is running -->
                <div class="pswp__preloader">
                    <div class="pswp__preloader__icn">
                        <div class="pswp__preloader__cut">
                            <div class="pswp__preloader__donut"></div>
                        </div>
                    </div>
                </div>
            </div>

            <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
                <div class="pswp__share-tooltip"></div>
            </div>

            <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
            </button>

            <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
            </button>

            <div class="pswp__caption">
                <div class="pswp__caption__center"></div>
            </div>

        </div>

    </div>

</div>

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/default-skin/default-skin.min.css">
<script src="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe-ui-default.min.js"></script>

<script>
    function viewer_init() {
        let pswpElement = document.querySelectorAll('.pswp')[0];
        let $imgArr = document.querySelectorAll(('.article-entry img:not(.reward-img)'))

        $imgArr.forEach(($em, i) => {
            $em.onclick = () => {
                // slider展开状态
                // todo: 这样不好，后面改成状态
                if (document.querySelector('.left-col.show')) return
                let items = []
                $imgArr.forEach(($em2, i2) => {
                    let img = $em2.getAttribute('data-idx', i2)
                    let src = $em2.getAttribute('data-target') || $em2.getAttribute('src')
                    let title = $em2.getAttribute('alt')
                    // 获得原图尺寸
                    const image = new Image()
                    image.src = src
                    items.push({
                        src: src,
                        w: image.width || $em2.width,
                        h: image.height || $em2.height,
                        title: title
                    })
                })
                var gallery = new PhotoSwipe(pswpElement, PhotoSwipeUI_Default, items, {
                    index: parseInt(i)
                });
                gallery.init()
            }
        })
    }
    viewer_init()
</script>

<!-- MathJax -->

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
      tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
      }
  });

  MathJax.Hub.Queue(function() {
      var all = MathJax.Hub.getAllJax(), i;
      for(i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
      }
  });
</script>

<script src="https://cdn.jsdelivr.net/npm/mathjax@2.7.6/unpacked/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script>
  var ayerConfig = {
    mathjax: true
  }
</script>

<!-- Katex -->

<!-- busuanzi  -->


<script src="/js/busuanzi-2.3.pure.min.js"></script>


<!-- ClickLove -->

<!-- ClickBoom1 -->

<!-- ClickBoom2 -->

<!-- CodeCopy -->


<link rel="stylesheet" href="/css/clipboard.css">

<script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js"></script>
<script>
  function wait(callback, seconds) {
    var timelag = null;
    timelag = window.setTimeout(callback, seconds);
  }
  !function (e, t, a) {
    var initCopyCode = function(){
      var copyHtml = '';
      copyHtml += '<button class="btn-copy" data-clipboard-snippet="">';
      copyHtml += '<i class="ri-file-copy-2-line"></i><span>COPY</span>';
      copyHtml += '</button>';
      $(".highlight .code pre").before(copyHtml);
      $(".article pre code").before(copyHtml);
      var clipboard = new ClipboardJS('.btn-copy', {
        target: function(trigger) {
          return trigger.nextElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        let $btn = $(e.trigger);
        $btn.addClass('copied');
        let $icon = $($btn.find('i'));
        $icon.removeClass('ri-file-copy-2-line');
        $icon.addClass('ri-checkbox-circle-line');
        let $span = $($btn.find('span'));
        $span[0].innerText = 'COPIED';
        
        wait(function () { // 等待两秒钟后恢复
          $icon.removeClass('ri-checkbox-circle-line');
          $icon.addClass('ri-file-copy-2-line');
          $span[0].innerText = 'COPY';
        }, 2000);
      });
      clipboard.on('error', function(e) {
        e.clearSelection();
        let $btn = $(e.trigger);
        $btn.addClass('copy-failed');
        let $icon = $($btn.find('i'));
        $icon.removeClass('ri-file-copy-2-line');
        $icon.addClass('ri-time-line');
        let $span = $($btn.find('span'));
        $span[0].innerText = 'COPY FAILED';
        
        wait(function () { // 等待两秒钟后恢复
          $icon.removeClass('ri-time-line');
          $icon.addClass('ri-file-copy-2-line');
          $span[0].innerText = 'COPY';
        }, 2000);
      });
    }
    initCopyCode();
  }(window, document);
</script>


<!-- CanvasBackground -->


    
  </div>
<script src="/live2dw/lib/L2Dwidget.min.js?094cbace49a39548bed64abff5988b05"></script><script>L2Dwidget.init({"pluginRootPath":"live2dw/","pluginJsPath":"lib/","pluginModelPath":"assets/","tagMode":false,"debug":false,"model":{"jsonPath":"live2d-widget-model-wanko"},"display":{"position":"left","width":150,"height":300,"hOffset":80,"vOffset":-70},"mobile":{"show":false,"scale":0.5},"log":false});</script></body>

</html>