<!DOCTYPE html>


<html lang="zh-CN">


<head>
  <meta charset="utf-8" />
    
  <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1" />
  <title>
    æ¨¡å‹å­¦ä¹ ç¬”è®° |  VincereZhou&#39;s blog
  </title>
  <meta name="generator" content="hexo-theme-ayer">
  
  <link rel="shortcut icon" href="/images/mojie.jpg" />
  
  
<link rel="stylesheet" href="/dist/main.css">

  <link rel="stylesheet" href="https://cdn.jsdelivr.net/gh/Shen-Yu/cdn/css/remixicon.min.css">
  
<link rel="stylesheet" href="/css/custom.css">

  
  <script src="https://cdn.jsdelivr.net/npm/pace-js@1.0.2/pace.min.js"></script>
  
  

  

<link rel="alternate" href="/atom.xml" title="VincereZhou's blog" type="application/atom+xml">
</head>

</html>

<body>
  <div id="app">
    
      
    <main class="content on">
      <section class="outer">
  <article
  id="post-æ¨¡å‹å­¦ä¹ ç¬”è®°"
  class="article article-type-post"
  itemscope
  itemprop="blogPost"
  data-scroll-reveal
>
  <div class="article-inner">
    
    <header class="article-header">
       
<h1 class="article-title sea-center" style="border-left:0" itemprop="name">
  æ¨¡å‹å­¦ä¹ ç¬”è®°
</h1>
 

    </header>
     
    <div class="article-meta">
      <a href="/posts/9664406d/" class="article-date">
  <time datetime="2021-04-25T06:30:25.000Z" itemprop="datePublished">2021-04-25</time>
</a> 
  <div class="article-category">
    <a class="article-category-link" href="/categories/%E7%90%86%E8%AE%BA%E5%AD%A6%E4%B9%A0/">ç†è®ºå­¦ä¹ </a> / <a class="article-category-link" href="/categories/%E7%90%86%E8%AE%BA%E5%AD%A6%E4%B9%A0/%E7%BA%BF%E6%80%A7%E6%A8%A1%E5%9E%8B/">çº¿æ€§æ¨¡å‹</a>
  </div>
  
<div class="word_count">
    <span class="post-time">
        <span class="post-meta-item-icon">
            <i class="ri-quill-pen-line"></i>
            <span class="post-meta-item-text"> å­—æ•°ç»Ÿè®¡:</span>
            <span class="post-count">22.9k</span>
        </span>
    </span>

    <span class="post-time">
        &nbsp; | &nbsp;
        <span class="post-meta-item-icon">
            <i class="ri-book-open-line"></i>
            <span class="post-meta-item-text"> é˜…è¯»æ—¶é•¿â‰ˆ</span>
            <span class="post-count">99 åˆ†é’Ÿ</span>
        </span>
    </span>
</div>
 
    </div>
      
    <div class="tocbot"></div>




  
    <div class="article-entry" itemprop="articleBody">
       
  <p>çº¿æ€§æ¨¡å‹å­¦ä¹ çš„å¤§æ‚çƒ©ï¼Œä¹‹åå†æ‹†åˆ†ã€‚</p>
<span id="more"></span>
<h1>Linear Models for the Prediction of Animal Breeding Values</h1>
<h2 id="ç¬¬ä¸€ç« -åŸºç¡€">ç¬¬ä¸€ç«  åŸºç¡€</h2>
<h3 id="é€šè¿‡ä¸ªä½“è‡ªèº«è®°å½•ä¼°è®¡è‚²ç§å€¼">é€šè¿‡ä¸ªä½“è‡ªèº«è®°å½•ä¼°è®¡è‚²ç§å€¼</h3>
<p><img src="single_record_by_own_performance.png" alt=""></p>
<blockquote>
<p>The corrlelation between the selection criterion, in this case the phenotypic value, and the true breeding value is known as the accuracy of prediction.</p>
</blockquote>
<p>è¿™æ®µè¯æ˜¯è¯´åœ¨è¿™ç§æƒ…å†µä¸‹ï¼Œè¡¨å‹å’Œè‚²ç§å€¼çš„ç›¸å…³ä¾¿æ˜¯é¢„æµ‹å‡†ç¡®æ€§ï¼ˆthe accuracy of predictionï¼‰ã€‚ä½†æ˜¯è¿™é‡Œçš„è‚²ç§å€¼ç”¨çš„å°±æ˜¯the true breeding valueï¼Œä¹Ÿå°±æ˜¯å¹³å¸¸è¯´çš„ä¼°è®¡è‚²ç§å€¼ï¼Ÿ<br>
è€ŒæŸäº›æƒ…å†µä¸‹ï¼Œå¯¹å‡†ç¡®æ€§è¿›è¡Œå¹³æ–¹ï¼Œç§°ä¸ºå¯é æ€§ï¼ˆreliability or repeatability,r^2ï¼‰ï¼Œå³ç›¸å…³ç³»æ•°çš„å¹³æ–¹ã€‚</p>
<h1>Computing Strategies for Solving Mixed Model Equations</h1>
<p>å¯¹å¤§é‡æ–¹ç¨‹ç»„çš„æ±‚è§£é€šå¸¸é‡‡ç”¨è¿­ä»£æŠ€æœ¯å› ä¸ºæ•°æ®å¤ªå¤§ï¼Œä½¿å¾—ç›´æ¥æ±‚é€†æ—¶æ— æ³•å…¨éƒ¨å‚¨å­˜å†å†…å­˜ä¸­ã€‚ï¼ˆSolution of a  large set of equations is usually by iteration techniques because they are too large to store in the memory of the com- puter for direct inversionï¼‰<br>
è™½ç„¶BLUPå¯èƒ½æ˜¯æœ€å¥½çš„åˆ†æçº¿æ€§æ¨¡å‹çš„æ–¹æ³•ï¼Œä½†æ˜¯å…¶è®¡ç®—çš„å¤æ‚åº¦ä½¿å¾—ç ”ç©¶è€…éœ€è¦ä½¿ç”¨ç®€å•çš„æ¨¡å‹ï¼Œæˆ–è€…é‡‡ç”¨é™¤BLUPå¤–çš„å…¶ä»–æ–¹æ³•ã€‚å› æ­¤ï¼Œå¯è®¡ç®—æ€§ï¼ˆcomputabilityï¼‰éå¸¸é‡è¦ã€‚<br>
æœ¬æ–‡æ˜¯å•çº¯<strong>ä»ç†è®ºä¸Š</strong>æ¯”è¾ƒä¸åŒçš„è®¡ç®—MMEçš„ç­–ç•¥</p>
<h2 id="æ¨¡å‹">æ¨¡å‹</h2>
<p>ç”¨çš„æ˜¯ç‰›çš„æ•°æ®ï¼Œå›ºå®šæ•ˆåº”ä¸ºåœºå¹´å­£ï¼ˆHYS,herd-year-season)ï¼Œéšæœºæ•ˆåº”ä¸ºå…¬ç‰›æ•ˆåº”ã€‚</p>
<h2 id="è®¡ç®—æ–¹æ³•">è®¡ç®—æ–¹æ³•</h2>
<p>å‡è®¾æœ‰50ä¸‡ä¸ªå¥¶ç‰›è§‚æµ‹å€¼ï¼Œå‡æ¥è‡ªäº 1ä¸‡å¤´å…¬ç‰›ï¼Œ5ä¸‡ä¸ªåœºå¹´å­£æ°´å¹³ã€‚å‡å®šæ¯ä¸ªåœºå¹´å­£æ°´å¹³åŒ…æ‹¬æ¥è‡ªäº6å¤´ä¸åŒçš„å…¬ç‰›çš„10ä¸ªå¥³å„¿æˆç»©ã€‚é‚£ä¹ˆå°±ä¼šæœ‰30ä¸‡ä¸ªå…¬ç‰›Ã—åœºå¹´å­£äºšæ°´å¹³ã€‚<br>
åƒæ•°æ®é‡è¿™ä¹ˆåºå¤§çš„æƒ…å†µä¸‹ï¼Œç›´æ¥æ±‚é€†æ˜¯ä¸å¯èƒ½çš„ã€‚</p>
<h2 id="ç¬¬ä¸€ç§æ–¹æ³•-â€“-ç›´æ¥æ–¹æ³•-Straightforward-Approach">ç¬¬ä¸€ç§æ–¹æ³• â€“ ç›´æ¥æ–¹æ³•(Straightforward  Approach)</h2>
<p>å°†ç³»æ•°çŸ©é˜µçš„æ‰€æœ‰é0å€¼å­˜å‚¨åœ¨ç¡¬ç›˜ä¸­ï¼Œæ¯ä¸€è¡Œä¸ºæ‰€åœ¨è¡Œidï¼Œæ‰€åœ¨åˆ—idï¼Œå€¼ã€‚(æ„Ÿè§‰ç±»ä¼¼ä¸ºDMUè·‘å‡ºçš„Gé˜µ)ã€‚<br>
<strong>y = Wh + Zs + e</strong><br>
MMEä¸­<strong>Wâ€™W</strong>æ˜¯åŒ…å«äº†50,000ä¸ªéé›¶å…ƒç´ çš„å¯¹è§’çº¿çŸ©é˜µï¼ˆå…¶ä¸­çš„å€¼ä¸ºæ¯ä¸ªåœºå¹´å­£æ°´å¹³çš„é‡å¤æ•°ï¼‰ã€‚<strong>Zâ€™Z</strong>ä¹Ÿæ˜¯åŒ…å«äº†10,000ä¸ªéé›¶å…ƒç´ çš„å¯¹è§’çº¿çŸ©é˜µï¼ˆåœ¨åŠ¨ç‰©æ¨¡å‹ä¸­ï¼Œå¯¹è§’çº¿å…ƒç´ åªæœ‰0å’Œ1ï¼Œè¡¨ç¤ºæœ‰æ— è¡¨å‹ï¼‰ã€‚<strong>Wâ€™Z</strong>å’Œ<strong>Zâ€™W</strong>äº’ä¸ºè½¬ç½®çŸ©é˜µï¼Œä¸”å„åŒ…å«300,000ä¸ªéé›¶å€¼ï¼ˆå®é™…ä¸ºHYSÃ—sireçš„äºšæ°´å¹³çš„é‡å¤æ•°ï¼‰ã€‚é‚£ä¹ˆç³»æ•°çŸ©é˜µå°±æ€»å…±æœ‰600,000ä¸ªéé›¶å…ƒç´ ã€‚Any iterative solution program must read each element once during each round of iteration. Conveniently, unless there are covariates in the model, these nonzero elements are integer variables, but the time to solve the equations depends on the speed with  which  the 660,000 records can be retrieved from disk.ï¼ˆè¿™ä¸€æ®µå¤§æ¦‚è¯´ï¼Œæ¯ä¸€æ¬¡è¿­ä»£éƒ½è¦è¯»å–æ¯ä¸€ä¸ªæ•°å­—ï¼Œå› æ­¤ä»ç¡¬ç›˜è¯»å–çš„é€Ÿåº¦å’Œæ•°å­—å¤§å°å°±å½±å“äº†è®¡ç®—é€Ÿåº¦ï¼Œä¸çŸ¥é“æ±‚é€†æ€ä¹ˆè¿­ä»£ï¼Ÿï¼‰ã€‚For example, Wâ€™W and Zâ€™Z are diagonal matrices and the diagonals could be stored in program memory rather than disk, and thus, only 600,000   nonzero elements need to be stored on diskï¼ˆä½œè€…ç»™å‡ºçš„æ„è§æ˜¯ï¼Œå› ä¸ºWâ€™Wå’ŒZâ€™Zæ˜¯å¯¹è§’çº¿å…ƒç´ ï¼Œå› æ­¤å¯ä»¥ç›´æ¥æ”¾åœ¨å†…å­˜ä¸­ï¼Œå…¶ä»–è¿˜æ”¾åœ¨ç¡¬ç›˜ä¸­ã€‚ï¼‰<br>
ç¬¬å››éƒ¨åˆ†è¿˜è¦åŠ ä¸ŠKA^-1^ã€‚</p>
<h2 id="ç¬¬äºŒç§æ–¹æ³•-Absorption-Approach">ç¬¬äºŒç§æ–¹æ³•-- Absorption Approach</h2>
<p>The absorption or elimination technique whereby a large diagonal submatrix can be absorbed into the other equations is frequently used for the solution of large sets of equations. Absorption of HYS equations into sire equations gives Zâ€™MZ and Zâ€™My, where M = I - W(Wâ€™W)^-I^Wâ€™. The resulting equations to solve are<br>
(Zâ€™MZ + A^-1^k)s^ = Zâ€™My (s^å³sä¼°è®¡å€¼ï¼‰<br>
â€¦â€¦<br>
Consequently, even though the absorption approach yields only 10,000 equations, we may have more records to read during each round of iteration, and therefore, <strong>computing costs would increase</strong>. Also, the extra program steps to sort and add together â€œlikeâ€ coefficients is a disadvantage of the absorption approach.<br>
çœ‹ä¸æ‡‚ï¼Œä¸è¿‡çœ‹æœ€åçš„ç»“è¯­ï¼Œè²Œä¼¼è®¡ç®—éš¾åº¦è¿˜æé«˜äº†ã€‚</p>
<h2 id="ç¬¬ä¸‰ç§æ–¹æ³•â€“Indirect-Approach">ç¬¬ä¸‰ç§æ–¹æ³•â€“Indirect Approach</h2>
<p>In this method data are arranged in HYS by sire subclasses sorted by HYS. Each record consists of HYS and sire identification with the corresponding number of daughters and the sum of daughter Observations in the subclass. Thus, we would have 300,000 subclasses for the example.<br>
If Y~Ij.~ is the subclass total and n~ij~ is the number of daughters for the jth sire in the ith<br>
HYS, then we accumulate the quantity, over sires within HYS:<br>
Y~ij.~ - n~ij~s~j~^</p>
<p>where s~j~^ is the current  solution for sire j. Then the new solution for the i^th^ HYS is<br>
â€¦â€¦åé¢çœŸçœ‹ä¸æ‡‚</p>
<h2 id="å°ç»“">å°ç»“</h2>
<p>Prediction error variances of sire evaluations can be approximated in the same manner as<br>
with the absorption approach. The number of effective daughters, or the diagonals of Zâ€™MZ, can be obtained while reading the HYS subclasses,but another vector of length equal to the number of sires would be needed to accumulate the information across HYS.<br>
ï¼ˆ<strong>PEV</strong>å¯ä»¥é€šè¿‡ç¬¬äºŒç§æ–¹æ³•Absorption Approachè¿‘ä¼¼ä¼°è®¡ã€‚ï¼‰<br>
The main advantage of the  indirect approach is that only one relatively simple computer program is necessary to obtain solutions for mixed model equations. At this point,  rate of convergence of the indirect approach compared with other methods is unknown. However, in the sire model  of [1], if the solutions for HYS are close to the mean of records in that HYS, then convergence should be relatively quick because HYS means are the solutions  after  the first iteration. Knowing Iâ€™A^-1^s  should be zero can also speed convergence<br>
(æ²¡çœ‹æ‡‚ï¼Œå¥½åƒæ˜¯è¯´ç¬¬ä¸‰ç§é—´æ¥æ–¹æ³•åªé€‚ç”¨äºæ¨¡å‹ä¸å˜æ—¶ï¼Ÿ)</p>
<h1>Indirect Solution of Mixed Model Equations</h1>
<p>Two iterative procedures were investigated, Gauss-Seidel and Jacobi.With a balanced dara set, putting restrictions on fixed effects is not effective for improving convergence ratcs in GaussSeidel but is essential in Jacobi.</p>
<h2 id="å¼•è¨€">å¼•è¨€</h2>
<p>æ„å»ºMMEéœ€è¦èŠ±å¾ˆé•¿æ—¶é—´ã€‚Schaeffer and Kennedy å±•ç¤ºäº†ä¸€ç§ä¸ç”¨æ„å»ºthe system of equationsçš„ç­–ç•¥ã€‚ä»–ä»¬çš„ç­–ç•¥ï¼Œç§°ä¹‹ä¸ºindirect approach,performs Gauss-Seidel (G-S) or successive overrelaxation (SOR) iteration while reading the data files rather than the matrix of coefficients.The data files, read once per round of iteration, consist of two copies of an observation file, each sorted for a different factor with a large number of levels, and three<br>
copies of a relationship file, also sorted differently.<br>
Misztal (2, 3) proposed another indirect approach using Jacobi (J) iteration (8). Programming<br>
is simpler, because only one unsorted copy of each of the observation and relationship files need to be read per round of iteration.Basic J iteration has a poor convergence rate,especially for models with many factors or with a random factor including t~ numerator relationship matrix. However, putting restrictions on the effects of levels of fixed factors and extending the method to secondÂ·order J made it competitive in running time relative to other methods; convergence rate was approximately half ,haâ€™ of SOR (8)<br>
This paper presents the indirect approach in terms of succesive averaging, investigates its<br>
properties, and describes several computing options.z<br>
è¿˜æ˜¯çœ‹ä¸æ‡‚ï¼Œè¿™ä¸ªçœŸçš„æ˜¯çº¯æ–¹æ³•ã€‚ä¸¤ç§è¿­ä»£æ–¹å¼ã€‚ä¼˜ç‚¹å°±æ˜¯ä¸ç”¨æ„å»ºMMEäº†ã€‚</p>
<h1>python è¿›è¡ŒçŸ©é˜µæ±‚è§£</h1>
<p>å®‹å¸ˆå…„æ˜¯ä»from scipy import linalgï¼Œä½†å…¶å®numpy æœ¬èº«ä¹Ÿæœ‰è¿™ä¸ªlinalg, ä¹Ÿå¯ä»¥å®ç°ç›¸åŒçš„åŠŸèƒ½ã€‚<br>
å®‹å¸ˆå…„python- GBLUP</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># -*- coding: UTF-8 -*-</span></span><br><span class="line"><span class="keyword">import</span> numpy <span class="keyword">as</span> np</span><br><span class="line"><span class="keyword">from</span> pysnptools.snpreader <span class="keyword">import</span> Bed</span><br><span class="line"><span class="keyword">from</span> scipy <span class="keyword">import</span> linalg</span><br><span class="line"><span class="keyword">import</span> pandas <span class="keyword">as</span> pd</span><br><span class="line"></span><br><span class="line"><span class="comment">###æ„å»ºGçŸ©é˜µ###</span></span><br><span class="line"><span class="comment"># PLINK bedæ–‡ä»¶å</span></span><br><span class="line">bed_file = <span class="string">&#x27;plink2&#x27;</span></span><br><span class="line"><span class="comment"># è¯»å–æ•°æ®</span></span><br><span class="line">M = Bed(bed_file, count_A1=<span class="literal">False</span>).read()</span><br><span class="line"><span class="comment"># è®¡ç®—æ¯ä¸ªä½ç‚¹çš„ç­‰ä½åŸºå› é¢‘ç‡</span></span><br><span class="line">freq = np.<span class="built_in">sum</span>(M.val, axis=<span class="number">0</span>) / (<span class="number">2</span> * M.iid_count)</span><br><span class="line"><span class="comment"># è®¡ç®—æ ‡å‡†åŒ–å› å­</span></span><br><span class="line">scale = np.<span class="built_in">sum</span>(<span class="number">2</span> * freq * (<span class="number">1</span> - freq))</span><br><span class="line"><span class="comment"># è®¡ç®—ZçŸ©é˜µ</span></span><br><span class="line">Z = M.val - <span class="number">2</span> * freq</span><br><span class="line"><span class="comment"># è®¡ç®—GçŸ©é˜µ</span></span><br><span class="line">G = np.dot(Z, Z.T) / scale</span><br><span class="line"><span class="comment">#å¯¹è§’çº¿å‘é‡åŠ 0.001</span></span><br><span class="line">d = np.diag(G) + <span class="number">0.001</span></span><br><span class="line"><span class="comment">#æ›¿æ¢åŸæ¥çš„GçŸ©é˜µå¯¹è§’çº¿</span></span><br><span class="line">np.fill_diagonal(G, d)</span><br><span class="line"><span class="comment">#æ±‚é€†</span></span><br><span class="line">Ginv = linalg.inv(G)</span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#è¯»å–æ•°æ®æ–‡ä»¶</span></span><br><span class="line">data_file = <span class="string">&#x27;phe&#x27;</span></span><br><span class="line">data = pd.read_table(data_file, header=<span class="number">0</span>, sep = <span class="string">&#x27;\s+&#x27;</span>)</span><br><span class="line"><span class="built_in">print</span> data.head()</span><br><span class="line"><span class="comment">#æ•°æ®æ–‡ä»¶æœ‰6åˆ—ï¼ŒIDã€meanã€sexã€ageã€ã€pheã€‚</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#è¡¨å‹å‘é‡</span></span><br><span class="line">y = np.array(data[<span class="string">&#x27;phe&#x27;</span>], dtype = np.<span class="built_in">float</span>)</span><br><span class="line"><span class="comment">#å›ºå®šæ•ˆåº”è®¾è®¡çŸ©é˜µ</span></span><br><span class="line">X = np.array(data.loc[:,<span class="string">&#x27;mean&#x27;</span>:<span class="string">&#x27;treat&#x27;</span>], dtype = np.<span class="built_in">float</span>)</span><br><span class="line">	<span class="comment"># data.locä¸­è¡¨ç¤ºï¼Œæå–meanåˆ—åˆ°treatåˆ—çš„æ‰€æœ‰è¡Œ</span></span><br><span class="line">	<span class="comment"># treatæ²¡å‡ºç°åœ¨å‰é¢ï¼Œä¼°è®¡æ˜¯ç¬¬5åˆ—</span></span><br><span class="line">	<span class="comment"># è¿™ä¸ªä¹Ÿä¸ç¬¦åˆå›ºå®šæ•ˆåº”å…³è”çŸ©é˜µçš„å®šä¹‰å•Šï¼Œæä¸æ‡‚</span></span><br><span class="line">	<span class="comment"># åæ­£è¿™é‡Œçš„Xé˜µï¼Œåº”è¯¥ä¸æ˜¯ç†è®ºä¸Šçš„å›ºå®šæ•ˆåº”è®¾è®¡çŸ©é˜µï¼Œå®ƒçš„è¡Œæ•°å’Œè¡¨å‹æ•°ç›®ä¸€è‡´ï¼Œæ‰€ä»¥ä¸‹æ–‡ä¸­ï¼Œæ‰èƒ½å¯¹Xé˜µ,Zé˜µæŒ‰åˆ—è¿›è¡Œåˆå¹¶ï¼Œå½¢æˆä»€ä¹ˆWé˜µã€‚</span></span><br><span class="line"><span class="comment">#éšæœºæ•ˆåº”çŸ©é˜µä¸ºå•ä½é˜µ</span></span><br><span class="line">num = <span class="built_in">len</span>(y)  <span class="comment">#ä¸ªä½“æ•°</span></span><br><span class="line">Z =  np.eye(num) </span><br><span class="line">	<span class="comment">#ä»–è¿™é‡Œçš„ZçŸ©é˜µæ˜¯è¡¨å‹ä¸ªæ•°çš„å•ä½é˜µ</span></span><br><span class="line">	<span class="comment">#æŒ‰å¸¸è§„BLUPï¼ŒZçŸ©é˜µæ˜¯å¯¹è§’é˜µï¼Œä½†æ˜¯æœ‰è¡¨å‹çš„æ˜¯1ï¼Œæ²¡è¡¨å‹çš„æ˜¯0</span></span><br><span class="line">	<span class="comment">#éš¾é“è¿™é‡Œå®‹å¸ˆå…„éƒ½æ˜¯æœ‰è¡¨å‹çš„ä¸ªä½“ï¼Œä¸å¯¹å§ï¼Ÿé‚£é¢„æµ‹å•¥ä¸ªä½“å•Šï¼Ÿ</span></span><br><span class="line"><span class="comment">#WçŸ©é˜µ</span></span><br><span class="line">W = np.concatenate((X, Z), axis = <span class="number">1</span>)</span><br><span class="line">	<span class="comment">#concatenate æ•°ç»„æ‹¼æ¥å‡½æ•° axis=1,æ˜¯æŒ‰åˆ—åˆå¹¶</span></span><br><span class="line">	<span class="comment">#å°±æ˜¯è¯´åˆå¹¶çš„ä¸¤ä¸ªçŸ©é˜µè¦æœ‰ç›¸åŒçš„è¡Œæ•°ã€‚</span></span><br><span class="line">	<span class="comment">#å—¯å—¯ï¼ŒXå’ŒZè¡Œæ•°å°±æ˜¯ç›¸åŒçš„</span></span><br><span class="line"><span class="comment">#æ„å»ºWâ€™R-1Wï¼Œç”±äºæ®‹å·®æ–¹å·®æ˜¯ç‹¬ç«‹åŒåˆ†å¸ƒçš„ï¼Œå¯ä»¥å†™ä¸º</span></span><br><span class="line">C = np.dot(W.T, W)/<span class="number">0.139</span></span><br><span class="line">	<span class="comment">#np.dot æ­£å¸¸çš„çŸ©é˜µç›¸ä¹˜ï¼Œè¿™é‡Œå°±æ˜¯Wâ€˜*Wã€‚</span></span><br><span class="line">	<span class="comment">#0.139ä¸çŸ¥é“ä»€ä¹ˆæ„æ€ï¼Œæ˜¯ä¸æ˜¯æ®‹å·®æ–¹å·®</span></span><br><span class="line">	<span class="comment">#æ‡‚äº†ï¼Œå› ä¸ºæ®‹å·®çŸ©é˜µç­‰äºå•ä½é˜µä¹˜ä»¥æ®‹å·®æ–¹å·®ï¼Œå› æ­¤è¿™é‡Œå¯ä»¥ä½œä¸ºå¸¸æ•°æå‡ºæ¥ã€‚</span></span><br><span class="line">	<span class="comment">#å› ä¸ºæ˜¯Ré€†ï¼Œæ‰€ä»¥æ˜¯1/æ®‹å·®æ–¹å·®ã€‚</span></span><br><span class="line"><span class="comment">#åŠ å…¥éšæœºæ•ˆåº”éƒ¨åˆ†çš„æ–¹å·®åæ–¹å·®çŸ©é˜µçš„é€†ï¼Œæ„å»ºç³»æ•°çŸ©é˜µ</span></span><br><span class="line">C[X.shape[<span class="number">1</span>]:,X.shape[<span class="number">1</span>]:] += Ginv/<span class="number">0.088</span></span><br><span class="line">	<span class="comment">#Cé‡Œé¢æ˜¯Xçš„åˆ—æ•°ï¼Œæ¯”å¦‚å¦‚æœæ˜¯4ï¼Œå°±æ˜¯ä»ç¬¬5è¡Œï¼Œç¬¬5åˆ—å¼€å§‹ã€‚</span></span><br><span class="line">	<span class="comment">#æˆ‘æ‡‚äº†ï¼Œè¿™é‡Œæ˜¯æŠŠğº^(âˆ’1)âˆ•ğœ_ğ‘^2 åŠ ä¸Šæ¥ï¼Œå½¢æˆä¸€ä¸ªå®Œæ•´çš„ç³»æ•°çŸ©é˜µã€‚</span></span><br><span class="line"><span class="comment">#ç³»æ•°çŸ©é˜µå³æ‰‹é¡¹</span></span><br><span class="line">RHS = np.dot(W.T, y)/ <span class="number">0.139</span></span><br><span class="line"><span class="comment">#æ±‚è§£æ··åˆæ¨¡å‹æ–¹ç¨‹ç»„</span></span><br><span class="line">effect = linalg.solve(C, RHS)</span><br><span class="line"></span><br><span class="line"><span class="comment">#æå–æ•ˆåº”</span></span><br><span class="line">fix_eff = effect[:X.shape[<span class="number">1</span>]]  <span class="comment">#å›ºå®šæ•ˆåº”éƒ¨åˆ†</span></span><br><span class="line">ran_eff = effect[X.shape[<span class="number">1</span>]:,] <span class="comment">#éšæœºæ•ˆåº”éƒ¨åˆ†</span></span><br><span class="line"><span class="comment">####ä¸‹é¢ä»‹ç»å¦‚ä½•å¯¹å›ºå®šæ•ˆåº”æ£€éªŒï¼Œå¦‚ä½•æ±‚å¯é æ€§</span></span><br><span class="line"><span class="comment">#ç³»æ•°çŸ©é˜µæ±‚é€†</span></span><br><span class="line">Ci = linalg.inv(C)</span><br><span class="line"><span class="comment">#å›ºå®šæ•ˆåº”æ•´ä½“æ£€éªŒï¼ˆåŒ…æ‹¬å‡å€¼ï¼‰ï¼Œå¤‡æ‹©å‡è®¾ä¸ºè‡³å°‘æœ‰ä¸€ä¸ªæ˜¾è‘—</span></span><br><span class="line">covi = linalg.inv(Ci[:X.shape[<span class="number">1</span>], :X.shape[<span class="number">1</span>]]) <span class="comment">#ä¼°è®¡å€¼æ–¹å·®åæ–¹å·®çŸ©é˜µçš„é€†</span></span><br><span class="line">chi_val = np.dot(fix_eff.T, np.dot(covi, fix_eff)) <span class="comment">#å¡æ–¹å€¼</span></span><br><span class="line"><span class="keyword">from</span> scipy.stats <span class="keyword">import</span> chi2 <span class="comment">#åŠ è½½å¡æ–¹åˆ†å¸ƒçš„åº“</span></span><br><span class="line">p1 = chi2.sf(chi_val, <span class="number">4</span>) <span class="comment">#è½¬åŒ–ä¸ºPå€¼</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#è‚²ç§å€¼çš„å¯é æ€§è®¡ç®—</span></span><br><span class="line">PEV = np.diag(Ci[X.shape[<span class="number">1</span>]:, X.shape[<span class="number">1</span>]:]) <span class="comment">#å–ç³»æ•°çŸ©é˜µéšæœºæ•ˆåº”éƒ¨åˆ†çš„å¯¹è§’çº¿ï¼Œå³ä¸ºPEV</span></span><br><span class="line">R2 = <span class="number">1</span> - PEV/<span class="number">0.088</span></span><br><span class="line"><span class="comment">#è¾“å‡ºå¯é æ€§çš„æœ€å¤§ã€æœ€å°å€¼</span></span><br><span class="line"><span class="built_in">print</span> <span class="built_in">max</span>(R2), <span class="built_in">min</span>(R2)  <span class="comment">#å¯é æ€§æ¯”è¾ƒé«˜ï¼Œå¯èƒ½ç”±äºæ‰€æœ‰ä¸ªä½“éƒ½è¡¨å‹</span></span><br></pre></td></tr></table></figure>
<h1>Efficient Methods to Compute Genomic Predictions â€“ VanRaden 2008</h1>
<h2 id="MçŸ©é˜µï¼ˆ-1ï¼Œ0ï¼Œ1ï¼‰">MçŸ©é˜µï¼ˆ-1ï¼Œ0ï¼Œ1ï¼‰</h2>
<p>MçŸ©é˜µæ˜¯n(ä¸ªä½“æ•°)*mï¼ˆsnpæ•°ï¼‰çš„çŸ©é˜µï¼Œå†…å®¹ä¸º-1ï¼Œ0ï¼Œ1ï¼ˆ-1è¡¨ç¤ºMajor frequentçš„çº¯åˆå­ï¼Œ0è¡¨ç¤ºæ‚åˆå­ï¼Œ1è¡¨ç¤ºMinor frequent çš„çº¯åˆå­ï¼‰ã€‚<br>
MMâ€™ çš„å¯¹è§’çº¿å…ƒç´ ä¸ºæ¯ä¸ªä¸ªä½“çš„çº¯å’Œå­çš„æ•°ç›®ï¼Œéå¯¹è§’çº¿å…ƒç´ æ²¡æœ‰å®é™…æ„ä¹‰ï¼ˆå®é™…ç­‰äºä¸¤ä¸ªä¸ªä½“ç›¸åŒçš„çº¯å’Œå­æ•°ç›®ï¼ˆå¦‚AAä¸AAï¼‰-ä¸åŒçš„çº¯åˆå­æ•°ç›®ï¼ˆAAä¸TTï¼‰,æ‚åˆå­æ²¡å½±å“ï¼‰ï¼ˆoff-diagonals measure the number of alleles shared by relatives.ï¼‰</p>
<p>Mâ€™M çš„å¯¹è§’çº¿è¡¨ç¤ºæ¯ä¸ªä½ç‚¹çº¯åˆå­çš„ä¸ªä½“æ•°ï¼Œéå¯¹è§’çº¿å…ƒç´ æ²¡æœ‰å®é™…æ„ä¹‰ï¼Œè®¡ç®—ç±»ä¸Šã€‚</p>
<h2 id="PçŸ©é˜µï¼ˆ2-p-0-5">PçŸ©é˜µï¼ˆ2*(p-0.5))</h2>
<p>PçŸ©é˜µä¹Ÿæ˜¯n<em>mçš„çŸ©é˜µï¼Œå…¶ä¸­æ¯ä¸€åˆ—çš„å†…å®¹éƒ½æ˜¯ç›¸åŒçš„ï¼ˆå¯¹åº”çš„æ˜¯è¯¥snpçš„2</em>(p-0.5)ï¼‰ã€‚<br>
å…¶ä¸­pçš„è®¡ç®—æ–¹å¼æ˜¯ï¼ˆ(1<em>æ‚åˆå­+2</em>Minor frequentçš„çº¯åˆå­)/2nï¼ˆnä¸ºä¸ªä½“æ•°ï¼‰ï¼‰ã€‚</p>
<h2 id="ZçŸ©é˜µ-ä¸­å¿ƒåŒ–">ZçŸ©é˜µ ä¸­å¿ƒåŒ–</h2>
<p>Z = M - P (which sets mean values of the allele effects to 0.)<br>
æ‡‚äº†ï¼Œæ‰€è°“çŸ©é˜µä¸­å¿ƒåŒ–å°±æ˜¯è¯´ï¼Œå°±æ˜¯å‡å»å‡å€¼ï¼Œä½¿æ•°æ®çš„å‡å€¼çš„ä¸º0ã€‚<br>
è€Œæ ‡å‡†åŒ–çš„è¿‡ç¨‹ï¼Œæ˜¯å‡å»å‡å€¼ï¼Œç„¶åé™¤ä»¥æ ‡å‡†å·®ã€‚ä½¿æ•°æ®æ»¡è¶³ï¼ˆ0ï¼Œ1ï¼‰çš„æ ‡å‡†æ­£æ€åˆ†å¸ƒã€‚<br>
ä¸ºä»€ä¹ˆå‡å»PçŸ©é˜µï¼Œå°±èƒ½å®ç°ä¸­å¿ƒåŒ–å‘¢ï¼Ÿ<br>
ä¸ªäººç†è§£ã€‚<br>
å¦‚æœæ‰€æœ‰åŸºå› æ»¡è¶³<strong>å“ˆæ¸©å¹³è¡¡</strong>ï¼Œé‚£ä¹ˆ</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">D &#x3D; (1-p)^2  # D æŒ‡ä»£çš„æ˜¯åŸºå› é¢‘ç‡å¤§çš„çº¯å’Œå­</span><br><span class="line">H &#x3D; 2\*(1-p)\*(p)</span><br><span class="line">R &#x3D; p^2</span><br><span class="line">ç”±äºD H R åœ¨MçŸ©é˜µä¸­åˆ†åˆ«ä¸º-1,0,1</span><br><span class="line">é‚£ä¹ˆå‡å€¼ä¸º -1\*(1-p)^2 + 0\* 2\*(1-p)\*(p) + 1\*p^2 &#x3D; 2(p-0.5)</span><br><span class="line">ä½†æ˜¯å‰ææ˜¯å¿…é¡»æ»¡è¶³å“ˆæ¸©å¹³è¡¡</span><br></pre></td></tr></table></figure>
<p>æ–‡ç« ä¸­ä¹ŸæåŠäº†é€‰æ‹©çš„ç­‰ä½åŸºå› ä¸èƒ½ç»è¿‡<strong>é€‰æ‹©</strong>ï¼Œè¿™ä¹Ÿå°±æ˜¯ä»ä¾§é¢è¯´è¦æ»¡è¶³å“ˆæ¸©å¹³è¡¡ã€‚</p>
<blockquote>
<p>Allele frequencies  in  P  should  be  from  the  unselected  base  population rather than those that occur after selection or inbreeding.</p>
</blockquote>
<h2 id="Gé˜µ">Gé˜µ</h2>
<p><img src="G_matrix.png" alt="G_matrix"></p>
<p>é™¤ä»¥åˆ†æ¯ï¼Œä½¿å¾—Gé˜µä¸Aé˜µå¯æ¯”ã€‚</p>
<blockquote>
<p>Division by  2(1 )âˆ‘âˆ’ppii scales  G to  be  analogous(å¯æ¯”çš„)  to  the  numerator  relationship  matrix  A.</p>
</blockquote>
<h2 id="å…¶ä»–æ¨å¯¼Gé˜µçš„æ–¹æ³•">å…¶ä»–æ¨å¯¼Gé˜µçš„æ–¹æ³•</h2>
<p>æ²¡å¤ªçœ‹æ‡‚ï¼Œåæ­£å¥½åƒå°±ç¬¬ä¸€ç§ç”¨çš„å¤šã€‚</p>
<h1>model.matrix_R</h1>
<h2 id="åˆ†ç±»å˜é‡ï¼ˆFactorsï¼‰ï¼šRä¸­ç”¨äºå­˜å‚¨åˆ†ç±»æ•°æ®çš„ç±»åˆ«ä¿¡æ¯">åˆ†ç±»å˜é‡ï¼ˆFactorsï¼‰ï¼šRä¸­ç”¨äºå­˜å‚¨åˆ†ç±»æ•°æ®çš„ç±»åˆ«ä¿¡æ¯</h2>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">f = factor(<span class="string">&#x27;a&#x27;</span>,<span class="string">&#x27;b&#x27;</span>,<span class="string">&#x27;a&#x27;</span>,<span class="string">&#x27;c&#x27;</span>)</span><br><span class="line"><span class="built_in">class</span>(f) <span class="comment"># &quot;factor&quot;</span></span><br></pre></td></tr></table></figure>
<h2 id="å“‘å˜é‡">å“‘å˜é‡</h2>
<p>è™šæ‹Ÿå˜é‡/å“‘å˜é‡ï¼ˆdummy variable) ï¼šé‡åŒ–éæ•°å€¼ç±»å‹çš„å˜é‡ï¼Œé€šå¸¸å–0/1ã€‚ä¾‹å¦‚ä¸€ä¸ªäººçš„æ€§åˆ«ã€‚</p>
<h2 id="è§£é‡Šå˜é‡">è§£é‡Šå˜é‡</h2>
<p>è§£é‡Šå˜é‡ï¼ˆexplanatory variable) : ç­‰åŒäºæ§åˆ¶å˜é‡/è‡ªå˜é‡ï¼Œä¸ä¹‹ç›¸å¯¹çš„æ˜¯<strong>è¢«è§£é‡Šå˜é‡</strong>ï¼ˆååº”å˜é‡/å› å˜é‡ï¼‰ã€‚</p>
<h2 id="è®¾è®¡çŸ©é˜µ">è®¾è®¡çŸ©é˜µ</h2>
<p>è®¾è®¡çŸ©é˜µï¼ˆdesign matrixï¼‰: åˆå«<strong>æ¨¡å‹çŸ©é˜µ</strong>ï¼ˆmodel matrix) æˆ–è€… <strong>å›å½’çŸ©é˜µ</strong>ï¼ˆregressor matrix) ã€‚ ç”±è§£é‡Šå˜é‡å€¼ç»„æˆçš„çŸ©é˜µï¼šä¸€è¡Œä»£è¡¨ä¸€ä¸ªç‹¬ç«‹çš„è§‚æµ‹å¯¹è±¡ï¼ˆæ ·æœ¬ï¼‰ï¼Œä¸€åˆ—ä»£è¡¨å¯¹åº”çš„å˜é‡ï¼ˆç‰¹å¾å€¼ã€å…ƒæ•°æ®ï¼‰ï¼Œé€šå¸¸è®°ä¸ºXã€‚ç®€å•ç†è§£ï¼Œå°±æ˜¯æˆ‘ä»¬æ‰€è¯´çš„è¾“å…¥çŸ©é˜µï¼Œå¯ä»¥æ˜¯å…ƒæ•°æ®çš„ï¼Œä¹Ÿå¯ä»¥æ˜¯æ•°æ®çš„ã€‚</p>
<h1>PH525x series - Biomedical Data Science ç½‘é¡µ</h1>
<h2 id="choice-of-design">choice of design</h2>
<p>The choice of design matrix is a critical step in linear modeling since it encodes which coefficients will be fit in the model, as well as the inter-relationship between the samples. A common misunderstanding is that the choice of design follows straightforward from a description of which samples were included in the experiment. This is not the case. The basic information about each sample (whether control or treatment group, experimental batch, etc.) does not imply a single â€˜correctâ€™ design matrix. The design matrix additionally encodes various assumptions about how the variables in X explain the observed values in Y, on which the investigator must decide. (è¿™ä¸€æ®µéƒ½çœ‹ä¸å¤ªæ‡‚ï¼Œå¤§æ¦‚æ„æ€æ˜¯ç³»æ•°çŸ©é˜µçš„æ„å»ºéœ€è¦æ„å»ºè€…è‡ªèº«çŸ¥é“ï¼Œå“ªäº›<strong>X</strong>ä¸­çš„å› ç´ èƒ½å¤Ÿè§£é‡Š<strong>Y</strong>ä¸­çš„è§‚æµ‹å€¼ã€‚)</p>
<p>model.matrixä¸­ä½¿ç”¨~å·è¡¨ç¤ºå¼€å§‹ä¸€ä¸ªå…¬å¼ï¼ˆformulaï¼‰,è¿™ä¸ªç¬¦å·å³è¾¹çš„å˜é‡ï¼ŒæŒ‡ç¤ºæ¯ä¸ªä¸ªä½“åœ¨å“ªä¸ªç¾¤ä½“ä¸­ï¼ˆwhich tells us which samples are in which groupï¼‰ã€‚</p>
<h2 id="æˆªè·ï¼ˆInterceptï¼‰çš„å«ä¹‰">æˆªè·ï¼ˆInterceptï¼‰çš„å«ä¹‰</h2>
<p>For the examples we cover here, we use linear models to make comparisons between different groups. Hence, the design matrices that we ultimately work with will have at least two columns: an intercept column, which consists of a column of 1â€™s, and a second column, which specifies which samples are in a second group. In this case, two coefficients are fit in the linear model: <strong>the intercept, which represents the population average of the first group</strong>, and a second coefficient, which represents the difference between the population averages of the second group and the first group. The latter is typically the coefficient we are interested in when we are performing statistical tests: we want to know if their is a difference between the two groups.</p>
<p>ä»è¿™æ®µç”»å¯ä»¥çœ‹å‡ºï¼Œæˆªè·æ˜¯æœ‰å«ä¹‰çš„ã€‚è¿™é‡Œçš„æ¨¡å‹æ˜¯Yä¸ºè€é¼ çš„ä½“é‡ï¼Œè‡ªå˜é‡åªæœ‰ä¸€ä¸ªï¼Œé‚£å°±æ˜¯æ˜¯å¦æ¥å—äº†é«˜è„‚è‚ªé£Ÿç‰©ï¼ˆåªæœ‰ä¸¤ç§çŠ¶æ€0å’Œ1ï¼Œ0è¡¨ç¤ºæ²¡æœ‰ï¼Œ1è¡¨ç¤ºæ¥å—äº†é«˜è„‚è‚ªé£Ÿç‰©ï¼‰ã€‚è¿™é‡Œçš„æˆªè·Î²å°±æ˜¯æŒ‡å½“è€é¼ æ²¡æœ‰æ¥å—é«˜è„‚è‚ªé£Ÿç‰©æ—¶çš„å¹³å‡ä½“é‡ï¼Œè€ŒÎ²1æŒ‡çš„æ˜¯æ¥å—é«˜è„‚è‚ªé£Ÿç‰©åä½“é‡çš„å¹³å‡å˜åŒ–é‡ã€‚</p>
<p>ç®€å•çš„è¯´ï¼Œä¹Ÿå°±æ˜¯æˆªè·è¡¨ç¤ºçš„<strong>groupçš„ä¸€ç§çŠ¶æ€ï¼ˆéšä¾¿é‚£ç§è®¾ä¸ºåŸºæœ¬çŠ¶æ€ï¼ŒRé‡Œå¥½åƒæ˜¯æŠŠç¢°åˆ°çš„ç¬¬ä¸€ä¸ªåˆ†ç±»ç»™æˆªè·ï¼‰<strong>æ—¶çš„å‡å€¼ã€‚ä¹Ÿå°±æ˜¯æˆªè·å°±è¡¨ç¤ºäº†ä¸€ç§çŠ¶æ€ï¼ˆçŠ¶æ€0ï¼‰ã€‚å…¶ä»–çš„groupçš„å€¼æ˜¯åœ¨è¿™ä¸ªåŸºç¡€çŠ¶æ€çš„</strong>å˜åŒ–é‡</strong>ã€‚æ¯”å¦‚æ­¤æ—¶Î²1å°±æ˜¯æ¥å—é«˜è„‚è‚ªé£Ÿç‰©ç›¸è¾ƒäºæ²¡æœ‰æ¥å—çš„ä½“é‡å¢åŠ é‡ã€‚</p>
<p>å¦‚æœä¸è®¾æˆªè·ï¼Œä½¿ç”¨~group-1ï¼Œæ­¤æ—¶å¦‚æœå¦‚æœY = Î²x + eã€‚æ¯ä¸ªÎ²ï¼ˆä¸¤ç§çŠ¶æ€ï¼‰éƒ½æ˜¯<strong>ç»å¯¹å€¼</strong>ã€‚</p>
<p><img src="Design_Matrix1.png" alt=""></p>
<h2 id="ç»„åˆ«ä¸º2">ç»„åˆ«ä¸º2</h2>
<p>Letâ€™s try an example. Suppose we have two groups, control and high fat diet, with two samples each. For illustrative purposes, we will code these with 1 and 2 respectively. <strong>We should first tell R that these values should not be interpreted numerically, but as different levels of a factor.</strong> ï¼ˆåœ¨Rä¸­ï¼Œè®¾è®¡çŸ©é˜µé‚£ä¸€åˆ—å¿…é¡»è®¾å®šä¸º<strong>factor</strong>ï¼‰</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">group &lt;- factor(<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>,<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line">model.matrix(~group)</span><br><span class="line"><span class="comment">#model.matrix(formula(~ group)) #å¯ä»¥çœç•¥formula()</span></span><br><span class="line"><span class="comment">## ç»“æœå¦‚ä¸‹ï¼Œå¦‚æœä¸å†™-1ï¼Œé‚£ä¹ˆå°±ä¼šæœ‰ä¸€ä¸ªæˆªè·</span></span><br><span class="line"><span class="comment">## å› ä¸ºå› å­åŒ–äº†ï¼Œè¿™é‡Œçš„groupå°±å¥½åƒå˜æˆ0å’Œ1äº†ã€‚</span></span><br><span class="line"> (Intercept) group2</span><br><span class="line">1           <span class="number">1</span>      <span class="number">0</span></span><br><span class="line">2           <span class="number">1</span>      <span class="number">0</span></span><br><span class="line">3           <span class="number">1</span>      <span class="number">1</span></span><br><span class="line">4           <span class="number">1</span>      <span class="number">1</span></span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;assign&quot;</span>)</span><br><span class="line">[<span class="number">1</span>] <span class="number">0</span> <span class="number">1</span></span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;contrasts&quot;</span>)</span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;contrasts&quot;</span>)$group</span><br><span class="line">[<span class="number">1</span>] <span class="string">&quot;contr.treatment&quot;</span></span><br></pre></td></tr></table></figure>
<h3 id="group-1">group - 1</h3>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br></pre></td><td class="code"><pre><span class="line">group &lt;- factor(<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>,<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line">model.matrix(~group -<span class="number">1</span> )</span><br><span class="line"><span class="comment">#model.matrix(~group +0 )</span></span><br><span class="line"><span class="comment"># ~group - 1(+0) è¿™æ ·å°±æ²¡æœ‰åŸºç¡€å€¼Î²0äº†ã€‚</span></span><br><span class="line"><span class="comment"># This group now fits a separate coefficient for each group.</span></span><br><span class="line">  group1 group2</span><br><span class="line">1      <span class="number">1</span>      <span class="number">0</span></span><br><span class="line">2      <span class="number">1</span>      <span class="number">0</span></span><br><span class="line">3      <span class="number">0</span>      <span class="number">1</span></span><br><span class="line">4      <span class="number">0</span>      <span class="number">1</span></span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;assign&quot;</span>)</span><br><span class="line">[<span class="number">1</span>] <span class="number">1</span> <span class="number">1</span></span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;contrasts&quot;</span>)</span><br><span class="line"><span class="built_in">attr</span>(,<span class="string">&quot;contrasts&quot;</span>)$group</span><br><span class="line">[<span class="number">1</span>] <span class="string">&quot;contr.treatment&quot;</span></span><br><span class="line"></span><br><span class="line">å¦‚æœæˆ‘ä»¬ä¸è®¾å®šgroupä¸º**å› å­**çš„è¯</span><br><span class="line"></span><br><span class="line">```R</span><br><span class="line">group &lt;- c(1,1,2,2)</span><br><span class="line">model.matrix(~group)</span><br><span class="line"></span><br><span class="line">## ç»“æœï¼Œè¿™é‡Œgroupæ˜¯1å’Œ2ï¼Œä¸ç¬¦åˆè®¾è®¡çŸ©é˜µå…ƒç´ åªæœ‰0å’Œ1</span><br><span class="line">## æ‰€è°“è®¾è®¡çŸ©é˜µï¼Œåªæ˜¯ä¸€ä¸ª*indicator*,æŒ‡ç¤ºä¸ªä½“æ‰€å±çš„groupï¼Œå› æ­¤åªæœ‰0å’Œ1ï¼Œè€Œä¸”ä¸€åˆ—åªæœ‰ä¸€ä¸ª1ï¼Œå…¶ä»–éƒ½æ˜¯0ã€‚</span><br><span class="line">## æ‰€ä»¥ï¼Œmodel.matrixæ‰€åœ¨åˆ—çš„å®é™…å†…å®¹æ— å…³ç´§è¦ï¼Œæ˜¯æ•°å­—ä¹Ÿå¥½ï¼Œå­—ç¬¦ä¸²ä¹Ÿå¥½ã€‚éƒ½æ²¡å½±å“ï¼Œéƒ½åªæ˜¯åˆ†ç±»å˜é‡ï¼Œè®¾è®¡çŸ©é˜µåªæ˜¯æç¤ºè¿™ä¸ªä¸ªä½“å±äºå“ªä¸ªåˆ†ç±»ï¼ˆgroupï¼‰</span><br><span class="line">## å› æ­¤ï¼Œmodel.matrixå‰å¿…é¡»è¿›è¡Œå› å­åŒ–ã€‚</span><br><span class="line">  (Intercept) group</span><br><span class="line">1           1     1</span><br><span class="line">2           1     1</span><br><span class="line">3           1     2</span><br><span class="line">4           1     2</span><br><span class="line">attr(,&quot;assign&quot;)</span><br><span class="line">[1] 0 1</span><br></pre></td></tr></table></figure>
<p>å°†åˆ†ç±»åˆ—çš„å†…å®¹æ¢æˆå­—ç¬¦ä¸²ï¼Œç»“æœç›¸åŒ</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">group &lt;- factor(<span class="built_in">c</span>(<span class="string">&quot;control&quot;</span>,<span class="string">&quot;control&quot;</span>,<span class="string">&quot;highfat&quot;</span>,<span class="string">&quot;highfat&quot;</span>))</span><br><span class="line">model.matrix(~ group)</span><br><span class="line"></span><br><span class="line"><span class="comment">##   (Intercept) grouphighfat</span></span><br><span class="line"><span class="comment">## 1           1            0</span></span><br><span class="line"><span class="comment">## 2           1            0</span></span><br><span class="line"><span class="comment">## 3           1            1</span></span><br><span class="line"><span class="comment">## 4           1            1</span></span><br><span class="line"><span class="comment">## attr(,&quot;assign&quot;)</span></span><br><span class="line"><span class="comment">## [1] 0 1</span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)</span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)$group</span></span><br><span class="line"><span class="comment">## [1] &quot;contr.treatment&quot;</span></span><br></pre></td></tr></table></figure>
<h3 id="More-groups">More groups</h3>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">group &lt;- factor(<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>,<span class="number">2</span>,<span class="number">2</span>,<span class="number">3</span>,<span class="number">3</span>))</span><br><span class="line">model.matrix(~ group)</span><br><span class="line"><span class="comment">##   (Intercept) group2 group3</span></span><br><span class="line"><span class="comment">## 1           1      0      0</span></span><br><span class="line"><span class="comment">## 2           1      0      0</span></span><br><span class="line"><span class="comment">## 3           1      1      0</span></span><br><span class="line"><span class="comment">## 4           1      1      0</span></span><br><span class="line"><span class="comment">## 5           1      0      1</span></span><br><span class="line"><span class="comment">## 6           1      0      1</span></span><br><span class="line"><span class="comment">## attr(,&quot;assign&quot;)</span></span><br><span class="line"><span class="comment">## [1] 0 1 1</span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)</span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)$group</span></span><br><span class="line"><span class="comment">## [1] &quot;contr.treatment&quot;</span></span><br></pre></td></tr></table></figure>
<p>Now we have a third column which specifies which samples belong to the third group.ï¼ˆç°åœ¨æˆ‘ä»¬æœ‰äº†ç¬¬ä¸‰åˆ—ï¼Œæ¥è¡¨ç¤ºç¬¬ä¸‰ç§groupï¼‰</p>
<h2 id="æ›´å¤šçš„å˜é‡">æ›´å¤šçš„å˜é‡</h2>
<p>è¿™é‡Œåªè€ƒè™‘äº†dietï¼Œå®é™…æƒ…å†µä¸­ä¸€èˆ¬ä¼šæœ‰æ›´å¤šçš„å˜é‡ã€‚åœ¨è¿™é‡Œï¼Œæˆ‘ä»¬å¢åŠ äº†<strong>æ€§åˆ«</strong>å˜é‡ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">diet &lt;- factor(<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">1</span>,<span class="number">2</span>,<span class="number">2</span>,<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line">sex &lt;- factor(<span class="built_in">c</span>(<span class="string">&quot;f&quot;</span>,<span class="string">&quot;f&quot;</span>,<span class="string">&quot;m&quot;</span>,<span class="string">&quot;m&quot;</span>,<span class="string">&quot;f&quot;</span>,<span class="string">&quot;f&quot;</span>,<span class="string">&quot;m&quot;</span>,<span class="string">&quot;m&quot;</span>))</span><br><span class="line">table(diet,sex)</span><br><span class="line"><span class="comment">##     sex</span></span><br><span class="line"><span class="comment">## diet f m</span></span><br><span class="line"><span class="comment">##    1 2 2</span></span><br><span class="line"><span class="comment">##    2 2 2</span></span><br></pre></td></tr></table></figure>
<p>å¦‚æœæˆ‘ä»¬å‡è®¾æ—¥ç²®æ•ˆåº”å¯¹äºå…¬é¼ å’Œæ¯é¼ çš„æ•ˆåº”æ˜¯ç›¸åŒçš„ï¼ˆä¹Ÿå°±æ˜¯äºŒè€…æ²¡æœ‰äº’ä½œï¼‰ã€‚é‚£ä¹ˆçº¿æ€§æ¨¡å‹ä¸º<br>
Y~i~ = Î²~0~ + Î²~1~x~i,1~ + Î²~2~x~2,i~ + e~i~</p>
<p>é‚£ä¹ˆï¼Œmodel.matrixè®¾ç½®å¦‚ä¸‹</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">diet &lt;- factor(c(1,1,1,1,2,2,2,2))</span><br><span class="line">sex &lt;- factor(c(&quot;f&quot;,&quot;f&quot;,&quot;m&quot;,&quot;m&quot;,&quot;f&quot;,&quot;f&quot;,&quot;m&quot;,&quot;m&quot;))</span><br><span class="line">model.matrix(~ diet + sex)</span><br><span class="line">##   (Intercept) diet2 sexm</span><br><span class="line">## 1           1     0    0</span><br><span class="line">## 2           1     0    0</span><br><span class="line">## 3           1     0    1</span><br><span class="line">## 4           1     0    1</span><br><span class="line">## 5           1     1    0</span><br><span class="line">## 6           1     1    0</span><br><span class="line">## 7           1     1    1</span><br><span class="line">## 8           1     1    1</span><br><span class="line">## attr(,&quot;assign&quot;)</span><br><span class="line">## [1] 0 1 2</span><br><span class="line">## attr(,&quot;contrasts&quot;)</span><br><span class="line">## attr(,&quot;contrasts&quot;)$diet</span><br><span class="line">## [1] &quot;contr.treatment&quot;</span><br><span class="line">## </span><br><span class="line">## attr(,&quot;contrasts&quot;)$sex</span><br><span class="line">## [1] &quot;contr.treatment&quot;</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>å¦‚æœè€ƒè™‘äº’ä½œæ•ˆåº”</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">model.matrix(~ diet + sex + diet:sex)</span><br><span class="line"><span class="comment">#model.matrix(~ diet*sex)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##   (Intercept) diet2 sexm diet2:sexm</span></span><br><span class="line"><span class="comment">## 1           1     0    0          0</span></span><br><span class="line"><span class="comment">## 2           1     0    0          0</span></span><br><span class="line"><span class="comment">## 3           1     0    1          0</span></span><br><span class="line"><span class="comment">## 4           1     0    1          0</span></span><br><span class="line"><span class="comment">## 5           1     1    0          0</span></span><br><span class="line"><span class="comment">## 6           1     1    0          0</span></span><br><span class="line"><span class="comment">## 7           1     1    1          1</span></span><br><span class="line"><span class="comment">## 8           1     1    1          1</span></span><br><span class="line"><span class="comment">## attr(,&quot;assign&quot;)</span></span><br><span class="line"><span class="comment">## [1] 0 1 2 3</span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)</span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)$diet</span></span><br><span class="line"><span class="comment">## [1] &quot;contr.treatment&quot;</span></span><br><span class="line"><span class="comment">## </span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)$sex</span></span><br><span class="line"><span class="comment">## [1] &quot;contr.treatment&quot;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="Releveling-è®¾ç½®å‚è€ƒæ°´å¹³ï¼Œreference-level">Releveling (è®¾ç½®å‚è€ƒæ°´å¹³ï¼Œ<em>reference level</em>)</h3>
<p>The level which is chosen for the reference level is the level which is contrasted against. By default, this is simply the first level alphabetically. We can specify that we want group 2 to be the reference level by either using the <em>relevel</em> function:<br>
ç¿»è¯‘ä¸€ä¸‹ï¼Œæ‰€è°“reference levelï¼ˆå‚è€ƒæ°´å¹³ï¼Œä¹Ÿå°±æ˜¯æˆªè·çš„æ°´å¹³ï¼‰ï¼Œå³æ˜¯å…¶ä»–æ°´å¹³ä¸ä¹‹ç›¸æ¯”è¾ƒçš„æ°´å¹³ã€‚é»˜è®¤æƒ…å†µä¸‹ï¼Œå‚è€ƒæ°´å¹³å°±æ˜¯<strong>åç§°æŒ‰å­—æ¯æ’åº</strong>çš„ç¬¬ä¸€ä¸ªæ°´å¹³ã€‚æˆ‘ä»¬å¯ä»¥é€šè¿‡<strong>relevel</strong>å‡½æ•°è®¾å®šreference levelã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">group &lt;- factor(<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>,<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line">group &lt;- relevel(group, <span class="string">&quot;2&quot;</span>)</span><br><span class="line">model.matrix(~ group)</span><br><span class="line"><span class="comment">##   (Intercept) group1</span></span><br><span class="line"><span class="comment">## 1           1      1</span></span><br><span class="line"><span class="comment">## 2           1      1</span></span><br><span class="line"><span class="comment">## 3           1      0</span></span><br><span class="line"><span class="comment">## 4           1      0</span></span><br><span class="line"><span class="comment">## attr(,&quot;assign&quot;)</span></span><br><span class="line"><span class="comment">## [1] 0 1</span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)</span></span><br><span class="line"><span class="comment">## attr(,&quot;contrasts&quot;)$group</span></span><br><span class="line"><span class="comment">## [1] &quot;contr.treatment&quot;</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h2 id="è¿ç»­å˜é‡">è¿ç»­å˜é‡</h2>
<p>In this chapter, we focus on models based on indicator values. In certain designs, however, we will be interested in using numeric variables in the design formula, as opposed to converting them to factors first. For example, in the falling object example, time was a continuous variable in the model and time squared was also included:<br>
ç¿»è¯‘ä¸€ä¸‹ï¼Œå°±æ˜¯ä¸€èˆ¬è®¾è®¡çŸ©é˜µéƒ½æ˜¯ç”¨äºåˆ†ç±»å˜é‡ï¼Œå½“æœ‰æ—¶æˆ‘ä»¬ä¹Ÿä¼šæ„Ÿå…´è¶£ä¸€äº›è¿ç»­å˜é‡ã€‚æ¯”å¦‚åœ¨è½ä½“è¿åŠ¨ï¼Œæ—¶é—´å’Œæ—¶é—´çš„å¹³æ–¹å°±åº”è¯¥è€ƒè™‘è¿›å…¥æ¨¡å‹ä¸­ã€‚<br>
è¿ç»­å˜é‡ç”¨äºè®¾è®¡çŸ©é˜µï¼Œè¡¨ç¤ºçš„æ˜¯ä¸€ç§<strong>å‰‚é‡æ•ˆåº”</strong>ã€‚æ­¤æ—¶æ±‚è§£å‡ºæ¥çš„Î²è¡¨ç¤ºçš„æ˜¯æ¯å¢åŠ ä¸€å•ä½çš„è¿ç»­å˜é‡ï¼Œyçš„å¹³å‡æ”¹å˜é‡ã€‚<br>
è¿™ä¹Ÿå°±æ˜¯å¸¦æ¥äº†ä¸€ç§è¿ç»­å˜é‡åº”ç”¨çš„å‡è®¾æˆ–å‰æï¼Œä½ å¾—çŸ¥é“è¿™ä¸ªå˜é‡å¯¹ååº”å˜é‡çš„å½±å“æ˜¯<strong>çº¿æ€§</strong>çš„ã€‚ä¸ç„¶å¾—å‡ºçš„ç»“æœæ˜¯é”™çš„ã€‚<br>
ä»¥ä¸‹è‹±æ–‡å…·ä½“é˜è¿°äº†è¿™ç§è§‚ç‚¹ã€‚ä¾‹å¦‚è½ä½“è·ç¦»ä¸æ—¶é—´çš„å…³ç³»ï¼Œå› ä¸ºæœ‰ç»å…¸çš„ç‰©ç†çš„å…¬å¼ï¼Œæ‰€ä»¥æˆ‘ä»¬ç¡®ä¿¡å®ƒä»¬ä¹‹é—´æ˜¯<strong>çº¿æ€§å…³ç³»</strong>ã€‚ä½†æ˜¯å¯¹äºå…¶ä»–å˜é‡ï¼Œå¦‚å¹´é¾„ç­‰å¯èƒ½ä¸æ˜¯çº¿æ€§å…³ç³»çš„å˜é‡ï¼Œæˆ‘ä»¬ä¸é¼“åŠ±è¿™ä¹ˆä½¿ç”¨ã€‚<br>
The assumptions imposed by including continuous data as variables are typically hard to defend and motivate than the indicator function variables. Why the indicator variables simply assume a different mean between two groups, continuous variables assume a very specific relationship between the outcome and predictor variables.</p>
<p>In cases like the falling object, we have the theory of gravitation supporting the model. In the father-son height example, because the data is bi variate normal, it follows that there is a linear relationship if we condition. However, we find that continuous variables are included in linear models without justification to â€œadjustâ€ for variables such as age. We highly discourage this practice unless the data support the model being used.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># I()ç”¨äºåœ¨æ¨¡å‹ä¸­è¿›è¡Œæ•°å­¦è½¬æ¢æ—¶ã€‚</span></span><br><span class="line"><span class="comment">#å› ä¸ºåœ¨æ¨¡å‹çš„å®šä¹‰ä¸­ï¼Œå¸¸è§çš„ç¬¦å·+ï¼ˆä¿©å˜é‡ï¼‰ã€*ï¼ˆè€ƒè™‘äº’ä½œï¼‰ã€ï¼šï¼ˆäº’ä½œï¼‰ã€^(ä¸æ¸…æ¥š)æœ‰äº†ç‰¹æ®Šå«ä¹‰ï¼Œä¸å†æ˜¯åŸæ¥çš„æ•°å­¦è¿ç®—ç¬¦å·ã€‚</span></span><br><span class="line"><span class="comment">#^ç¬¦å·è²Œä¼¼æ˜¯è€ƒè™‘äº’ä½œï¼Ÿ åœ¨æ¨¡å‹ä¸­ï¼Œï¼ˆx+y)^2 è¡¨ç¤ºçš„æ˜¯ x + y + x:y</span></span><br><span class="line"><span class="comment">#?fromula ä¸­The ^ operator indicates crossing to the specified degree. For example (a+b+c)^2 is identical to (a+b+c)*(a+b+c) which in turn expands to a formula containing the main effects for a, b and c together with their second-order interactions. </span></span><br><span class="line"><span class="comment">#å³(a+b+c)^2 è¡¨ç¤º a + b + c + a:b + a:c + b:cã€‚å—¯ï¼Œå¯¹çš„</span></span><br><span class="line"><span class="comment"># -1 æˆ– +0 éƒ½è¡¨ç¤ºé€šè¿‡åŸç‚¹ï¼Œå³æ²¡æœ‰æˆªè·ã€‚</span></span><br><span class="line"><span class="comment">#å˜é‡ä¿æŠ¤å‡½æ•°ï¼ˆI( )ï¼‰ï¼šæ”¾åœ¨Iå†…çš„è¡¨è¾¾å¼æŒ‰ç…§ç®—æœ¯æ„ä¹‰è¿›è¡Œè§£é‡Š</span></span><br><span class="line">	<span class="comment">#a+b è¡¨ç¤ºå˜é‡aå’Œå˜é‡béƒ½åŒ…å«åœ¨å…¬å¼ä¸­</span></span><br><span class="line">	<span class="comment">#I(a+bï¼‰è¡¨ç¤º a+b è¿™ä¸ªå’Œåº”è¯¥åŒ…å«åœ¨å…¬å¼ä¸­</span></span><br><span class="line"><span class="comment"># Rä¸­å…³äºæ¨¡å‹è®¾å®šçš„é—®é¢˜ï¼Œæ‰¾?formula.</span></span><br><span class="line">tt &lt;- seq(<span class="number">0</span>,<span class="number">3.4</span>,len=<span class="number">4</span>) </span><br><span class="line">model.matrix(~ tt + I(tt^<span class="number">2</span>))</span><br><span class="line"></span><br><span class="line"><span class="comment">##   (Intercept)       tt   I(tt^2)</span></span><br><span class="line"><span class="comment">## 1           1 0.000000  0.000000</span></span><br><span class="line"><span class="comment">## 2           1 1.133333  1.284444</span></span><br><span class="line"><span class="comment">## 3           1 2.266667  5.137778</span></span><br><span class="line"><span class="comment">## 4           1 3.400000 11.560000</span></span><br><span class="line"><span class="comment">## attr(,&quot;assign&quot;)</span></span><br><span class="line"><span class="comment">## [1] 0 1 2</span></span><br></pre></td></tr></table></figure>
<h1>MMEæ„å»ºä¸æ±‚è§£-é‚“é£</h1>
<p>setwd(â€˜D:\Desktop\R_modelâ€™)<br>
ped &lt;- read.table(â€˜example_ped.txtâ€™,header = F)<br>
dat &lt;- read.table(â€˜example_dat.txtâ€™,header = F)</p>
<h2 id="å¤„ç†ç³»è°±">å¤„ç†ç³»è°±</h2>
<p>library(nadiv)<br>
pped = prepPed(ped)#ç»™foundersä¹Ÿå¢åŠ ä¸€è¡Œç³»è°±ï¼Œç„¶åæ’åºç³»è°±<br>
pped</p>
<h1>è¾“å…¥ç³»è°±æ–‡ä»¶ç¼ºå¤±å€¼ä¸ºNA,0,*</h1>
<h1>è¾“å‡ºæ–‡ä»¶ç¼ºå¤±å€¼å…¨æœªNA</h1>
<h1>A = as.matrix(makeA(pped))</h1>
<h1>A</h1>
<h1>A_inv = solve(A) #solve(A) å³æ±‚é€†</h1>
<h1>A_inv</h1>
<p>A_inv = makeAinv(pped)<img src="https://math.now.sh?inline=Ainv%23%E5%8F%AF%E7%9B%B4%E6%8E%A5%E6%B1%82A%E9%80%86%0AA_inv%0A%23%E7%BB%93%E6%9E%9C%E4%B8%BAlist%0A%23" style="display:inline-block;margin: 0;"/>Ainvä¸ºç¨€ç–çŸ©é˜µ<br>
#$listAinvä¸ºä¸‰åˆ—çš„æ ¼å¼<br>
#fä¸ºè¿‘äº¤ç³»æ•°ï¼Œä¸è®¡ç®—<br>
#logDet the log determinant of the A matrix<br>
#dii the (non-zero) elements of the diagonal D matrix of the A=TDTâ€™ decomposition.</p>
<h2 id="ä¿®è¦å¯¹Aé€†æŒ‰ç…§ä¸€å®šé¡ºåºé‡æ’">ä¿®è¦å¯¹Aé€†æŒ‰ç…§ä¸€å®šé¡ºåºé‡æ’</h2>
<p>A_inv = as.matrix(A_inv)<br>
#rownames(A_inv) = as.numeric(rownames(A_inv))<br>
#A_invçš„è¡Œåä¸ºå­—ç¬¦ä¸²ï¼Œè€Œä¸”ä¸èƒ½æ”¹æˆæ•°å­—ï¼Œä¸çŸ¥é“æ˜¯ä¸ºä»€ä¹ˆã€‚</p>
<p>###ä½¿ç”¨z_phe A_invä¸éœ€è¦é‡æ’<br>
colnames(A_inv) = rownames(A_inv)<br>
new_rowname &lt;- sort(rownames(A_inv))<br>
new_rowname<br>
A_inv<br>
A_inv_new &lt;- A_inv[new_rowname,new_rowname]<br>
#è¿™é‡Œç”¨1:8ï¼Œæˆ–è€…seq(â€˜1â€™,â€˜8â€™)ï¼Œéƒ½ä¸è¡Œï¼Œå› ä¸ºæ˜¯æ•°å­—<br>
#è€Œå®é™…çš„è¡Œåæˆ–åˆ—åæ˜¯å­—ç¬¦ä¸²ã€‚<br>
A_inv_new # ç°åœ¨ä¸Z_allå°±ä¸€è‡´äº†ã€‚</p>
<p>str(dat)<br>
colnames(dat) &lt;- c(â€˜idâ€™,â€˜sexâ€™,â€˜wwgâ€™) #ä¸ªä½“å·-æ€§åˆ«-åˆç”Ÿé‡<br>
dat<img src="https://math.now.sh?inline=id%20%3C-%20as.factor%28dat" style="display:inline-block;margin: 0;"/>id)<br>
dat<img src="https://math.now.sh?inline=sex%20%3C-%20as.factor%28dat" style="display:inline-block;margin: 0;"/>sex)<br>
X &lt;- model.matrix(~sex-1,data = dat) #å’Œä¹¦ä¸Šä¸€è‡´<br>
Z_phe &lt;- model.matrix(~id-1,data = dat) # 5*5çš„å•ä½é˜µï¼Œä¸€è‡´<br>
zero_matrix &lt;- matrix(rep(0,15),nrow = 5)<br>
Z_all &lt;- cbind(zero_matrix,Z_phe)<br>
Z_all # Z_all ä¸ºæ·»åŠ äº†ä¸‰åˆ—çš„æ‰€æœ‰ä¸ªä½“çš„ZçŸ©é˜µã€‚</p>
<p>y = as.vector(dat$wwg)</p>
<h3 id="Z-all-ç»“æœ">Z_all ç»“æœ</h3>
<p>alpha = 2 # alphaè®¾ç½®ä¸º2</p>
<p>XpX=crossprod(X) #Xâ€™X</p>
<p>XpZ=crossprod(X,Z_all) #Xâ€™Z</p>
<p>ZpX=crossprod(Z_all,X) #Zâ€™X</p>
<p>ZpZ=crossprod(Z_all) #Zâ€™Z</p>
<p>Xpy=crossprod(X,y) #Xâ€™y</p>
<p>Zpy=crossprod(Z_all,y) #Zâ€™y</p>
<p>LHS=rbind(cbind(XpX,XpZ),cbind(ZpX,ZpZ+A_inv_new*alpha)) #LHS<br>
LHS<br>
LHS_inv1 = solve(LHS)<br>
LHS_inv1</p>
<p>RHS=rbind(Xpy,Zpy) #ä¸éœ€è¦<br>
RHS</p>
<p>sol=solve(LHS)%*%RHS #MMEæ±‚è§£<br>
sol</p>
<h3 id="Z-phe-ç»“æœ">Z_phe ç»“æœ</h3>
<p>a = as.character(dat$id)<br>
A_inv2 = A_inv_new[a,a] #æˆªå–4ï¼š8ä¸ªä½“çš„é€†é˜µ<br>
A_inv2</p>
<p>XpX=crossprod(X) #Xâ€™X</p>
<p>XpZ=crossprod(X,Z_phe) #Xâ€™Z</p>
<p>ZpX=crossprod(Z_phe,X) #Zâ€™X</p>
<p>ZpZ=crossprod(Z_phe) #Zâ€™Z</p>
<p>Xpy=crossprod(X,y) #Xâ€™y</p>
<p>Zpy=crossprod(Z_phe,y) #Zâ€™y</p>
<p>LHS=rbind(cbind(XpX,XpZ),cbind(ZpX,ZpZ+A_inv2*alpha)) #LHS<br>
LHS ##LHSå’ŒZ_allä¸€è‡´<br>
LHS_inv2 = solve(LHS)<br>
LHS_inv2 #ä½†æ˜¯é€†çŸ©é˜µçš„ç»“æœå°±ä¸ä¸€è‡´äº†ã€‚</p>
<p>RHS=rbind(Xpy,Zpy) #ä¸éœ€è¦<br>
RHS #å³æ‰‹é¡¹ä¹Ÿæ˜¯å»é™¤äº†ç›¸åº”çš„è¡Œè€Œå·²ï¼Œå…¶ä»–ä¸€è‡´</p>
<p>sol=solve(LHS)%*%RHS #MMEæ±‚è§£<br>
sol<br>
#ç»“æœç•¥æœ‰ä¸åŒã€‚</p>
<p>#å°ç»“<br>
#è¿˜æ˜¯åº”è¯¥ç”¨æ‰€æœ‰ä¸ªä½“ï¼Œæ›´æ–¹ä¾¿ã€‚ä¹Ÿæ›´ç¬¦åˆå®é™…è¯„ä¼°çš„æƒ…å†µã€‚<br>
#ä½¿ç”¨å…¨éƒ¨ä¸ªä½“åŠæœ‰è¡¨å‹ä¸ªä½“ï¼Œè·å¾—çš„LHSæ˜¯ä¸€æ ·çš„<br>
#ä»…ä»…æ˜¯æœ‰è¡¨å‹ä¸ªä½“çš„LHSï¼Œä»å…¨éƒ¨ä¸ªä½“çš„LHSä¸­åˆ é™¤äº†æ— è¡¨å‹ä¸ªä½“çš„è¡Œå’Œåˆ—ã€‚</p>
<p>#é—®é¢˜<br>
#ZçŸ©é˜µå’ŒAé€†çš„æ„å»ºé—®é¢˜ã€‚<br>
#ZçŸ©é˜µå’Œè¡¨å‹æœ‰å…³ï¼Œåˆå’ŒAé€†æœ‰å…³ã€‚<br>
#è¿˜æ˜¯å…ˆæ„å»ºAé€†</p>
<p>#æˆ‘å‘ç°ä¸€ç›´å¯¹è¡¨å‹å’Œç³»è°±çš„é‡ç¼–å·é¡ºåºæ˜¯é”™çš„ã€‚<br>
#ä¹‹å‰éƒ½æ˜¯å…ˆé‡ç¼–å·ï¼Œå†è¿½ç³»è°±ã€‚è¿™æ ·ä¼šåˆ æ‰ä¸€éƒ¨åˆ†è¡¨å‹ã€‚<br>
#åº”è¯¥å…ˆè¿½ç³»è°±ï¼ˆè¿™æ ·ç¡®ä¿æ‰€æœ‰è¡¨å‹ï¼ˆå‚éªŒç¾¤ä½“ï¼‰å‡åœ¨ç³»è°±ä¸­ï¼‰ï¼Œå†é‡ç¼–å·ã€‚</p>
<p>#å¦‚æœæŒ‰ç…§allçš„æ€è·¯ï¼Œåº”è¯¥æ˜¯è¿™æ ·ã€‚<br>
#ä»å•åœºé‡ç¼–å·ç»“æœï¼ˆä»1åˆ°n,è¦ç¡®ä¿ä¸é—´æ–­ï¼Œå³ç³»è°±æ€»å…±å°±æ˜¯nä¸ªä¸ªä½“ï¼‰ã€‚<br>
#ç„¶åæ²¡æœ‰è¡¨å‹çš„ä¸ªä½“å°±æ˜¯n-m(mä¸ºè¡¨å‹è¡Œæ•°)ã€‚<br>
#ç„¶åæ„å»ºZé˜µ æœ‰n-måˆ—å…¨ä¸º0,åé¢måˆ—ä¸º&quot;å•ä½é˜µ&quot;ã€‚ä¸å¯¹ï¼Œæˆ‘åšçš„æ˜¯ä¸¤ä¸ªåœºã€‚æ‰€ä»¥åé¢Zé˜µå†é‡å¤ä¸€éï¼Œå…ˆæ˜¯n2-m2åˆ—å…¨ä¸º0ï¼Œåé¢m2åˆ—ä¸º&quot;å•ä½é˜µ&quot;<br>
#å¾—åˆ°Aé€†ï¼Œé‡æ’ï¼ŒæŒ‰ä¸¤ä¸ªåœºåˆ†å¼€ï¼Œæ¯ä¸ªåœºéƒ½æ˜¯å‰é¢æ˜¯æ²¡è¡¨å‹çš„ï¼Œåé¢æ˜¯æœ‰è¡¨å‹çš„ã€‚</p>
<p>#å¦ä¸€ç§æ€è·¯<br>
#Aé€†å¯ä»¥ä¸é‡æ’ï¼Œå®Œå…¨æ ¹æ®Aé€†çš„é¡ºåºæ„å»ºZé˜µ<br>
#å¯¹äºæ¯ä¸€ä¸ªAä¸­çš„ä¸ªä½“ï¼Œæ²¡æœ‰è¡¨å‹ï¼Œåˆ™Zä¸­ç›¸åº”çš„è¿™ä¸€åˆ—ï¼ˆè¡Œï¼‰å…¨ä¸º0<br>
#å¦‚æœAé€†ä¸­çš„ä¸ªä½“æœ‰è¡¨å‹ï¼Œé‚£ä¹ˆæ‰¾åˆ°è¿™ä¸ªä¸ªä½“åœ¨è¡¨å‹ä¸­çš„è¡Œæ•°ï¼Œæ­¤æ—¶Zä¸­è¿™ä¸ªä¸ªä½“ç›¸åº”çš„è¡Œæ•°çš„åœ°æ–¹ä¸º1ï¼Œå…¶ä»–ä¸º0<br>
#è¿™ä¸ªåœ¨pythonä¸­ç”¨å­—å…¸å¯ä»¥åšåˆ°ï¼ŒRæ„Ÿè§‰ä¸æ–¹ä¾¿ã€‚</p>
<p>##ç¬¬äºŒç§allçš„æ€è·¯æ˜¯å¯¹çš„ã€‚ä½†æ˜¯Aé€†è¿˜æ˜¯è¦é‡æ’ï¼Œæœ€èµ·ç å¾—ç¬¬ä¸€ä¸ªåœºçš„ä¸ªä½“éƒ½åœ¨å‰é¢ï¼Œç¬¬äºŒä¸ªåœºçš„ä¸ªä½“å µåœ¨åé¢ã€‚<br>
##ä¸èƒ½ä¸¤ä¸ªåœºçš„ä¸ªä½“æ··åœ¨ä¸€å—ã€‚</p>
<p>##æ€»ç»“ä¸€ä¸‹ï¼Œallçš„æ€è·¯å¤æ‚ï¼Œåœ¨Rä¸­ä¸å¥½å®ç°ã€‚åœ¨pythonä¸­å¯èƒ½æ–¹éå®ç°ã€‚</p>
<p>#Z_phe<br>
#é‚£å°±å¾—æŒ‰ç…§è¡¨å‹çš„é¡ºåºæŒ‘Aé€†ï¼Œå³å¯ã€‚<br>
#Zé˜µå°±æ˜¯ä¸€ä¸ªm*mçš„å•ä½é˜µï¼ˆmä¸ºä¸¤ä¸ªåœºçš„è¡¨å‹æ€»æ•°ï¼‰<br>
#è¿™ç§æ–¹æ³•éå¸¸ç®€å•ï¼Œå®ç”¨ã€‚</p>
<p>#è¿™é‡Œæœ‰ä¸€ä¸ªæ½œåœ¨çš„é—®é¢˜æ˜¯ï¼Œè¡¨å‹ä¸èƒ½å†è¢«å‰”é™¤äº†ã€‚<br>
#ä¹Ÿå°±æ˜¯è¯´ä»è¡¨å‹çš„åŸå§‹æ–‡ä»¶åˆ°é‡ç¼–å·ï¼Œä¸èƒ½åˆ è¡¨å‹<br>
#ä¹Ÿå°±æ˜¯è¦å¯¹åŸå§‹æ–‡ä»¶æå‰å¤„ç†å¥½ã€‚</p>
<h1>æ€»çš„æ¥è¯´ï¼Œåªä½¿ç”¨æœ‰è¡¨å‹ä¸ªä½“çš„æ•°æ®ç®€å•ï¼Œç›®å‰å…ˆç”¨è¿™ç§æ–¹æ³•ã€‚</h1>
<h1>ç°åœ¨èƒ½é€šè¿‡Aé˜µç®—å‡ºæ¥ä¸¤ä¸ªåœºçš„LHSé€†äº†ï¼Œå°±å·®ä¸´é—¨ä¸€è„šPEBDäº†ã€‚</h1>
<p>ä½¿ç”¨å®é™…æ•°æ®ï¼Œåªè€ƒè™‘æœ‰è¡¨å‹ä¸ªä½“ï¼Œé€šè¿‡Aé˜µæ„å»ºç³»æ•°çŸ©é˜µæ•´ç†å¦‚ä¸‹ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># è¿™é‡Œåªç”¨æœ‰è¡¨å‹çš„ä¸ªä½“</span></span><br><span class="line"></span><br><span class="line">setwd(<span class="string">&#x27;D:\\Desktop\\R_model\\true_practice\\HAYF19647_HAQG17137&#x27;</span>)</span><br><span class="line"></span><br><span class="line">ped &lt;- read.table(<span class="string">&#x27;pedigree_code&#x27;</span>,header = <span class="built_in">F</span>)</span><br><span class="line">dat &lt;- read.table(<span class="string">&#x27;phe_code&#x27;</span>,header = <span class="built_in">F</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">## å¤„ç†ç³»è°±</span></span><br><span class="line">new_ped &lt;- ped[,<span class="number">1</span>:<span class="number">3</span>] <span class="comment">#ä¸è¦ç¬¬å››åˆ—</span></span><br><span class="line"></span><br><span class="line">library(nadiv)</span><br><span class="line">pped = prepPed(new_ped)<span class="comment">#ç»™foundersä¹Ÿå¢åŠ ä¸€è¡Œç³»è°±ï¼Œç„¶åæ’åºç³»è°±</span></span><br><span class="line"></span><br><span class="line">A_inv = makeAinv(pped)$Ainv<span class="comment">#å¯ç›´æ¥æ±‚Aé€†</span></span><br><span class="line"></span><br><span class="line">A_inv = as.matrix(A_inv)</span><br><span class="line"></span><br><span class="line">colnames(A_inv) = rownames(A_inv)</span><br><span class="line"></span><br><span class="line">str(dat)</span><br><span class="line">colnames(dat) &lt;- <span class="built_in">c</span>(<span class="string">&#x27;id&#x27;</span>,<span class="string">&#x27;gu&#x27;</span>,<span class="string">&#x27;wo&#x27;</span>,<span class="string">&#x27;age&#x27;</span>,<span class="string">&#x27;bf&#x27;</span>)</span><br><span class="line">dat$id &lt;- as.factor(dat$id) </span><br><span class="line">dat$gu &lt;- as.factor(dat$gu)</span><br><span class="line">X &lt;- model.matrix(~gu-<span class="number">1</span>,data = dat) </span><br><span class="line">Z &lt;- model.matrix(~id-<span class="number">1</span>,data = dat) <span class="comment"># Zä¸ºå•ä½é˜µ</span></span><br><span class="line"></span><br><span class="line">phe_id = <span class="built_in">as.character</span>(dat$id)</span><br><span class="line">A_inv2 = A_inv[phe_id,phe_id] <span class="comment">#æˆªå–è¡¨å‹ä¸ªä½“çš„Aé€†</span></span><br><span class="line"></span><br><span class="line">alpha = <span class="number">2.09</span> <span class="comment"># alphaç”±æ–¹å·®ç»„åˆ†å¾—æ¥ã€‚</span></span><br><span class="line"></span><br><span class="line">XpX=crossprod(X) <span class="comment">#Xâ€™X </span></span><br><span class="line"></span><br><span class="line">XpZ=crossprod(X,Z) <span class="comment">#Xâ€™Z</span></span><br><span class="line"></span><br><span class="line">ZpX=t(XpZ)</span><br><span class="line"></span><br><span class="line"><span class="comment">#ZpX=crossprod(Z,X) #Zâ€™X</span></span><br><span class="line"></span><br><span class="line">ZpZ=crossprod(Z) <span class="comment">#Zâ€™Z</span></span><br><span class="line"></span><br><span class="line">LHS=rbind(cbind(XpX,XpZ),cbind(ZpX,ZpZ+A_inv2*alpha)) <span class="comment">#LHS</span></span><br><span class="line"></span><br><span class="line">LHS_inv = solve(LHS)</span><br><span class="line"></span><br><span class="line">write.table(LHS,file = <span class="string">&#x27;HAYF_HAQG_LHS.txt&#x27;</span>,<span class="built_in">quote</span> = <span class="literal">FALSE</span>,sep = <span class="string">&quot; &quot;</span>,row.names = <span class="literal">FALSE</span>,col.names = <span class="literal">FALSE</span>)</span><br><span class="line">write.table(LHS_inv,file = <span class="string">&#x27;HAYF_HAQG_LHS_inv.txt&#x27;</span>,<span class="built_in">quote</span> = <span class="literal">FALSE</span>,sep = <span class="string">&quot; &quot;</span>,row.names = <span class="literal">FALSE</span>,col.names = <span class="literal">FALSE</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h1>Data Analysis for the Life Sciences</h1>
<h2 id="introduction">introduction</h2>
<p>statistical inference(ç»Ÿè®¡æ¨æ–­)ï¼šæ¨æ–­å°±æ˜¯ä½¿ç”¨å¯èƒ½æ€§ä»æ•°æ®ä¸­è·å–ç¾¤ä½“ç‰¹å¾å€¼ï¼ˆInference is the use of probability to learn population characteristic from dataï¼‰ã€‚ä¸€ä¸ªå…¸å‹çš„ä¾‹å­å°±æ˜¯æ¨æ–­ä¸¤ä¸ªç¾¤ä½“çš„å¹³å‡å€¼æ˜¯å¦ç›¸åŒã€‚</p>
<h2 id="getting-started">getting started</h2>
<p>ä¸»è¦æ˜¯ä»‹ç»äº†ä¸€äº›Rçš„åŸºç¡€éƒ¨åˆ†å’Œæ•°å­¦åŸºç¡€ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">###ç¬¬ä¸€ç«  getting started </span></span><br><span class="line">install.packages(<span class="string">&quot;swirl&quot;</span>)</span><br><span class="line">install.packages(<span class="string">&quot;rafalib&quot;</span>)</span><br><span class="line">install.packages(<span class="string">&quot;downloader&quot;</span>)</span><br><span class="line">setwd(<span class="string">&#x27;D:\\Desktop\\ä¸‹è½½è‹±æ–‡æ–‡çŒ®\\æ¨¡å‹\\R_model&#x27;</span>)</span><br><span class="line">dat &lt;- read.table(<span class="string">&#x27;female_mouse_weights.txt&#x27;</span>,header = <span class="built_in">T</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">###é€šè¿‡downloaderç›´æ¥ä¸‹è½½ç½‘ä¸Šçš„æ•°æ®</span></span><br><span class="line">library(downloader) <span class="comment">##use install.packages to install ##é€šè¿‡downloader</span></span><br><span class="line">url &lt;- <span class="string">&quot;https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extd\</span></span><br><span class="line"><span class="string">ata/femaleMiceWeights.csv&quot;</span>  <span class="comment">## æ•°æ®çš„å®Œæ•´è·¯å¾„</span></span><br><span class="line">filename &lt;- <span class="string">&quot;femaleMiceWeights.csv&quot;</span></span><br><span class="line">download(url, destfile=filename) <span class="comment">## ä¸‹è½½ï¼Œåé¢æ›´çš„æ˜¯ä¸‹è½½åçš„åç§°</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">### ä¸ºäº†ä»GitHub ä¸Šç›´æ¥ä¸‹è½½æ–‡ä»¶å¤¹ï¼Œä½ éœ€è¦ä¸‹è½½devtoolsåŒ…</span></span><br><span class="line">library(devtools)</span><br><span class="line">install_github(<span class="string">&quot;genomicsclass/dagdata&quot;</span>)</span><br><span class="line"></span><br><span class="line">dir &lt;- system.file(package = <span class="string">&quot;dagdata&quot;</span>) <span class="comment"># è·å–æ–‡ä»¶å¤¹ä½ç½®</span></span><br><span class="line">dir  <span class="comment">## &quot;D:/software/R/R-3.6.1/library/dagdata&quot;</span></span><br><span class="line">list.files(dir) <span class="comment"># æŸ¥çœ‹dirä¸‹é¢çš„æ–‡ä»¶å’Œå­æ–‡ä»¶å¤¹</span></span><br><span class="line">list.files(dir,<span class="string">&quot;extdata&quot;</span>) <span class="comment"># æŸ¥çœ‹extdataä¸‹çš„å­æ–‡ä»¶</span></span><br><span class="line">filename&lt;-file.path(dir,<span class="string">&quot;extdata/femaleMiceWeights.csv&quot;</span>) <span class="comment">## æ–‡ä»¶çš„å®Œæ•´è·¯å¾„</span></span><br><span class="line">dat&lt;-read.csv(filename) <span class="comment">## æœ€ç»ˆæ‰“å¼€æ–‡ä»¶</span></span><br><span class="line"></span><br><span class="line"><span class="comment">####åˆæ¢dplyr </span></span><br><span class="line">library(<span class="string">&quot;dplyr&quot;</span>)</span><br><span class="line"><span class="comment">##dplyråŒ…ä¸­filteræ˜¯è¿›è¡Œæ•°æ®ç­›é€‰</span></span><br><span class="line">chow &lt;- filter(dat,Diet==<span class="string">&quot;chow&quot;</span>) <span class="comment">## ç­›é€‰Dietåˆ—ä¸­ä¸ºchowçš„æ•°æ®</span></span><br><span class="line">head(chow)</span><br><span class="line"></span><br><span class="line"><span class="comment">## selectè²Œä¼¼æ˜¯é€‰åˆ—</span></span><br><span class="line">chowVals &lt;- select(chow,Bodyweight) <span class="comment">## é€‰æ‹©chowä¸­çš„Bodyweightåˆ—</span></span><br><span class="line">head(chowVals)</span><br><span class="line"></span><br><span class="line"><span class="comment">## è¿˜å¯ä»¥ä½¿ç”¨é€šé“ç¬¦ %&gt;% denote a pipe </span></span><br><span class="line"><span class="comment">## å› æ­¤ä¸Šè¿°ä¸¤æ­¥å¯ä»¥åˆæˆä¸€æ­¥</span></span><br><span class="line">chowVals &lt;- filter(dat,Diet==<span class="string">&quot;chow&quot;</span>) %&gt;% select(Bodyweight)</span><br><span class="line">head(chowVals)</span><br><span class="line"><span class="comment">## dplyræ¥æ”¶çš„æ˜¯data.frame ï¼Œè¿”å›çš„ä¹Ÿæ˜¯data.frame</span></span><br><span class="line"><span class="built_in">class</span>(chowVals) <span class="comment">#data.frame</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## ä¸ºäº†ä½¿ç»“æœä¸ºå‘é‡ï¼Œå¯ä»¥ä½¿ç”¨unlistå‡½æ•°ã€‚å°†list(åŒ…æ‹¬data.frame)è½¬æ¢æˆnumeric vectors. </span></span><br><span class="line">chowVals &lt;- filter(dat,Diet==<span class="string">&quot;chow&quot;</span>) %&gt;% select(Bodyweight) %&gt;% unlist</span><br><span class="line"><span class="comment">## é€šé“å°‘å†™ä¸€ä¸ªå‚æ•°ï¼Œä½†æ˜¯è¿™é‡Œçš„unlistè¿()éƒ½çœäº†</span></span><br><span class="line"><span class="built_in">class</span>(chowVals) <span class="comment">#numeric </span></span><br><span class="line"></span><br><span class="line"><span class="comment">## ä¸ç”¨dplyrçš„å†™æ³•</span></span><br><span class="line">chowVals &lt;- dat[dat$Diet==<span class="string">&quot;chow&quot;</span>,<span class="string">&quot;Bodyweight&quot;</span>]</span><br><span class="line">head(chowVals)</span><br><span class="line"></span><br><span class="line"><span class="comment">## å°ç»“ï¼Œæ„Ÿè§‰Ræ•´ç†æ•°æ®å°±æ˜¯ä¸å¤ªæ–¹ä¾¿ã€‚</span></span><br></pre></td></tr></table></figure>
<h2 id="inference">inference</h2>
<p>ä»¥ä¸‹ä¸ºrandom variable,  null distribution,  p-value çš„ ç®€è¦ä»‹ç»ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br><span class="line">107</span><br><span class="line">108</span><br><span class="line">109</span><br><span class="line">110</span><br><span class="line">111</span><br><span class="line">112</span><br><span class="line">113</span><br><span class="line">114</span><br><span class="line">115</span><br><span class="line">116</span><br><span class="line">117</span><br><span class="line">118</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#inference</span></span><br><span class="line">setwd(<span class="string">&#x27;D:\\Desktop\\ä¸‹è½½è‹±æ–‡æ–‡çŒ®\\æ¨¡å‹\\R_model&#x27;</span>)</span><br><span class="line">dat &lt;- read.table(<span class="string">&#x27;female_mouse_weights.txt&#x27;</span>,header = <span class="built_in">T</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#####didtribution çš„è§£æ</span></span><br><span class="line"><span class="comment">#distributionï¼šå¯ä»¥ç®€å•åœ°ç†è§£ä¸ºä¸€å †æ•°ã€‚</span></span><br><span class="line"><span class="comment">#æ¯”å¦‚ä½ æµ‹é‡äº†ä¸€ä¸ªç¾¤ä½“ä¸­æ‰€æœ‰äººçš„èº«é«˜ã€‚</span></span><br><span class="line"><span class="comment">#è¿™ä¸ªæ—¶å€™ä½ è¦å’Œå…¶ä»–äººä¸æ‡‚çš„äººï¼ˆsuch as analien that has never visited Earthï¼‰æè¿°è¿™äº›æ•°å­—ã€‚</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#å‡å¦‚ä»¥ä¸‹å°±æ˜¯ä½ å–å¾—çš„èº«é«˜æ•°æ®</span></span><br><span class="line">library(UsingR)</span><br><span class="line">x &lt;- father.son$fheight</span><br><span class="line"></span><br><span class="line"><span class="comment">#æœ€ç®€å•çš„å°±æ˜¯åˆ—å‡ºæ¥ç»™å¤–æ˜Ÿäººçœ‹</span></span><br><span class="line"><span class="built_in">round</span>(sample(x,<span class="number">10</span>),<span class="number">1</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#ä½†è¿™ä¸å¤Ÿå……åˆ†ï¼Œæˆ‘ä»¬å¯ä»¥å®šä¹‰ä¸€ä¸ªç´¯ç§¯åˆ†å¸ƒå‡½æ•°ï¼ˆCumulative Distribution Functionï¼‰</span></span><br><span class="line"><span class="comment">#F(a) == Pr(x&lt;a) #å°äºç­‰äºå·</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#ç”»å›¾å¦‚ä¸‹</span></span><br><span class="line"><span class="comment"># floor ceiling æ˜¯å–æ•´ç”¨çš„</span></span><br><span class="line"><span class="comment"># floor æ˜¯ç›´æ¥å–æ•´æ•°ï¼Œceiling æ˜¯ â€œå¤©èŠ±æ¿â€ </span></span><br><span class="line"><span class="comment"># å¦‚ floor(2.99) = 2 ceiling(2.99) = 3</span></span><br><span class="line"><span class="comment"># ç”¨åœ¨ç”»å›¾è¿˜æ˜¯æŒºå¥½çš„</span></span><br><span class="line"></span><br><span class="line">smallest&lt;-<span class="built_in">floor</span>( <span class="built_in">min</span>(x) )</span><br><span class="line">largest&lt;-<span class="built_in">ceiling</span>( <span class="built_in">max</span>(x) )</span><br><span class="line">values&lt;-seq(smallest, largest,len=<span class="number">300</span>)</span><br><span class="line">heightecdf&lt;-ecdf(x)</span><br><span class="line">plot(values, heightecdf(values), type=<span class="string">&quot;l&quot;</span>,xlab=<span class="string">&quot;a (Height in inches)&quot;</span>,ylab=<span class="string">&quot;Pr(x &lt;= a)&quot;</span>)</span><br><span class="line"><span class="comment"># ä½†æ˜¯ç°åœ¨æ›´å¸¸è§çš„æ˜¯ç”»ç›´æ–¹å›¾</span></span><br><span class="line"><span class="comment"># æŠŠç›´æ–¹å›¾ç»™å¤–æ˜Ÿäººä¸€çœ‹ï¼ŒåŸºæœ¬å°±å¾ˆæ¸…æ™°äº†ã€‚</span></span><br><span class="line">hist(x)</span><br><span class="line">bins&lt;-seq(smallest, largest)</span><br><span class="line">hist(x,breaks=bins,xlab=<span class="string">&quot;Height (in inches)&quot;</span>,main=<span class="string">&quot;Adult men heights&quot;</span>)</span><br><span class="line"><span class="comment">#breaksè®¾ç½®ä¸­ï¼Œä¸ç”¨floor,ceiling ç”»ä¸æˆå›¾</span></span><br><span class="line"></span><br><span class="line"><span class="comment">####å›åˆ°æ­£é¢˜</span></span><br><span class="line">library(dplyr)</span><br><span class="line">control &lt;- filter(dat,Diet==<span class="string">&quot;chow&quot;</span>) %&gt;% select(Bodyweight) %&gt;% unlist</span><br><span class="line">treatment  &lt;- filter(dat,Diet==<span class="string">&quot;hf&quot;</span>) %&gt;% select(Bodyweight) %&gt;% unlist</span><br><span class="line">mean(treatment)</span><br><span class="line">mean(control)</span><br><span class="line">obsdiff &lt;- mean(treatment) - mean(control)</span><br><span class="line">obsdiff</span><br><span class="line"><span class="comment">#[1] 3.020833</span></span><br><span class="line"><span class="comment">#åªçœ‹å‡å€¼å’Œå·®å€¼æ˜¯ä¸å¤Ÿçš„</span></span><br><span class="line"><span class="comment">#å› ä¸ºå¦‚æœä½ å†å–24åªè€é¼ é‡æ–°åšè¯•éªŒï¼Œæ¯ä¸€æ¬¡éƒ½èƒ½å¾—åˆ°ä¸€ä¸ªæ–°çš„ç»“æœã€‚</span></span><br><span class="line"><span class="comment">#æˆ‘ä»¬æŠŠå…·æœ‰è¿™ç§ç‰¹æ€§çš„æ•°æ®ç§°ä¸ºéšæœºå˜é‡ã€‚</span></span><br><span class="line"></span><br><span class="line"><span class="comment">####æ›´æ·±åˆ»åœ°ç†è§£éšæœºå˜é‡###########</span></span><br><span class="line"><span class="comment"># å‡è®¾æˆ‘ä»¬æœ‰æ‰€æœ‰çš„controlç¾¤ä½“ï¼Œå¦‚ä¸‹</span></span><br><span class="line">library(downloader)</span><br><span class="line">url&lt;-<span class="string">&quot;https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/femaleControlsPopulation.csv&quot;</span></span><br><span class="line">filename&lt;-<span class="string">&quot;femaleControlsPopulation.csv&quot;</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span>(!file.exists(filename)) download(url,destfile=filename)</span><br><span class="line"><span class="comment">#å¦‚æœæ–‡ä»¶ä¸å­˜åœ¨åˆ™ä¸‹è½½</span></span><br><span class="line"></span><br><span class="line">population&lt;-read.csv(filename)</span><br><span class="line"></span><br><span class="line">population &lt;- unlist(population) <span class="comment">#ä»data.frame è½¬ä¸º vector</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#ä»æ€»ä½“ä¸­æ¯æ¬¡éšæœºæŠ½å–12åªè€é¼ ï¼Œé‡å¤ä¸‰æ¬¡ï¼Œæ¯æ¬¡å‡å€¼å‡ä¸åŒ</span></span><br><span class="line"></span><br><span class="line">control&lt;-sample(population,<span class="number">12</span>)</span><br><span class="line">mean(control)</span><br><span class="line"><span class="comment">#[1] 23.4925</span></span><br><span class="line"></span><br><span class="line">control&lt;-sample(population,<span class="number">12</span>)</span><br><span class="line">mean(control)</span><br><span class="line"><span class="comment">#[1] 24.96583</span></span><br><span class="line"></span><br><span class="line">control&lt;-sample(population,<span class="number">12</span>)</span><br><span class="line">mean(control)</span><br><span class="line"><span class="comment">#[1] 22.90083</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#å› æ­¤å“ªæ€•åŒä¸€ä¸ªç¾¤ä½“ï¼ŒæŠ½å–ä¸åŒçš„æ ·æœ¬ï¼Œä¹‹é—´ä¹Ÿæ˜¯æœ‰å·®å¼‚çš„ã€‚</span></span><br><span class="line"><span class="comment">#æ‰€ä»¥å¯¹äºä¸Šé¢ä¸¤ä¸ªç¾¤ä½“çš„obsdiffï¼Œæˆ‘ä»¬å­˜æœ‰æ€€ç–‘ï¼Œè¿™ç©¶ç«Ÿæ˜¯ä¸æ˜¯å› ä¸ºå½±å“å› ç´ ï¼ˆdietï¼‰é€ æˆçš„</span></span><br><span class="line"><span class="comment">#å¦‚æœæˆ‘ä»¬ç»™è¿™24åªè€é¼ ç›¸åŒçš„dietï¼Œå·®å¼‚è¿˜æœ‰æœ‰è¿™ä¹ˆå¤§å—ï¼Ÿ</span></span><br><span class="line"><span class="comment">#å‡ºäºæ€€ç–‘ç²¾ç¥ï¼Œä¸€èˆ¬ç»Ÿè®¡æ¨æ–­çš„é›¶å‡è®¾ï¼ˆnull hypothesisï¼‰ä¼šè®¾å®šä¸ºä¸¤ä¸ªç¾¤ä½“æ²¡æœ‰åŒºåˆ«ã€‚</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#ç”±äºæˆ‘ä»¬æœ‰äº†controlçš„æ€»ä½“ï¼Œæ‰€ä»¥æˆ‘ä»¬å¯ä»¥æ— é™æ¬¡æŠ½æ ·ã€‚</span></span><br><span class="line"><span class="comment">#æ¯æ¬¡ä»controlçš„ç¾¤ä½“ä¸­ï¼ŒæŠ½2æ‰¹12åªè€é¼ ï¼Œæ±‚å‡å€¼ã€‚ä»£ç å¦‚ä¸‹</span></span><br><span class="line"></span><br><span class="line">control&lt;-sample(population,<span class="number">12</span>)</span><br><span class="line"><span class="comment">##another 12 control mice that we act as if they were not</span></span><br><span class="line">treatment &lt;-sample(population,<span class="number">12</span>)</span><br><span class="line">print(mean(treatment)-mean(control))</span><br><span class="line"><span class="comment">#[1] -1.175</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#ç°åœ¨è®©æˆ‘ä»¬é‡å¤è¿™ä¸€æ­¥éª¤10000æ¬¡</span></span><br><span class="line"></span><br><span class="line">null &lt;- vector()</span><br><span class="line"><span class="keyword">for</span> (i <span class="keyword">in</span> <span class="number">1</span>:<span class="number">10000</span>)&#123;</span><br><span class="line">  control&lt;-sample(population,<span class="number">12</span>)</span><br><span class="line">  treatment &lt;-sample(population,<span class="number">12</span>)</span><br><span class="line">  null[i] &lt;- mean(treatment)-mean(control)</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line"><span class="comment"># ç°åœ¨åœ¨nullä¸­çš„å€¼æˆ‘ä»¬ç§°ä¹‹ä¸ºnull distribution</span></span><br><span class="line"></span><br><span class="line">mean(null&gt;=obsdiff)</span><br><span class="line"><span class="comment"># [1] 0.0147</span></span><br><span class="line"><span class="comment">#meané‡Œé¢çš„æ˜¯ä¸€ä¸ªå¸ƒå°”å‘é‡ï¼Œç”±TRUEå’ŒFALSEç»„æˆ</span></span><br><span class="line"><span class="comment">#ä½†æ˜¯åœ¨è®¡ç®—çš„è¿‡ç¨‹ä¸­ï¼ŒTRUEè§†ä¸º1ï¼ŒFALSEè§†ä¸º0ã€‚</span></span><br><span class="line"><span class="comment">#æ‰€ä»¥è¿™ä¸ª0.0147æ˜¯è¿™10000æ¬¡ä¸­å·®è·å¤§äºobsdiffçš„æ¯”ä¾‹ã€‚</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#å› æ­¤ä½œä¸ºä¸€ä¸ªæ€€ç–‘ä¸»ä¹‰è€…ï¼Œæˆ‘ä»¬å¯ä»¥å¾—å‡ºç»“è®º</span></span><br><span class="line"><span class="comment">#å¦‚æœæ²¡æœ‰diet effectï¼Œæˆ‘ä»¬çœ‹åˆ°åƒobsdiffè¿™ä¹ˆçš„å·®è·çš„å¯èƒ½æ€§æ˜¯1.47%</span></span><br><span class="line"><span class="comment">#è¿™å°±æ˜¯æ‰€è°“çš„På€¼ï¼</span></span><br><span class="line"><span class="comment">#på€¼å³åŸå‡è®¾æˆç«‹çš„å‰æä¸‹ï¼Œæ­¤äº‹ä»¶å‘ç”Ÿçš„æ¦‚ç‡ã€‚</span></span><br><span class="line"><span class="comment">#æ·±å…¥æµ…å‡ºï¼Œå‰å®³ï¼</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#å°ç»“ä¸€ä¸‹ï¼Œåƒä¸Šè¿°è¿™æ ·è®¡ç®—På€¼å¾ˆç®€å•ã€‚ä½†æ˜¯ç”Ÿæ´»ä¸­ä¸å¯è¡Œã€‚</span></span><br><span class="line"><span class="comment">#å› ä¸ºä½ ä¸è‚¯çœŸçš„æŠŠæµ‹é‡æ•´ä¸ªç¾¤ä½“çš„ä¿¡æ¯ï¼Œä»è€Œæ„æˆnull distributionã€‚</span></span><br><span class="line"><span class="comment">#ä¸€èˆ¬ä¹Ÿè¿˜æ˜¯é€šè¿‡é‡‡æ ·ã€‚</span></span><br><span class="line"><span class="comment">#ä¹Ÿå°±æ˜¯è¿˜æ˜¯åªé€šè¿‡æ•°æ®ä¸­çš„24åªè€é¼ çš„æ•°æ®è¿›è¡Œåˆ¤æ–­ã€‚</span></span><br></pre></td></tr></table></figure>
<h3 id="population-parameters">population parameters</h3>
<p>ç»Ÿè®¡æ¨æ–­çš„ç¬¬ä¸€æ­¥æ˜¯å¼„æ¸…æ¥šä½ æ„Ÿå…´è¶£çš„ç¾¤ä½“ã€‚ä½œè€…é€‰è¿™ä¸ªè¿™ä¸ªå°é¼ çš„æ•°æ®é›†çš„åŸå› ä¹‹ä¸€ï¼Œæ˜¯å› ä¸ºä½œè€…æœ‰æ‰€æœ‰è¿™ç§ç±»å‹çš„å°é¼ çš„ä½“é‡ä¿¡æ¯ï¼ˆå³è¿™ä¸ªç¾¤ä½“çš„æ‰€æœ‰ç¾¤ä½“ï¼‰ã€‚</p>
<p>å¦‚æœä½ èƒ½æ”¶é›†ä½ æ„Ÿå…´è¶£çš„ç¾¤ä½“çš„æ‰€æœ‰æ•°æ®ï¼Œé‚£ä¹ˆå¾—åˆ°çš„å‚æ•°ï¼ˆç¾¤ä½“å‡å€¼ã€æ–¹å·®ã€æ ‡å‡†å·®ï¼‰å°±è¢«ç§°ä¸ºç¾¤ä½“å‚æ•°ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">library(downloader)</span><br><span class="line">url&lt;-<span class="string">&quot;https://raw.githubusercontent.com/genomicsclass/dagdata/master/inst/extdata/mice_pheno.csv&quot;</span></span><br><span class="line">filename&lt;-<span class="string">&quot;mice_pheno.csv&quot;</span> </span><br><span class="line">download(url,destfile=filename)</span><br><span class="line">dat&lt;-read.csv(filename) <span class="comment">#è¿™æ˜¯æ‰€æœ‰è¿™ä¸ªå°é¼ ç¾¤ä½“çš„æ•°æ®</span></span><br><span class="line"></span><br><span class="line">library(dplyr)</span><br><span class="line">controlPopulation &lt;- filter(dat,Sex==<span class="string">&quot;F&quot;</span>&amp;Diet==<span class="string">&quot;chow&quot;</span>) </span><br><span class="line"><span class="built_in">length</span>(controlPopulation$Bodyweight)</span><br><span class="line"></span><br><span class="line">hfPopulation&lt;-filter(dat,Sex==<span class="string">&quot;F&quot;</span>&amp;Diet==<span class="string">&quot;hf&quot;</span>)</span><br><span class="line"><span class="built_in">length</span>(hfPopulation$Bodyweight)</span><br><span class="line"><span class="comment">#æ­¤æ—¶ç¾¤ä½“çš„å‡å€¼ã€æ–¹å·®ã€æ ‡å‡†å·®ç­‰å°±è¢«ç§°ä¸ºç¾¤ä½“å‚æ•°ï¼ˆpopulation parametersï¼‰</span></span><br><span class="line"><span class="comment">#æ­¤æ—¶å›åˆ°æˆ‘ä»¬çš„é—®é¢˜ï¼Œä¸¤ä¸ªç¾¤ä½“çš„å‡å€¼æ˜¯å¦ç›¸ç­‰ã€‚</span></span><br><span class="line"><span class="comment">#ç”±äºæˆ‘ä»¬æ‹¿åˆ°äº†ä¸¤ä¸ªç¾¤ä½“çš„æ‰€æœ‰æ•°æ®ï¼Œé‚£ä¹ˆç”¨Î¼y-uxï¼ˆx:control y:hfï¼‰,ç›´æ¥çœ‹å·®å€¼æ˜¯å¦ä¸º0å³å¯ã€‚</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä½†æ˜¯ï¼Œåœ¨å®é™…ç”Ÿæ´»ä¸­ï¼Œæˆ‘ä»¬æ‹¿ä¸åˆ°ç¾¤ä½“çš„æ‰€æœ‰æ•°æ®ã€‚æ¯”å¦‚ï¼Œå¯èƒ½ä¹°ä¸‹ä¸€ä¸ªç¾¤ä½“æ‰€æœ‰çš„å°é¼ æ˜¯å¾ˆè´µçš„ã€‚<br>
å› æ­¤ï¼Œæˆ‘ä»¬éœ€è¦å­¦ä¼šå¦‚ä½•ä»æ ·æœ¬æ•°æ®æ¥å›ç­”æˆ‘ä»¬çš„é—®é¢˜ã€‚è¿™å°±æ˜¯ç»Ÿè®¡æ¨æ–­çš„æœ¬è´¨ã€‚</p>
<h3 id="Sample-estimates">Sample estimates</h3>
<p>åœ¨ä¹‹å‰çš„ç« èŠ‚ä¸­ï¼Œæˆ‘ä»¬ä»ä¸¤ä¸ªç¾¤ä½“ä¸­å„éšæœºæŠ½å–äº†12åªå°é¼ ã€‚æˆ‘ä»¬ä¸€èˆ¬ç”¨å¤§å†™å­—æ¯æ¥è¡¨ç¤ºæŠ½æ ·ä¸ªä½“ï¼Œä»¥æ­¤è¡¨æ˜ä»–ä»¬æ˜¯éšæœºæŠ½å–çš„ã€‚è¿™ä¹Ÿæ˜¯ç»Ÿè®¡ä¸­çš„ä¸€ç§å¸¸è§„åšæ³•ã€‚å› æ­¤æŠ½æ ·ä¸ªä½“è¡¨ç¤ºä¸ºX~1~ â€¦â€¦X~12~ å’Œ Y~1~ â€¦â€¦ Y~12~ã€‚<br>
è€Œå¯¹äºä¸Šé¢çš„ç¾¤ä½“ä¸­çš„ä¸ªä½“ï¼Œç”¨çš„æ˜¯å°å†™æ•°å­—ã€‚å› ä¸ºè¿™äº›ä¸ªä½“æ˜¯ç»™å®šçš„ï¼Œä¸æ˜¯éšæœºçš„ã€‚<br>
æ­¤æ—¶ï¼Œç ”ç©¶ç›®æ ‡æ˜¯ä¸¤ä¸ªæ ·æœ¬ç¾¤ä½“å‡å€¼çš„å·®å€¼ã€‚<br>
è€Œè¿™ä¸ªå·®å€¼ä¹Ÿæ˜¯ä¸€ä¸ªéšæœºå˜é‡ï¼ˆrandom variableï¼‰ã€‚å‰é¢ï¼Œæˆ‘ä»¬é‡‡ç”¨ä¸€ç§ä»åŸå§‹æ•°æ®ä¸­é‡å¤æŠ½æ ·çš„æ‰‹æ®µï¼ˆ10000æ¬¡ï¼‰ã€‚ä½†æ˜¯ï¼Œè¿™åœ¨å®é™…ä¸­ï¼ˆin practiceï¼‰æ˜¯ä¸å¯å®ç°çš„ï¼Œæ¯”å¦‚åœ¨æˆ‘ä»¬è¿™ä¸ªä¾‹å­ä¸­ï¼Œå°±æ˜¯è¦ä¸æ–­åœ°ä¹°24åªå°é¼ åšè¯•éªŒã€‚<br>
ä¸‹é¢ä¼šä»‹ç»ç¾¤ä½“å‡å€¼å·®ä¸æ ·æœ¬å‡å€¼å·®çš„å…³ç³»ã€‚è¿™æ¶‰åŠåˆ°äº†ä¸­å¿ƒæé™å®šç†ã€‚</p>
<h3 id="Central-Limit-Theorem-and-t-distribution">Central Limit Theorem and t-distribution</h3>
<p><strong>ä¸­å¿ƒæé™å®šç†</strong>ï¼š<strong>æ— è®ºç¾¤ä½“æ€»ä½“æœä»ä»€ä¹ˆåˆ†å¸ƒ</strong>ï¼Œå½“æŠ½æ ·ä¸ªä½“æ•°è¶³å¤Ÿå¤§æ—¶ï¼ˆä¸€èˆ¬è®¤ä¸ºï¼Œæ¯ç»„å¤§äºç­‰äº30ä¸ªï¼‰ï¼ŒæŠ½æ ·ç¾¤ä½“çš„å‡å€¼ï¼ˆYæ‹”ï¼‰æœä»æ­£æ€åˆ†å¸ƒã€‚å…¶å‡å€¼ä¸ºç¾¤ä½“å‡å€¼ï¼Œå…¶æ ‡å‡†å·®ç­‰äºç¾¤ä½“æ ‡å‡†å·®é™¤ä»¥æ ¹å·Nï¼ˆN ä¸º simple sizeï¼‰ã€‚æˆ‘ä»¬æŠŠè¿™ä¸ªéšæœºå˜é‡åˆ†å¸ƒçš„æ ‡å‡†å·®ç§°ä¸ºæ ‡å‡†è¯¯ï¼ˆ<em>standard error</em>ï¼‰ã€‚</p>
<ul>
<li>å¯¹äºä¸€ä¸ªéšæœºå˜é‡ï¼ŒåŠ å‡å¸¸æ•°ï¼ˆX+aï¼‰ï¼Œåˆ™è¯¥éšæœºå˜é‡çš„å‡å€¼ä¹Ÿéšä¹‹å˜åŒ–ï¼ˆu~X~+aï¼‰</li>
<li>å¯¹äºä¸€ä¸ªéšæœºå˜é‡ï¼Œä¹˜ä»¥ä¸€ä¸ªå¸¸æ•°ï¼ˆX*aï¼‰ï¼Œåˆ™è¯¥éšæœºå˜é‡çš„å‡å€¼å’Œæ ‡å‡†å·®ä¹Ÿä¹˜ä»¥æ­¤å¸¸æ•°</li>
<li>å¯¹äºä¸¤ä¸ªç‹¬ç«‹çš„éšæœºå˜é‡ç›¸åŠ ï¼ˆX+Yï¼‰ï¼Œå…¶å‡å€¼ä¸ºÎ¼~X~+Î¼~Y~ï¼Œå…¶æ–¹å·®ä¸ºäºŒè€…ä¹‹å’Œã€‚å¦‚æœæ˜¯ä¸¤ä¸ªç‹¬ç«‹çš„éšæœºå˜é‡ç›¸å‡ï¼Œå…¶å‡å€¼ä¸ºäºŒè€…ä¹‹å·®ï¼Œå…¶æ–¹å·®ä»æœª<strong>äºŒè€…ä¹‹å’Œ</strong>ã€‚</li>
<li>å¦‚æœä¸¤ä¸ªéšæœºå˜é‡éƒ½æ»¡è¶³æ­£æ€åˆ†å¸ƒï¼Œé‚£ä¹ˆå®ƒä»¬åªå’Œä»æ»¡è¶³æ­£æ€åˆ†å¸ƒã€‚</li>
</ul>
<p>åœ¨æœ¬ä¾‹ä¸­ï¼Œæˆ‘ä»¬å…³å¿ƒçš„ä¸¤ä¸ªç¾¤ä½“çš„å‡å€¼ï¼Œç”±äºä¸­å¿ƒæé™å®šç†ï¼Œæœä»æ­£æ€åˆ†å¸ƒã€‚å› æ­¤å®ƒä»¬çš„å·®å€¼ä¹Ÿæœä»æ­£æ€åˆ†å¸ƒï¼Œå‡å€¼ä¸ºä¸¤ä¸ªéšæœºå˜é‡ä¹‹å·®ï¼Œæ–¹å·®ä¸º<strong>ä¸¤ä¸ªéšæœºå˜é‡ä¹‹å’Œ</strong>ã€‚ç”±äºåŸå‡è®¾æ˜¯ä¸å­˜åœ¨å·®å¼‚ï¼Œé‚£ä¹ˆä¸¤ä¸ªç¾¤ä½“çš„å‡å€¼çš„å‡å€¼åº”è¯¥å¤§è‡´æœä»ä¸€ä¸ªå‡å€¼ä¸º0ï¼Œæ ‡å‡†å·®ä¸ºâˆšï¼ˆÏƒ2X+Ïƒ2Yï¼‰/âˆšNçš„æ­£æ€åˆ†å¸ƒã€‚<br>
ä¹Ÿå°±æ˜¯å¦‚ä¸‹å¼</p>
<p><img src="clt.png" alt="clt"></p>
<p>è¿™ä¸ªè¿›è¡Œäº†æ ‡å‡†åŒ–çš„éšæœºå˜é‡åº”è¯¥æ»¡è¶³ï¼ˆ0ï¼Œ1ï¼‰çš„æ ‡å‡†æ­£æ€åˆ†å¸ƒã€‚ï¼ˆæ·±å…¥æµ…å‡ºï¼ï¼‰</p>
<p>æ ¹æ®è¿™ä¸ªå…¬å¼ï¼Œæˆ‘ä»¬å¯ä»¥æ›´å¿«åœ°æ±‚å‡ºPå€¼ï¼ˆå·²çŸ¥æ ‡å‡†æ­£æ€åˆ†å¸ƒï¼‰ã€‚ä¾‹å¦‚ï¼Œæ­£æ€åˆ†å¸ƒä¸­å¤§äº2çš„å€¼åªæœ‰5%ï¼ˆin absolute valueï¼Œåº”è¯¥æ˜¯åŒå°¾çš„æ„æ€ï¼‰ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">pnorm(-<span class="number">2</span>)+(<span class="number">1</span>-pnorm(<span class="number">2</span>))</span><br><span class="line"><span class="comment">## [1] 0.04550026</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>å› æ­¤ï¼Œæˆ‘ä»¬ä¸ç”¨å†ä¹°æ›´å¤šçš„å°é¼ ï¼Œ24åªå¤Ÿäº†ã€‚</p>
<p>ç„¶è€Œï¼Œäº‹æƒ…åˆ°è¿™ä¸€æ­¥è¿˜æ²¡æœ‰ç»“æŸã€‚å› ä¸ºæˆ‘ä»¬ä¸çŸ¥é“ç¾¤ä½“çš„æ ‡å‡†å·®ï¼ŒÏƒ~X~å’ŒÏƒ~Y~ã€‚è¿™äº›éƒ½æ˜¯ä¸å¯çŸ¥çš„ç¾¤ä½“å‚æ•°ã€‚ä½†æ˜¯æˆ‘ä»¬å¯ä»¥ç”¨æ ·æœ¬æ ‡å‡†å·®å»ä¼°è®¡ç¾¤ä½“æ ‡å‡†å·®ï¼Œç§°ä¸ºs~X~ å’Œ s~Y~ã€‚å®šä¹‰å¦‚ä¸‹ï¼š</p>
<p><img src="s2.png" alt="s2"></p>
<p>è¿™é‡Œæ˜¯é™¤ä»¥M-1 å’Œ N-1ï¼ˆå‡ºäºç†è®ºåŸå› ï¼Œæ²¡è§£é‡Šï¼‰ã€‚</p>
<p>å› æ­¤ï¼Œæˆ‘ä»¬å¯ä»¥æŠŠä¸Šå¼æ”¹ä¸ºï¼š</p>
<p><img src="clt_new.png" alt="t_value"></p>
<p>ä¸­å¿ƒæé™å®šç†å‘Šè¯‰æˆ‘ä»¬ï¼Œå½“M,Nè¾ƒå¤§æ—¶ï¼Œè¿™ä¸ªéšæœºå˜é‡æœä»æ ‡å‡†æ­£æ€åˆ†å¸ƒï¼ˆé›¶å‡è®¾ä¸‹ï¼‰ã€‚å› æ­¤ï¼Œæˆ‘ä»¬ç°åœ¨å¯ä»¥è®¡ç®—På€¼äº†ã€‚</p>
<h3 id="t-åˆ†å¸ƒ">t åˆ†å¸ƒ</h3>
<p>ä¸­å¿ƒæé™å®šç†éœ€è¦è¾ƒå¤§çš„æŠ½æ ·ä¸ªä½“æ•°ï¼Œè¿™ä¸ªæˆ‘ä»¬ç§°ä¹‹ä¸º<em>asymptotic results</em>ã€‚å½“è¿™ä¸ªæ¡ä»¶ä¸æ»¡è¶³æ—¶ï¼Œè¿˜æœ‰ä¸€ç§æ–¹æ³•å¯ä»¥ä½¿ç”¨ã€‚å¦‚æœæ€»ä½“æ•°æ®æœä»ä¸€ä¸ªå‡å€¼ä¸º0çš„æ­£æ€åˆ†å¸ƒï¼Œé‚£ä¹ˆå¯¹äºæˆ‘ä»¬çš„æŠ½æ ·ä¸ªä½“å‡å€¼ï¼Œå¯ä»¥è®¡ç®—ä»¥ä¸‹åˆ†å¸ƒï¼Œç§°ä¸ºtåˆ†å¸ƒï¼š</p>
<p><img src="t-distribution.png" alt="t-distribution"></p>
<p>è¿™é‡Œå¾ˆé‡è¦çš„ä¸€ç‚¹ï¼Œæˆ‘ä»¬æåˆ°çš„æœä»æ­£æ€åˆ†å¸ƒæ˜¯æ€»ä½“ï¼ˆy1,y2,â€¦â€¦ynï¼‰æœä»æ­£æ€åˆ†å¸ƒï¼Œè€Œä¸æ˜¯Yæ‹”ã€‚è™½ç„¶æ— æ³•å®é™…è¯æ˜ï¼Œä½†æ˜¯æˆ‘ä»¬çœ‹ä¸€çœ‹æŠ½æ ·ä¸ªä½“çš„åˆ†å¸ƒã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">control &lt;- controlPopulation$Bodyweight</span><br><span class="line">hf &lt;- hfPopulation$Bodyweight</span><br><span class="line"><span class="comment"># ç›´æ–¹å›¾</span></span><br><span class="line">op &lt;- par(mfrow = <span class="built_in">c</span>(<span class="number">2</span>,<span class="number">1</span>))</span><br><span class="line">hist(control)</span><br><span class="line">hist(hf)</span><br><span class="line">par(op)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>æ›´ä¸ºç›´è§‚çš„æ˜¯ï¼Œç”»ä¸€ä¸ªQQï¼ˆquantieï¼Œåˆ†ä½æ•°ï¼‰å›¾ã€‚åˆ†ä½æ•°çš„å®ä¾‹ä¹‹ä¸€å°±æ˜¯å››åˆ†ä½æ•°ã€‚QQå›¾çš„ä¸ªäººç†è§£ä¸ºï¼Œå¯¹äºxè½´ä¸Šçš„ä¸€ç‚¹ï¼Œèƒ½å¾—åˆ°xåˆ†å¸ƒä¸­å°äºç­‰äºè¿™ä¸ªç‚¹çš„æ¦‚ç‡ï¼ˆæ¯”å¦‚5%ï¼‰; å†æ‰¾åˆ°yåˆ†å¸ƒä¸­å¯¹åº”5%æ¦‚ç‡çš„å®é™…æ•°å€¼ï¼Œå³yå€¼ï¼Œè¿™æ ·å°±å¾—åˆ°äº†ä¸€ä¸ªç‚¹ï¼ˆx,y)ã€‚</p>
<ul>
<li>å¦‚æœæ˜¯ç¦»æ•£å˜é‡ï¼Œ</li>
</ul>
<blockquote>
<p>åœ¨ç»Ÿè®¡å­¦ä¸­ï¼ŒQQå›¾[1] ï¼ˆQä»£è¡¨åˆ†ä½æ•°Quantileï¼‰æ˜¯ä¸€ç§é€šè¿‡æ¯”è¾ƒä¸¤ä¸ªæ¦‚ç‡åˆ†å¸ƒçš„åˆ†ä½æ•°å¯¹è¿™ä¸¤ä¸ªæ¦‚ç‡åˆ†å¸ƒè¿›è¡Œæ¯”è¾ƒçš„æ¦‚ç‡å›¾æ–¹æ³•ã€‚é¦–å…ˆé€‰å®šåˆ†ä½æ•°çš„å¯¹åº”æ¦‚ç‡åŒºé—´é›†åˆï¼Œåœ¨æ­¤æ¦‚ç‡åŒºé—´ä¸Šï¼Œç‚¹(x,y)å¯¹åº”äºç¬¬ä¸€ä¸ªåˆ†å¸ƒçš„ä¸€ä¸ªåˆ†ä½æ•°xå’Œç¬¬äºŒä¸ªåˆ†å¸ƒåœ¨å’Œxç›¸åŒæ¦‚ç‡åŒºé—´ä¸Šç›¸åŒçš„åˆ†ä½æ•°ã€‚å› æ­¤ç”»å‡ºçš„æ˜¯ä¸€æ¡å«å‚æ•°çš„æ›²çº¿ï¼Œå‚æ•°ä¸ºæ¦‚ç‡åŒºé—´çš„åˆ†å‰²æ•°ã€‚</p>
</blockquote>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">op &lt;- par(mfrow = <span class="built_in">c</span>(<span class="number">2</span>,<span class="number">1</span>))</span><br><span class="line">qqnorm(hf)</span><br><span class="line">qqline(hf)</span><br><span class="line">qqnorm(control)</span><br><span class="line">qqline(control)</span><br><span class="line">par(op)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>åœ¨Rä¸­ï¼Œvar() å’Œ sd() å‡½æ•°éƒ½æ˜¯ç”¨äºè®¡ç®—æ ·æœ¬å‚æ•°ï¼ˆå³é™¤ä»¥n-1ï¼‰ã€‚ä¸é€‚ç”¨äºè®¡ç®—ç¾¤ä½“å‚æ•°ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">x &lt;- control</span><br><span class="line">N &lt;- <span class="built_in">length</span>(x)</span><br><span class="line">populationvar &lt;- mean((x-mean(x))^<span class="number">2</span>)</span><br><span class="line">identical(var(x),populationvar)</span><br><span class="line"><span class="comment">#identical åˆ¤æ–­ä¸¤ä¸ªå€¼æ˜¯å¦ç›¸ç­‰ã€‚</span></span><br><span class="line"><span class="comment">#Rä¸­çš„var() å’Œ sd() éƒ½æ˜¯é™¤ä»¥n-1ã€‚</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä½†æ˜¯å®é™…ä¸­ï¼Œæˆ‘ä»¬æ²¡æœ‰ç”¨åˆ°è®¡ç®—æ€»ä½“å‚æ•°çš„åœ°æ–¹ï¼ˆå¾—ä¸åˆ°ï¼‰ã€‚è¿™é‡Œä¸ºäº†æ•™å­¦ï¼Œç”¨äº†popvar(), popsd() å‡½æ•° è®¡ç®—æ€»ä½“å‚æ•°ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">&#x27;rafalib&#x27;</span>)</span><br><span class="line">sd_hf &lt;- popsd(hf)</span><br><span class="line">sd_control &lt;- popsd(control)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä¸€èˆ¬æˆ‘ä»¬ä¹Ÿåªèƒ½å¾—åˆ°æ ·æœ¬ï¼Œå¦‚ä¸‹ï¼š</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">N &lt;- 12</span><br><span class="line">hf_sample &lt;- sample(hf,<span class="number">12</span>)</span><br><span class="line">control_sample &lt;- sample(control,<span class="number">12</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä¸ºäº† è¿›ä¸€æ­¥äº†è§£ä¸­å¿ƒæé™å®šç†ï¼Œè¿™é‡Œé‡‡ç”¨ä¸åŒçš„æŠ½æ ·å¤§å°è¿›è¡ŒæŠ½æ ·</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#å¯¹ä¸åŒåœ°æŠ½æ ·ä¸ªä½“æ•°ï¼ŒæŠ½æ ·10000æ¬¡</span></span><br><span class="line"><span class="comment">#sapple å¯¹Nsä¸­çš„æ¯ä¸€ä¸ªæ•°ï¼Œæ‰§è¡Œåé¢çš„å‡½æ•°</span></span><br><span class="line">Ns&lt;-<span class="built_in">c</span>(<span class="number">3</span>,<span class="number">12</span>,<span class="number">25</span>,<span class="number">50</span>)</span><br><span class="line">B &lt;-  10000   <span class="comment">#number of simulations</span></span><br><span class="line">res&lt;-sapply(Ns,<span class="keyword">function</span>(n) &#123;</span><br><span class="line">  replicate(B,mean(sample(hf,n))-mean(sample(control,n)))</span><br><span class="line">&#125;)</span><br><span class="line">res</span><br><span class="line"></span><br><span class="line"><span class="comment">## ç°åœ¨æˆ‘ä»¬å¯ä»¥ç”¨qqå›¾çœ‹ä¸­å¿ƒæé™å®šç†çš„æ•ˆæœå¦‚ä½•</span></span><br><span class="line">op &lt;- par(mfrow=<span class="built_in">c</span>(<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line"><span class="keyword">for</span>(i <span class="keyword">in</span> seq(along=Ns)) &#123;    <span class="comment"># ç­‰äº seq(length(Ns))</span></span><br><span class="line">  titleavg&lt;-<span class="built_in">signif</span>(mean(res[,i]),<span class="number">3</span>) <span class="comment">#signifå’Œroundå·®ä¸å¤šï¼Œä¸è¿‡åé¢ä¸æ˜¯å°æ•°ä½æ•°ï¼Œè€Œæ˜¯æœ‰æ•ˆæ•°å­—ã€‚</span></span><br><span class="line">  titlesd&lt;-<span class="built_in">signif</span>(popsd(res[,i]),<span class="number">3</span>)</span><br><span class="line">  title&lt;-paste0(<span class="string">&quot;N=&quot;</span>,Ns[i],<span class="string">&quot;Avg=&quot;</span>,titleavg,<span class="string">&quot;SD=&quot;</span>,titlesd)</span><br><span class="line">  qqnorm(res[,i],main=title)</span><br><span class="line">  qqline(res[,i],col=<span class="number">2</span>)</span><br><span class="line">&#125;</span><br><span class="line"><span class="comment">#éšç€æŠ½æ ·æ•°ç›®çš„å¢åŠ ï¼Œæ ‡å‡†è¯¯è¶Šå°ã€‚åŸºæœ¬ç¬¦åˆæ ¹å·Nçš„è§„å¾‹ã€‚</span></span><br><span class="line"><span class="comment">#å³ä¾¿æŠ½æ ·æ•°ç›®ä¸º3ï¼Œä¹ŸåŸºæœ¬ç¬¦åˆæ­£æ€åˆ†å¸ƒã€‚</span></span><br><span class="line"><span class="comment">#è¿™å’Œæ€»ä½“åˆ†å¸ƒæœ‰å…³ã€‚å› ä¸ºæˆ‘ä»¬è¿™ä¸ªå°é¼ ä½“é‡åˆ†å¸ƒæœ¬èº«å°±æ¥è¿‘æ­£æ€åˆ†å¸ƒã€‚å“ªæ€•é‡‡æ ·ä¸ªä½“ä¸º1ï¼Œä¹Ÿä¼šç¬¦åˆæ­£æ€åˆ†å¸ƒã€‚</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä½†æ˜¯åœ¨å®è·µä¸­ï¼Œæˆ‘ä»¬ä¸ä¼šåªæ±‚å‡å€¼ï¼Œè€Œæ˜¯ä¼šé™¤ä»¥æ ‡å‡†å·®ã€‚æ­¤æ—¶é‡‡æ ·ä¸ªæ•°å°±é‡è¦äº†</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">Ns&lt;-<span class="built_in">c</span>(<span class="number">3</span>,<span class="number">12</span>,<span class="number">25</span>,<span class="number">50</span>)</span><br><span class="line">B&lt;-<span class="number">10000</span><span class="comment">#number of simulations#</span></span><br><span class="line"><span class="comment">#function to compute a t-stat</span></span><br><span class="line">computetstat&lt;-<span class="keyword">function</span>(n) &#123;</span><br><span class="line">  y&lt;-sample(hfPopulation,n)</span><br><span class="line">  x&lt;-sample(controlPopulation,n)</span><br><span class="line">  (mean(y)-mean(x))/<span class="built_in">sqrt</span>(var(y)/n+var(x)/n)</span><br><span class="line">  &#125;</span><br><span class="line">res&lt;-sapply(Ns,<span class="keyword">function</span>(n) &#123;</span><br><span class="line">  replicate(B,computetstat(n))</span><br><span class="line">  &#125;)</span><br><span class="line">op &lt;- par(mfrow=<span class="built_in">c</span>(<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line"><span class="keyword">for</span>(i <span class="keyword">in</span> seq(along=Ns)) &#123;</span><br><span class="line">  qqnorm(res[,i],main=Ns[i])</span><br><span class="line">  qqline(res[,i],col=<span class="number">2</span>)</span><br><span class="line">  &#125;</span><br><span class="line"><span class="comment">#æ­¤æ—¶N=3å·²ç»æœ‰äº›åç¦»ï¼Œè€ŒN=12åªæœ‰è½»å¾®åç¦»</span></span><br><span class="line"><span class="comment">#å› æ­¤N=12å·²ç»å¤Ÿäº†</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä½†æ˜¯è¿™ç§æ¨¡æ‹Ÿåœ¨å®é™…ä¸­æ— æ³•åšåˆ°ã€‚è¿™é‡Œåªç”¨äºè¯´æ˜CLTçš„æ¦‚å¿µå’Œå±€é™ã€‚</p>
<h3 id="t-tests-in-Partice">t-tests in Partice</h3>
<p>é€»è¾‘çº¿ï¼š</p>
<ol>
<li>é¦–å…ˆæˆ‘ä»¬å¾—åˆ°äº†ä¸¤ä¸ªæ ·æœ¬ç¾¤ä½“çš„å‡å€¼å·®ï¼Œè¿™æœ¬èº«ä¹Ÿæ˜¯ä¸€ä¸ªéšæœºå˜é‡ã€‚è€Œä¸”æ ¹æ®é›¶å‡è®¾ï¼Œè¿™ä¸ªå˜é‡åˆ†å¸ƒçš„å‡å€¼ä¸º0ã€‚</li>
<li>é‚£ä¹ˆè¿™ä¸ªå‡å€¼å·®åˆ†å¸ƒçš„æ ‡å‡†è¯¯ï¼ˆå·®ï¼‰å‘¢ï¼Ÿ</li>
<li>é¦–å…ˆæˆ‘ä»¬æ ¹æ®ä¸­å¿ƒæé™å®šç†ï¼Œæ¯ä¸ªç¾¤ä½“å‡å€¼æœä»æ­£æ€åˆ†å¸ƒï¼Œæ ‡å‡†è¯¯ä¸ºæ€»ä½“æ ‡å‡†å·®ï¼ˆè¿™é‡Œç›´æ¥ç”¨æ ·æœ¬æ ‡å‡†å·®ä¼°è®¡æ€»ä½“æ ‡å‡†å·®ï¼‰é™¤ä»¥æ ¹å·Nã€‚</li>
<li>ä½†è¿™ä»…ä»…æ˜¯æ¯ä¸ªæ ·æœ¬å‡å€¼çš„æ ‡å‡†è¯¯ï¼ˆæˆ–è€…è¯´æ ‡å‡†å·®ï¼Œå…¶å®æˆ‘è§‰å¾—è¿™é‡Œè¯´æ ‡å‡†è¯¯å’Œæ ‡å‡†å·®æ˜¯ä¸€å›äº‹ï¼Œæ ‡å‡†è¯¯å°±æ˜¯å‡å€¼çš„æ ‡å‡†å·®ï¼Œæ‰€ä»¥ä¸è¿˜æ˜¯æ ‡å‡†å·®å—ï¼Ÿè¿™é‡Œè¯´çš„å‡å€¼å·®çš„æ ‡å‡†è¯¯å…¶å®ä¹Ÿå°±æ˜¯å‡å€¼å·®çš„æ ‡å‡†å·®ã€‚ï¼‰</li>
<li>æˆ‘ä»¬æ ¹æ®ç»Ÿè®¡ç†è®ºï¼Œæˆ‘ä»¬çŸ¥é“ä¸¤ä¸ªéšæœºå˜é‡å·®å€¼çš„æ–¹å·®ç­‰äºè¿™ä¸¤ä¸ªéšæœºå˜é‡çš„æ–¹å·®ä¹‹å’Œï¼ˆvar(x-y)=var(x)+var(y))</li>
<li>å› æ­¤se = sd( x-y ) = sqrt(var(x)+var(y)) =  sqrt( sd(x)/sqrt(n) + (sd(y)/sqrt(m) )</li>
<li>tstat &lt;- diff(-0)/se</li>
<li>è¿™å°±æ˜¯tç»Ÿè®¡é‡ï¼Œç”±äºé™¤ä»¥äº†seï¼Œè¿™ä¸ªç»Ÿè®¡é‡çš„SEä¸º1ã€‚</li>
<li>å› æ­¤æˆ‘ä»¬åªè¦çŸ¥é“è¿™ä¸ªtç»Ÿè®¡é‡çš„åˆ†å¸ƒï¼Œæˆ‘ä»¬å°±èƒ½å¾—åˆ°På€¼ã€‚</li>
<li>æ ¹æ®ä¸­å¿ƒæé™å®šç†ï¼Œå½“<strong>æŠ½æ ·ä¸ªä½“æ•°è¾ƒå¤§</strong>æ—¶ï¼Œä¸¤ä¸ªæŠ½æ ·å‡å€¼å‡ç¬¦åˆæ­£æ€åˆ†å¸ƒã€‚æ ¹æ®ç»Ÿè®¡ç†è®ºï¼Œé‚£ä¹ˆè¿™ä¸¤ä¸ªéšæœºå˜é‡ä¹‹å·®ä¹Ÿæœä»æ­£æ€åˆ†å¸ƒã€‚å› æ­¤æ­¤æ—¶è¿™ä¸ªtstatä¹Ÿæ˜¯æœä»<strong>å‡å€¼ä¸º0ï¼Œæ–¹å·®ä¸º1çš„æ­£æ€åˆ†å¸ƒ</strong></li>
<li>é‚£ä¹ˆè®¡ç®—ä¸ºäº†è®¡ç®—På€¼ï¼Œæˆ‘ä»¬éœ€è¦é—®ï¼šåœ¨æ­£æ€åˆ†å¸ƒä¸­ï¼Œå‡ºç°diffæˆ–æ¯”diffï¼ˆtstatï¼‰æ›´æç«¯çš„å€¼çš„æ¦‚ç‡æ˜¯å¤šå¤§ï¼Ÿï¼ˆ how often does a normally distributed randomvariable exceeddiff?ï¼‰æˆ‘ä»¬å¯ä»¥é€šè¿‡pnorm()å‡½æ•°è¿›è¡Œè®¡ç®—åŒå°¾æ¦‚ç‡ã€‚</li>
<li>åœ¨æœ¬ä¾‹ä¸­ï¼Œp-value å°äº 0.05ï¼Œ æˆ‘ä»¬è®¤ä¸ºå·®å¼‚æ˜¯æ˜¾è‘—çš„ã€‚ï¼ˆæ·±å…¥ç†è§£På€¼ï¼ŒPå€¼æ˜¯é›¶å‡è®¾æˆç«‹ä¸‹è¯¥äº‹ä»¶å‘ç”Ÿçš„æ¦‚ç‡ï¼‰</li>
<li>ä½†æ˜¯æˆ‘ä»¬åˆé¢ä¸´å¦ä¸€ä¸ªé—®é¢˜ã€‚ä¸­å¿ƒæé™å®šç†åªæœ‰åœ¨æ ·æœ¬é‡è¾ƒå¤§æ—¶é€‚ç”¨ï¼Œä½†æ˜¯12è¶³å¤Ÿå¤§äº†å—ï¼Ÿä¸€èˆ¬æˆ‘ä»¬è®¤ä¸ºæ ·æœ¬æ•°å¤§äº30æ‰æ¯”è¾ƒå¥½ã€‚æ‰€ä»¥æˆ‘ä»¬åˆšåˆšè®¡ç®—çš„På€¼åªæœ‰åœ¨è¿™ä¸ªæˆç«‹æ—¶æ‰æœ‰æ•ˆï¼Œæ‰€ä»¥è¿™é‡Œå¯èƒ½ä¸é€‚ç”¨ã€‚ç„¶è€Œï¼Œé™¤äº†ä½¿ç”¨ä¸­å¿ƒæé™å®šç†ï¼Œæˆ‘ä»¬è¿™é‡Œæœ‰å¦ä¸€ç§æ–¹æ³•ã€‚</li>
</ol>
<p>Rä»£ç å¦‚ä¸‹</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br></pre></td><td class="code"><pre><span class="line">dat&lt;-read.csv(<span class="string">&quot;femaleMiceWeights.csv&quot;</span>)<span class="comment">#previously downloaded</span></span><br><span class="line">control&lt;-filter(dat,Diet==<span class="string">&quot;chow&quot;</span>)%&gt;%select(Bodyweight)%&gt;%unlist</span><br><span class="line">treatment&lt;-filter(dat,Diet==<span class="string">&quot;hf&quot;</span>)%&gt;%select(Bodyweight)%&gt;%unlist</span><br><span class="line">diff&lt;-mean(treatment)-mean(control)</span><br><span class="line">print(diff)</span><br><span class="line"><span class="comment">## [1] 3.020833</span></span><br><span class="line"></span><br><span class="line">se&lt;-<span class="built_in">sqrt</span>(</span><br><span class="line">  var(treatment)/<span class="built_in">length</span>(treatment) + </span><br><span class="line">    var(control)/<span class="built_in">length</span>(control)</span><br><span class="line">  )</span><br><span class="line"></span><br><span class="line">tstat&lt;-diff/se</span><br><span class="line"><span class="comment">## [1] 2.055174</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##tstatåœ¨æ ‡å‡†æ­£æ€åˆ†å¸ƒä¸­å‡ºç°çš„æ¦‚ç‡(åŒå°¾)</span></span><br><span class="line"><span class="comment">##ä¹Ÿå°±æ˜¯åœ¨æ ‡å‡†æ­£æ€åˆ†å¸ƒä¸­å‡ºç°tstatåŠæ¯”tstatæ›´æç«¯çš„å€¼çš„æ¦‚ç‡</span></span><br><span class="line">righttail&lt;-<span class="number">1</span>-pnorm(<span class="built_in">abs</span>(tstat))</span><br><span class="line">lefttail&lt;-pnorm(-<span class="built_in">abs</span>(tstat))</span><br><span class="line">pval&lt;-lefttail+righttail</span><br><span class="line">print(pval)</span><br><span class="line"></span><br><span class="line"><span class="comment">## [1] 0.0398622</span></span><br></pre></td></tr></table></figure>
<h3 id="The-t-distribution-in-Practice">The t-distribution in Practice</h3>
<p>æ ¹æ®ä¸Šé¢çš„æ¨¡æ‹Ÿæƒ…å†µå’Œç»Ÿè®¡ç†è®ºã€‚ å¦‚æœ<strong>æ€»ä½“çš„åˆ†å¸ƒæ˜¯æ­£æ€çš„</strong>ï¼Œé‚£ä¹ˆæˆ‘ä»¬å¯ä»¥<strong>ä¸é€šè¿‡ä¸­å¿ƒæé™å®šç†</strong>ï¼Œè€Œå¼„æ¸…æ¥štç»Ÿè®¡é‡çš„ç¡®åˆ‡åˆ†å¸ƒã€‚ä½†æ˜¯å¦‚æœæŠ½æ ·ä¸ªä½“æ•°å¾ˆå°‘ï¼Œå¾ˆéš¾è¯æ˜æ€»ä½“åˆ†å¸ƒæ˜¯å¦ç¬¦åˆæ­£æ€ã€‚ä½†æ˜¯å¯¹äºä½“é‡è¿™ç§æ•°æ®ï¼Œæ ¹æ®ç»éªŒï¼Œæˆ‘ä»¬ä¸€èˆ¬è®¤ä¸ºæ€»ä½“åˆ†å¸ƒæå¯èƒ½å¤§è‡´ç¬¦åˆæ­£æ€åˆ†å¸ƒï¼Œå› æ­¤æˆ‘ä»¬å¯ä»¥åˆ©ç”¨è¿™ç§è¿‘ä¼¼ã€‚å¦å¤–ï¼Œæˆ‘ä»¬ä¹Ÿå¯ä»¥çœ‹ä¸€ä¸‹æ ·æœ¬çš„qqå›¾ï¼Œå¦‚ä¸‹æ‰€ç¤ºã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">op &lt;- par(mfrow=<span class="built_in">c</span>(<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line">qqnorm(treatment)</span><br><span class="line">qqline(treatment,col=<span class="number">2</span>)</span><br><span class="line">qqnorm(control)</span><br><span class="line">qqline(control,col=<span class="number">2</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>å¦‚æœæˆ‘ä»¬è¿‘ä¼¼è®¤ä¸º<strong>æ€»ä½“åˆ†å¸ƒç¬¦åˆæ­£æ€åˆ†å¸ƒ</strong>ï¼Œé‚£ä¹ˆåŸºäºç»Ÿè®¡ç†è®ºï¼Œtstatæœä»tåˆ†å¸ƒã€‚ç›¸è¾ƒäºæ­£æ€åˆ†å¸ƒï¼Œtåˆ†å¸ƒè¿˜æœ‰ä¸€ä¸ªè‡ªç”±åº¦çš„å‚æ•°ã€‚æˆ‘ä»¬ç›´æ¥é€šè¿‡ä¸‹é¢çš„å‡½æ•°æ±‚è§£På€¼ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">t.test(treatment,control)</span><br><span class="line"></span><br><span class="line">t.test(treatment,control)$p.value <span class="comment">### ç›´æ¥çœ‹På€¼</span></span><br><span class="line"><span class="comment">## [1] 0.05299888</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<blockquote>
<p>The p-value is slightly bigger now. This is to be expected because our CLT approximation consideredthe denominator oftstatpractically fixed (with large samples it practically is), while the t-distribution approximation takes into account that the denominator (the standard error of thedifference) is a random variable. The smaller the sample size, the more the denominator varies.</p>
</blockquote>
<p>è¿™æ®µè¯æ²¡çœ‹å¤ªæ‡‚ï¼Œå¤§æ¦‚ç†è§£æ˜¯ï¼šè¿™é‡Œç”¨tåˆ†å¸ƒè®¡ç®—å‡ºæ¥çš„på€¼ç›¸è¾ƒäºä¸Šé¢ç”¨æ­£æ€åˆ†å¸ƒè®¡ç®—å‡ºæ¥çš„På€¼è¾ƒå¤§ã€‚è¿™æ˜¯å› ä¸ºåœ¨CLTä¸­ï¼Œè®¤ä¸ºtstatçš„åˆ†æ¯éƒ¨åˆ†æ˜¯å›ºå®šçš„ï¼ˆä¹Ÿå°±æ˜¯è®¤ä¸ºé‡‡ç”¨äº†å¤§æ ·æœ¬ï¼‰ï¼›è€Œtåˆ†å¸ƒæ£€éªŒä¸­è®¤ä¸ºtstatçš„åˆ†æ¯éƒ¨åˆ†æ˜¯éšæœºå˜é‡ï¼ˆè¿˜ä¸æ˜¯å¾ˆæ‡‚ï¼‰ã€‚</p>
<p>è¿™å¯èƒ½ä¼šè®©äººå›°æƒ‘ï¼Œæ€ä¹ˆä¼šå¾—åˆ°ä¸¤ä¸ªPå€¼ï¼Ÿä½†æ˜¯ï¼Œè¿™åœ¨æ•°æ®åˆ†æä¸­æ˜¯å¾ˆå¸¸è§çš„ã€‚é€šè¿‡ä¸åŒçš„å‡è®¾ï¼Œä¸åŒçš„æ–¹æ³•ï¼Œæˆ‘ä»¬å°±ä¼šå¾—åˆ°ä¸åŒçš„ç»“æœã€‚</p>
<p>å…¶å®ï¼ŒåŸºäºä¸­å¿ƒæé™å®šç†å¾—åˆ°çš„ç»“æœæ›´å®¹æ˜“å‡ºç°<strong>å‡é˜³æ€§</strong>ï¼Œè€ŒåŸºäºtåˆ†å¸ƒçš„ç»“æœæ›´å®¹æ˜“å‡ºç°<strong>å‡é˜´æ€§</strong>ã€‚</p>
<h3 id="Confidence-Intervals">Confidence Intervals</h3>
<p>åœ¨ç”Ÿå‘½ç§‘å­¦ä¸­ï¼Œæˆ‘ä»¬æ€»æ˜¯éœ€è¦è®¡ç®—På€¼ã€‚ä½†æ˜¯æˆ‘ä»¬å¹¶ä¸æ¨å´‡æŠŠPå€¼å½“æˆå”¯ä¸€çš„ç»Ÿè®¡æŒ‡æ ‡ã€‚åŸå› å¾ˆç®€å•ï¼šç»Ÿè®¡æ„ä¹‰ä¸Šçš„æ˜¾è‘—å¹¶ä¸ä¿è¯ç§‘å­¦æ„ä¹‰ä¸Šçš„æ˜¾è‘—ã€‚å½“æŠ½æ ·ä¸ªä½“æ•°å¾ˆå¤§æ—¶ï¼Œå³ä¾¿ä¸¤ä¸ªç¾¤ä½“å·®å¼‚å¾ˆå°ï¼Œä½ ä»å¯ä»¥æ£€æµ‹åˆ°ä¸€ä¸ªç»Ÿè®¡æ„ä¹‰ä¸Šçš„æ˜¾è‘—å·®å¼‚ã€‚ä½†æ˜¯è¿™æœ‰æ„ä¹‰å—ï¼Ÿå¦‚æœåªçœ‹På€¼ï¼Œä¿¡æ¯æ˜¯ä¸å¤Ÿçš„ã€‚è¿˜éœ€è¦ç»™å‡ºæ•ˆåº”å¤§å°ï¼ˆè¿™é‡Œæ˜¯ä¸¤ä¸ªç¾¤ä½“å‡å€¼å·®ï¼‰ã€‚<br>
å¦ä¸€ç§æ–¹æ³•æ˜¯ç»™å‡ºç½®ä¿¡åŒºé—´ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#ä»¥controlç¾¤ä½“å‡å€¼çš„ç½®ä¿¡åŒºé—´ä¸ºä¾‹</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#å…ˆçœ‹æ€»ä½“çš„å‡å€¼</span></span><br><span class="line">dat&lt;-read.csv(<span class="string">&quot;mice_pheno.csv&quot;</span>)</span><br><span class="line">chowPopulation&lt;-dat[dat$Sex==<span class="string">&quot;F&quot;</span>&amp;dat$Diet==<span class="string">&quot;chow&quot;</span>,<span class="number">3</span>]</span><br><span class="line"></span><br><span class="line">mu_chow&lt;-mean(chowPopulation)</span><br><span class="line">print(mu_chow)</span><br><span class="line"><span class="comment">## [1] 23.89338</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##æˆ‘ä»¬æ˜¯è¦ä¼°è®¡è¿™ä¸ªå‚æ•°ï¼Œä½†æ˜¯æˆ‘ä»¬å®é™…æ‹¿ä¸åˆ°æ€»ä½“çš„æ•°æ®</span></span><br><span class="line"><span class="comment">## å…ˆæŠ½å–æ ·æœ¬é‡ä¸º30çš„æ ·æœ¬</span></span><br><span class="line"></span><br><span class="line">N&lt;-<span class="number">30</span></span><br><span class="line">chow&lt;-sample(chowPopulation,N)</span><br><span class="line">print(mean(chow))</span><br><span class="line"><span class="comment">## [1] 24.54567</span></span><br><span class="line"></span><br><span class="line"></span><br><span class="line"><span class="comment">#æˆ‘ä»¬çŸ¥é“æ ·æœ¬å‡å€¼æ˜¯ä¸€ä¸ªéšæœºå˜é‡</span></span><br><span class="line"><span class="comment">#æ‰€ä»¥ï¼Œè¿™ä¸ªå€¼ä¸æ˜¯ä¸€ä¸ªå®Œç¾çš„ä¼°è®¡</span></span><br><span class="line"><span class="comment">#é€šè¿‡ç½®ä¿¡åŒºé—´æ¥ä½“ç°è¿™ä¸ªéšæœºå˜é‡çš„variability</span></span><br><span class="line"></span><br><span class="line"><span class="comment">#å› ä¸ºæ ·æœ¬é‡ä¸º30ï¼Œæˆ‘ä»¬å¯ä»¥ä½¿ç”¨CLTã€‚</span></span><br><span class="line"><span class="comment">#ä¸­å¿ƒæé™å®šç†å‘Šè¯‰æˆ‘ä»¬ï¼Œæ ·æœ¬å‡å€¼æœä»ä¸€ä¸ªå‡å€¼ä¸ºæ€»ä½“å‡å€¼ï¼Œæ ‡å‡†å·®ä¸ºæ€»ä½“æ ‡å‡†å·®é™¤ä»¥æ ¹å·30çš„æ­£æ€åˆ†å¸ƒã€‚</span></span><br><span class="line"></span><br><span class="line">se&lt;-sd(chow)/<span class="built_in">sqrt</span>(N)</span><br><span class="line">print(se)</span><br><span class="line"><span class="comment">## [1] 0.6610814</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<blockquote>
<p>A 95% confidence interval (we can use percentages other than 95%) is a random interval with a95% probability of falling on the parameter we are estimating. Keep in mind that saying 95% ofrandom intervals will fall on the true value (our definition above) isnot the sameas saying there isa 95% chance that the true value falls in our interval.(æœ‰ç‚¹ç»•ï¼Œæ²¡å¤ªæ‡‚ã€‚æ‡‚äº†ï¼Œè¿™ä¸ªç½®ä¿¡åŒºé—´å…¶å®ä¹Ÿæ˜¯ä¸€ä¸ªéšæœºå˜é‡ï¼Œä½ å†æŠ½ä¸€ä¸ªæ ·ï¼Œè¿™ä¸ªç½®ä¿¡åŒºé—´ä¹Ÿä¼šå˜ã€‚æ‰€ä»¥ä½ åªèƒ½è¯´ç½®ä¿¡åŒºé—´ï¼ˆéšæœºå˜é‡ï¼‰ä»ç»Ÿè®¡ä¸Šè¯´æœ‰95%çš„æ¦‚ç‡åŒ…å«çœŸå€¼ï¼ˆä½ è¦æ˜¯å–10000ä¸ªæ ·æœ¬ï¼Œè®¡ç®—å‡º10000ä¸ªç½®ä¿¡åŒºé—´ï¼Œå…¶ä¸­åŒ…å«æ€»ä½“å‡å€¼çš„æ¦‚ç‡åœ¨95%å·¦å³ï¼‰ã€‚ä½†æ˜¯ï¼Œå¦‚æœæ˜¯ä¸€ä¸ªç»™å®šçš„åŒºé—´(ä½ æœ€åè®¡ç®—å‡ºæ¥çš„åŒºé—´)ï¼Œé‚£ä¹ˆçœŸå€¼å­˜åœ¨è¿™ä¸ªåŒºé—´çš„æ¦‚ç‡åªæœ‰0å’Œ1)</p>
</blockquote>
<p>ç”±ä¸­å¿ƒæé™å®šç†ï¼Œæˆ‘ä»¬å¯ä»¥å¯¹æ ·æœ¬å‡å€¼è¿›è¡Œæ ‡å‡†åŒ–ã€‚ç„¶åæˆ‘ä»¬çŸ¥é“æ ‡å‡†æ­£æ€åˆ†å¸ƒçš„95%åŒºé—´å¤§è‡´åœ¨[-2,2]ã€‚å› æ­¤å¯ä»¥å¾—å‡ºä¸‹å¼</p>
<p><img src="confidence_interval.png" alt="confidence_interval"></p>
<p>æˆ‘ä»¬ç›´æ¥å¾—å‡ºç½®ä¿¡åŒºé—´</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Q&lt;-qnorm(<span class="number">1</span>-<span class="number">0.05</span>/<span class="number">2</span>)</span><br><span class="line">interval&lt;-<span class="built_in">c</span>(mean(chow)-Q*se, mean(chow)+Q*se )</span><br><span class="line">interval</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>æ¨¡æ‹Ÿnæ¬¡ç½®ä¿¡åŒºé—´</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">library(rafalib)</span><br><span class="line">B&lt;-<span class="number">250</span></span><br><span class="line">mypar()</span><br><span class="line">plot(mean(chowPopulation)+<span class="built_in">c</span>(-<span class="number">7</span>,<span class="number">7</span>),<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>),type=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">	xlab=<span class="string">&quot;weight&quot;</span>,ylab=<span class="string">&quot;interval&quot;</span>,ylim=<span class="built_in">c</span>(<span class="number">1</span>,B))</span><br><span class="line">	abline(v=mean(chowPopulation))</span><br><span class="line">	<span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:B) &#123;</span><br><span class="line">	chow&lt;-sample(chowPopulation,N)</span><br><span class="line">	se&lt;-sd(chow)/<span class="built_in">sqrt</span>(N)</span><br><span class="line">	interval&lt;-<span class="built_in">c</span>(mean(chow)-Q*se, mean(chow)+Q*se)</span><br><span class="line">	covered&lt;-</span><br><span class="line">		mean(chowPopulation) &lt;=interval[<span class="number">2</span>]&amp;mean(chowPopulation)&gt;=interval[<span class="number">1</span>]</span><br><span class="line">	color&lt;-ifelse(covered,<span class="number">1</span>,<span class="number">2</span>)</span><br><span class="line">	lines(interval, <span class="built_in">c</span>(i,i),col=color)</span><br><span class="line">&#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>è¿™ä¸ªä»£ç ç”»å‡ºäº†250æ¬¡ç½®ä¿¡åŒºé—´çš„å›¾ï¼ˆä»£ç å†…å®¹æ²¡çœ‹ï¼‰ã€‚å…¶ä¸­å¤§æ¦‚æœ‰5%çš„ç½®ä¿¡åŒºé—´æ²¡æœ‰åŒ…å«æ€»ä½“å‡å€¼ã€‚</p>
<p>ä½†æ˜¯ä¸Šé¢æ˜¯å› ä¸ºæŠ½æ ·ä¸ªä½“è¾ƒå¤šï¼ˆN = 30ï¼‰ï¼Œæ»¡è¶³äº†ä¸­å¿ƒæé™å®šç†ã€‚å¦‚æœ N = 5ï¼Œé‚£ä¹ˆè¿™æ ·åšä¹Ÿæœ‰æ•ˆå—ï¼Ÿæˆ‘ä»¬å°±æ¨¡æ‹Ÿè¿™ç§N =5ï¼Œä½¿ç”¨CLTçš„æƒ…å†µã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">N &lt;- 5</span><br><span class="line">mypar()</span><br><span class="line">plot(mean(chowPopulation)+<span class="built_in">c</span>(-<span class="number">7</span>,<span class="number">7</span>),<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>),type=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">     xlab=<span class="string">&quot;weight&quot;</span>,ylab=<span class="string">&quot;interval&quot;</span>,ylim=<span class="built_in">c</span>(<span class="number">1</span>,B))</span><br><span class="line">abline(v=mean(chowPopulation))</span><br><span class="line"><span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:B)&#123;</span><br><span class="line">  chow&lt;-sample(chowPopulation,N)</span><br><span class="line">  se&lt;-sd(chow)/<span class="built_in">sqrt</span>(N)</span><br><span class="line">  interval&lt;-<span class="built_in">c</span>(mean(chow)-Q*se, mean(chow)+Q*se)</span><br><span class="line">  covered&lt;- mean(chowPopulation) &lt;=interval[<span class="number">2</span>]&amp;mean(chowPopulation)&gt;=interval[<span class="number">1</span>]</span><br><span class="line">  color &lt;- ifelse(covered,<span class="number">1</span>,<span class="number">2</span>)</span><br><span class="line">  lines(interval, <span class="built_in">c</span>(i,i),col=color)</span><br><span class="line">&#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>æˆ‘ä»¬å‘ç°ç½®ä¿¡åŒºé—´çš„å®½åº¦å¢åŠ äº†ï¼Œä½†æ˜¯æ²¡æœ‰è¦†ç›–çœŸå€¼çš„æƒ…å†µå´å¤šäº†å¾ˆå¤šã€‚è¿™æ˜¯å› ä¸ºè¿™é‡Œæˆ‘ä»¬é€šè¿‡CLTé”™è¯¯åœ°è®¤ä¸ºæ ·æœ¬å‡å€¼ç¬¦åˆæ­£æ€åˆ†å¸ƒï¼Œä½†æ˜¯å®é™…ä¸Šç›¸è¾ƒäºæ­£æ€å®ƒä¸¤ä¾§çš„å°¾å·´æ›´å¤§ï¼ˆIt has a fatter tailï¼‰ã€‚è¿™å½±å“åˆ°äº†æˆ‘ä»¬Qå€¼çš„è®¡ç®—ï¼ˆè¿™é‡Œæ˜¯-2åˆ°2ï¼‰ã€‚è¿™ç§æƒ…å†µä¸‹ï¼Œé‡‡ç”¨tåˆ†å¸ƒæ›´åˆé€‚ã€‚æ‰€ä»¥æˆ‘ä»¬åªéœ€ç”¨qt() å‡½æ•°æ¥è®¡ç®—Qå°±å¯ä»¥äº†ï¼Œå…¶å®ƒä¸å˜ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">N &lt;- 5</span><br><span class="line">Q&lt;-qt(<span class="number">1</span>-<span class="number">0.05</span>/<span class="number">2</span>, df=<span class="number">4</span>)</span><br><span class="line">mypar()</span><br><span class="line">plot(mean(chowPopulation)+<span class="built_in">c</span>(-<span class="number">7</span>,<span class="number">7</span>),<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">1</span>),type=<span class="string">&quot;n&quot;</span>,</span><br><span class="line">     xlab=<span class="string">&quot;weight&quot;</span>,ylab=<span class="string">&quot;interval&quot;</span>,ylim=<span class="built_in">c</span>(<span class="number">1</span>,B))</span><br><span class="line">abline(v=mean(chowPopulation))</span><br><span class="line"><span class="keyword">for</span>(i <span class="keyword">in</span> <span class="number">1</span>:B)&#123;</span><br><span class="line">  chow&lt;-sample(chowPopulation,N)</span><br><span class="line">  se&lt;-sd(chow)/<span class="built_in">sqrt</span>(N)</span><br><span class="line">  interval&lt;-<span class="built_in">c</span>(mean(chow)-Q*se, mean(chow)+Q*se)</span><br><span class="line">  covered&lt;- mean(chowPopulation) &lt;=interval[<span class="number">2</span>]&amp;mean(chowPopulation)&gt;=interval[<span class="number">1</span>]</span><br><span class="line">  color &lt;- ifelse(covered,<span class="number">1</span>,<span class="number">2</span>)</span><br><span class="line">  lines(interval, <span class="built_in">c</span>(i,i),col=color)</span><br><span class="line">&#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>æˆ‘ä»¬çœ‹åˆ°ç½®ä¿¡åŒºé—´çš„å®½åº¦<strong>è¿›ä¸€æ­¥</strong>ï¼ˆç›¸è¾ƒäºCLTï¼‰å¢åŠ äº†ï¼ˆå› ä¸ºtåˆ†å¸ƒå°¾å·´åšï¼Œå¾—åˆ°çš„Qå€¼æ›´å¤§ï¼‰ï¼Œä½†æ˜¯åŸºæœ¬ç¬¦åˆäº†95%çš„æ¦‚ç‡ã€‚</p>
<h3 id="ç½®ä¿¡åŒºé—´å’ŒPå€¼çš„å…³ç³»">ç½®ä¿¡åŒºé—´å’ŒPå€¼çš„å…³ç³»</h3>
<p>æˆ‘ä»¬æ¨èåœ¨å®é™…ä¸­æ›´å¤šçš„ä½¿ç”¨ç½®ä¿¡åŒºé—´ï¼Œè€Œä¸æ˜¯På€¼ã€‚</p>
<p>å¦‚æœä½ ç»™å‡ºPå€¼ï¼Œæˆ‘ä»¬ä¼šé—®å·®è·æ˜¯ä¸æ˜¯çœŸçš„åƒä½ è®¡ç®—æˆ–è§‚æµ‹å‡ºæ¥çš„å€¼é‚£ä¹ˆå¤§ã€‚<strong>å› ä¸ºå³ä¾¿ä¸¤ä¸ªç¾¤ä½“æ€»ä½“å‡å€¼ä¸º0ï¼Œå¾—åˆ°çš„æ ·æœ¬å‡å€¼å·®ä¹Ÿæœ‰å¯èƒ½å‡ºç°ä½ è®¡ç®—æˆ–è§‚æµ‹çš„ç»“æœ</strong>ã€‚å› æ­¤ï¼Œæˆ‘ä»¬å¯ä»¥ç»™å‡ºå‡å€¼å·®çš„ç½®ä¿¡åŒºé—´ï¼Œæ¥å›ç­”è¿™ä¸ªé—®é¢˜ã€‚</p>
<p>å¦‚æœå‡å€¼å·®çš„95%çš„ç½®ä¿¡åŒºé—´ä¸­ä¸åŒ…å«0ï¼Œé‚£ä¹ˆæˆ‘ä»¬å¯ä»¥è¯´å·®å¼‚ä¸æ˜¾è‘—ï¼ˆ0.05æ°´å¹³ä¸‹ï¼‰ï¼Œä¹Ÿå°±æ˜¯på€¼è‚¯å®šå°äº0.05ã€‚ï¼ˆ99%çš„ç½®ä¿¡åŒºé—´ç±»ä¼¼ï¼‰</p>
<p>t.testçš„ç»“æœä¸­åŒ…å«äº†95%çš„ç½®ä¿¡åŒºé—´ã€‚</p>
<h3 id="Power-Calculations">Power Calculations</h3>
<p>ç”±äºæˆ‘ä»¬æœ‰æ€»ä½“ä¿¡æ¯ï¼Œæˆ‘ä»¬ä»æ€»ä½“ä¿¡æ¯ä¸­å¯ä»¥ç›´æ¥å¾—çŸ¥ï¼Œè¿™ä¸¤ä¸ªç¾¤ä½“ç¡®å®æ˜¯æœ‰å·®è·çš„ï¼Œè€Œä¸”æŒ‰æ¯”ä¾‹å¤§çº¦æ˜¯10%ã€‚<br>
ä½†æ˜¯æˆ‘ä»¬å¦‚æœä»ä¸­å–æ ·ï¼Œç„¶åè¿›è¡Œt-testï¼Œæœ‰çš„æ—¶å€™På€¼ä¸æ€»æ˜¯å°äº0.05ã€‚</p>
<p>æ‰€ä»¥æˆ‘ä»¬çŠ¯é”™äº†å—ï¼Ÿé€šè¿‡ä¸æ‹’ç»åŸå‡è®¾ï¼Œæˆ‘ä»¬è¦è¯´dietæ²¡æœ‰å½±å“å—ï¼Ÿè¿™ä¸ªé—®é¢˜çš„ç­”æ¡ˆæ˜¯noã€‚All we can say is that we did not reject the null hypothesis. But this does not necessarily imply that the null is trueï¼ˆæˆ‘ä»¬åªæ˜¯è¯´ä¸æ‹’ç»åŸå‡è®¾ï¼Œä½†æˆ‘ä»¬æ²¡æœ‰è¯´åŸå‡è®¾æ˜¯çœŸçš„ï¼‰.</p>
<p>è¿™ä¸ªé—®é¢˜çš„æœ¬è´¨æ˜¯è¿™æ ·ï¼Œåœ¨è¿™æ¬¡çš„å®ä¾‹ä¸­ï¼Œæˆ‘ä»¬æ²¡æœ‰å¾—åˆ°è¶³å¤Ÿçš„powerï¼ˆä¸çŸ¥é“å•¥æ„æ€ï¼‰ã€‚åœ¨ç§‘å­¦ç ”ç©¶ä¸­ï¼Œä½ å¾ˆå¯èƒ½éœ€è¦åšæŸç§ç¨‹åº¦çš„power calculationã€‚åœ¨å¾ˆå¤šæƒ…å†µä¸‹ï¼Œè¿™æ˜¯ä¸€ç§é“å¾·ä¹‰åŠ¡ï¼Œæ¥é¿å…ä¸å¿…è¦åœ°ç‰ºç‰²æ›´å¤šå°é¼ ï¼Œæˆ–è€…ä½¿äººä»¬å› ä¸ºä½ çš„ç ”ç©¶é­é‡æ½œåœ¨çš„é£é™©ã€‚</p>
<h4 id="types-of-error">types of error</h4>
<blockquote>
<p>â… å‹é”™è¯¯ï¼šåŸå‡è®¾ä¸ºçœŸï¼Œæˆ‘ä»¬æ‹’ç»çš„æ¦‚ç‡ï¼Œç­‰äºPå€¼<br>
â…¡å‹é”™è¯¯ï¼šåŸå‡è®¾ä¸ºå‡ï¼Œæ²¡æœ‰æ‹’ç»åŸå‡è®¾çš„æ¦‚ç‡</p>
</blockquote>
<p>å½“æˆ‘ä»¬åœ¨è¿›è¡Œç»Ÿè®¡æ£€éªŒæ—¶ï¼Œæˆ‘ä»¬å§‹ç»ˆéœ€è¦æ³¨æ„æˆ‘ä»¬å¾ˆå¯èƒ½ä¼šçŠ¯é”™è¯¯ã€‚è¿™å°±æ˜¯ä¸ºä»€ä¹ˆæˆ‘ä»¬çš„på€¼ä¸æ˜¯0ã€‚åœ¨åŸå‡è®¾æˆç«‹çš„å‰æä¸‹ï¼Œæ€»æ˜¯å­˜åœ¨ä¸€ä¸ªå¯èƒ½å¾ˆå°‘ï¼Œä½†æ˜¯å¯èƒ½å‘ç”Ÿçš„æ¦‚ç‡ï¼Œä½¿å¾—æˆ‘ä»¬æ‹’ç»äº†æ­£ç¡®çš„åŸå‡è®¾ã€‚å¦‚æœPå€¼è®¾ä¸º0.05ï¼Œé‚£ä¹ˆè¿™ç§æƒ…å†µå‘ç”Ÿçš„æ¦‚ç‡ä¹Ÿå°±æ˜¯0.05ã€‚è¿™ç§é”™è¯¯å«åšâ… å‹é”™è¯¯ï¼ˆtypeâ… errorï¼‰</p>
<p>â… å‹é”™è¯¯ä¼šå‘ç”Ÿ<strong>å‡é˜³æ€§</strong>ã€‚é‚£ä¹ˆæˆ‘ä»¬ä¸ºä»€ä¹ˆåªè®¾ä¸º0.05ï¼Ÿä¸è®¾å®šä¸€ä¸ªéå¸¸å°çš„å€¼å‘¢ï¼Ÿå› ä¸ºè¿˜å­˜åœ¨å¦å¤–ä¸€ç§é”™è¯¯ï¼Œâ…¡å‹é”™è¯¯ï¼Œå³åŸå‡è®¾ä¸ºå‡ï¼Œæ²¡æœ‰æ‹’ç»åŸå‡è®¾çš„æ¦‚ç‡ï¼Œè¿™ä¼šé€ æˆ<strong>å‡é˜´æ€§</strong>ã€‚åƒä¸Šé¢æˆ‘ä»¬æåˆ°çš„é‚£ç§æƒ…å†µï¼Œå°±æ˜¯çŠ¯äº†äºŒå‹é”™è¯¯ã€‚å¦‚æœæˆ‘ä»¬æŠŠPå€¼æé«˜åˆ°0.25ï¼Œé‚£ä¹ˆä¸Šé¢çš„ä¾‹å­å°±èƒ½å¤Ÿæ˜¾è‘—ã€‚ä½†æ˜¯ä¸€èˆ¬æƒ…å†µä¸‹æˆ‘ä»¬æ„¿ä¸æ„¿æ„æé«˜På€¼åˆ°0.25å‘¢ï¼Ÿé€šå¸¸æˆ‘ä»¬ä¸ä¼šè¿™ä¹ˆåšã€‚</p>
<h4 id="0-05æˆ–0-01-çš„-på€¼-æ˜¯éšæ„åˆ’åˆ†çš„ï¼ˆThe-0-05-and-0-01-Cut-offs-Are-Arbitraryï¼‰">0.05æˆ–0.01 çš„ på€¼ æ˜¯éšæ„åˆ’åˆ†çš„ï¼ˆThe 0.05 and 0.01 Cut-offs Are Arbitraryï¼‰</h4>
<p>å¾ˆå¤šæ‚å¿—å’Œç›‘ç®¡éƒ¨é—¨å¼ºè°ƒè¦è¾¾åˆ°0.05æˆ–0.01æ°´å¹³çš„æ˜¾è‘—ã€‚ä½†æ˜¯è¿™ä¸¤ä¸ªæ•°å­—çš„è®¾å®šå…¶å®æ²¡æœ‰ç‰¹åˆ«çš„ï¼Œè¿™ä»…ä»…æ˜¯å› ä¸ºæœ€å¼€å§‹çš„æ–‡ç« ç”¨è¿™ä¸¤ä¸ªæ•°å­—ä½œä¸ºä¾‹å­è€Œå·²ã€‚æœ¬ä¹¦çš„éƒ¨åˆ†ç›®çš„ä¹‹ä¸€å°±æ˜¯è®©è¯»è€…å¯¹äºPå€¼å’Œç½®ä¿¡åŒºé—´æœ‰ä¸€ä¸ªæ›´åŠ æ¸…æ™°çš„è®¤è¯†ï¼ˆå—æ•™äº†ï¼‰ã€‚</p>
<h4 id="Power-Calculation">Power Calculation</h4>
<p>Power is the probability of rejecting the null when the null is falseï¼ˆPower æ˜¯æŒ‡å½“åŸå‡è®¾ä¸ºå‡æ—¶æ‹’ç»åŸå‡è®¾çš„æ¦‚ç‡ï¼‰ã€‚</p>
<p>ä¸‹é¢æˆ‘ä»¬é€šè¿‡Rä¸­çš„å®é™…ä»£ç å»ç†è§£è¿™ä¸ªæ¦‚å¿µã€‚</p>
<p>æˆ‘ä»¬ä»ç„¶å‡è®¾æ ·æœ¬å¤§å°ä¸º12ï¼Œæ˜¾è‘—æ°´å¹³æ—¶0.05ã€‚ç”±äºæˆ‘ä»¬å·²ç»ä¸¤ä¸ªç¾¤ä½“å‡å€¼æœ‰å·®å¼‚ï¼Œæˆ‘ä»¬å¯ä»¥é‡å¤æŠ½æ ·ï¼Œçœ‹æ¯æ¬¡æŠ½æ ·tæ£€éªŒæ˜¯å¦å°äº0.05</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">N&lt;-12</span><br><span class="line">alpha&lt;-0.05</span><br><span class="line">B&lt;-2000</span><br><span class="line"></span><br><span class="line">reject&lt;-function(N, alpha&#x3D;0.05)&#123;</span><br><span class="line">	hf&lt;-sample(hfPopulation,N)</span><br><span class="line">	control&lt;-sample(controlPopulation,N)</span><br><span class="line">	pval&lt;-t.test(hf,control)$p.value</span><br><span class="line">	pval&lt;alpha</span><br><span class="line">&#125;</span><br><span class="line">#reject ç»™å‡ºä¸€æ¬¡æŠ½æ ·æ˜¯å¦å°äº0.05</span><br><span class="line"></span><br><span class="line">reject(12)</span><br><span class="line">## [1] FALSE</span><br><span class="line"></span><br><span class="line">rejections &lt;- replicate(B,reject(N))</span><br><span class="line">#é‡å¤2000æ¬¡ï¼Œreplicate(n,expr) </span><br><span class="line"></span><br><span class="line">mean(rejetions) # è¿™å°±æ˜¯N&#x3D;12æ—¶Powerçš„å€¼</span><br><span class="line">## [1] 0.2215</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>è¿™å°±æ˜¯ä¸ºä»€ä¹ˆå½“æˆ‘ä»¬çŸ¥é“åŸå‡è®¾æ˜¯é”™çš„æ—¶ï¼Œt-testå´æ²¡æœ‰æ‹’ç»åŸå‡è®¾ã€‚å½“æ ·æœ¬å¤§å°åªæœ‰12æ—¶ï¼Œpoweråªæœ‰23%ã€‚ä¸ºäº†é™ä½<strong>å‡é˜³æ€§</strong>å‘ç”Ÿçš„æ¦‚ç‡ï¼Œæˆ‘ä»¬å°†På€¼è®¾åœ¨0.05æ°´å¹³ã€‚å…¶å®è¿™ä¸ªæ°´å¹³æœ‰ç‚¹è¿‡é«˜ï¼Œå¯¼è‡´å‘ç”Ÿ<strong>â…¡å‹é”™è¯¯</strong>çš„æ¦‚ç‡å¤§å¤§æé«˜ã€‚</p>
<blockquote>
<p>power å’Œ å‘ç”Ÿâ…¡å‹çš„æ¦‚ç‡ ç›¸åŠ ä¸º1ã€‚powerä½ï¼Œè¯´æ˜å‘ç”Ÿâ…¡å‹é”™è¯¯çš„æ¦‚ç‡é«˜ã€‚</p>
</blockquote>
<p>æ¥ä¸‹æ¥ï¼Œè®©æˆ‘ä»¬æ¥çœ‹çœ‹powerå’ŒæŠ½æ ·å¤§å°Nçš„å…³ç³»ã€‚è¿™é‡Œæˆ‘ä»¬ç”¨åˆ°äº†<strong>sapply(Ns,expr)</strong>ï¼Œè¿™ä¸ªå‡½æ•°ä¼šå¯¹å‘é‡ä¸­çš„æ¯ä¸€ä¸ªæ•°æ‰§è¡ŒåŒæ ·çš„functionã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">Ns&lt;-seq(<span class="number">5</span>,<span class="number">50</span>,<span class="number">5</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment"># So we use apply() like this</span></span><br><span class="line">power&lt;-sapply(Ns,<span class="keyword">function</span>(N)&#123;</span><br><span class="line">	rejections&lt;-replicate(B, reject(N))</span><br><span class="line">	mean(rejections)</span><br><span class="line">	&#125;)</span><br><span class="line">	</span><br><span class="line">plot(Ns,power,type = <span class="string">&#x27;b&#x27;</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>æˆ‘ä»¬å¯ä»¥çœ‹å‡ºï¼Œéšç€Nçš„å¢åŠ ï¼Œpowerä¸æ–­æé«˜ã€‚</p>
<p>æˆ‘ä»¬å›ºå®šNï¼Œçœ‹ä¸€çœ‹alphaçš„è®¾å®šå¯¹power çš„å½±å“</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">N&lt;-<span class="number">30</span></span><br><span class="line">alphas&lt;-<span class="built_in">c</span>(<span class="number">0.1</span>,<span class="number">0.05</span>,<span class="number">0.01</span>,<span class="number">0.001</span>,<span class="number">0.0001</span>)</span><br><span class="line">power&lt;-sapply(alphas,<span class="keyword">function</span>(alpha)&#123;</span><br><span class="line">	rejections &lt;- replicate(B,reject(N,alpha=alpha))</span><br><span class="line">	mean(rejections)</span><br><span class="line">&#125;)</span><br><span class="line">plot(alphas, power, xlab=<span class="string">&quot;alpha&quot;</span>, type=<span class="string">&quot;b&quot;</span>, <span class="built_in">log</span>=<span class="string">&quot;x&quot;</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>æˆ‘ä»¬å¯ä»¥çœ‹å‡ºéšç€alphaçš„å¢åŠ ï¼Œpowerä¹Ÿåœ¨ä¸æ–­å¢åŠ ã€‚</p>
<blockquote>
<p>There is no â€œrightâ€ power or â€œrightâ€ alpha level, but it is important that you understand what eachmeans.</p>
</blockquote>
<h5 id="p-values-are-Arbitrary-under-the-Alternative-Hypothesis">p-values are Arbitrary under the Alternative Hypothesis</h5>
<blockquote>
<p>Another consequence of what we have learned about power is that p-values are somewhat arbitrarywhen the null hypothesis is not true and therefore thealternativehypothesis is true (the differencebetween the population means is not zero). When the alternative hypothesis is true, we can make a p-value as small as we want simply by increasing the sample size (supposing that we have an infinitepopulation to sample from). We can show this property of p-values by drawing larger and largersamples from our population and calculating p-values. This works because, in our case, we knowthat the alternative hypothesis is true, since we have access to the populations and can calculate thedifference in their means.</p>
</blockquote>
<p>éšç€æ ·æœ¬å¤§å°çš„å¢åŠ ï¼ŒPå€¼ä¼šå‡å°ï¼ˆå‡è®¾ç¾¤ä½“æ— é™å¤§ï¼‰ã€‚è¯æ˜å¦‚ä¸‹</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">calculatePvalue&lt;-<span class="keyword">function</span>(N) &#123;</span><br><span class="line">	hf&lt;-sample(hfPopulation,N)</span><br><span class="line">	control&lt;-sample(controlPopulation,N)</span><br><span class="line">	t.test(hf,control)$p.value</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">Ns&lt;-seq(<span class="number">10</span>,<span class="number">200</span>,by=<span class="number">10</span>)</span><br><span class="line">Ns_rep&lt;-<span class="built_in">rep</span>(Ns, each=<span class="number">10</span>) <span class="comment">#each çš„å½¢å¼æ˜¯ 10 10 10â€¦â€¦ 20 â€¦â€¦</span></span><br><span class="line"></span><br><span class="line">pvalues&lt;-sapply(Ns_rep, calculatePvalue)</span><br><span class="line"></span><br><span class="line">plot(Ns_rep, pvalues, <span class="built_in">log</span>=<span class="string">&quot;y&quot;</span>, xlab=<span class="string">&quot;sample size&quot;</span>,</span><br><span class="line">ylab=<span class="string">&quot;p-values&quot;</span>)</span><br><span class="line">abline(h=<span class="built_in">c</span>(<span class="number">.01</span>,<span class="number">.05</span>), col=<span class="string">&quot;red&quot;</span>, lwd=<span class="number">2</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>æ ·æœ¬é‡ä»10å¢åŠ åˆ°200çš„è¿‡ç¨‹ä¸­ï¼Œpå€¼æœ€ä½ä¸‹é™åˆ°äº†10^-8^ã€‚ä½†æ˜¯på€¼å¤ªä½ä¹Ÿä¸è§å¾—æ˜¯ä¸€ä»¶å¾ˆå¥½çš„äº‹ã€‚ä¸€æ—¦æˆ‘ä»¬ç¡®ä¿¡äº†ä¸€ä¸ªreasonable alphaï¼ˆæ¯”å¦‚0.05ï¼‰ï¼Œé‚£ä¹ˆå¦‚æœæˆ‘ä»¬è·å¾—äº†ä¸€ä¸ªæ¯”è¿™ä¸ªé˜ˆå€¼ä½å¾—å¤šçš„på€¼ï¼Œé‚£å°±è¯´æ˜<strong>æˆ‘ä»¬ç”¨äº†è¿‡é‡çš„å°é¼ </strong>ã€‚é€‚å½“åœ°æé«˜æ ·æœ¬é‡ï¼Œç¡®å®æœ‰åŠ©äºä¼°è®¡çš„å‡†ç¡®æ€§ã€‚ä½†æ˜¯ï¼Œå®é™…ä¸Šï¼Œéšç€æ ·æœ¬é‡çš„å¢åŠ på€¼çš„é™ä½æ˜¯ä¸€ä»¶è‡ªç„¶çš„äº‹ã€‚</p>
<blockquote>
<p>The p-values get smaller and smaller with increasing sample size because the numerator ofthe t-statistic haspN(for equal sized groups, and a similar effect occurs whenMÌ¸=N). Therefore,ifâˆ†is non-zero, the t-statistic will increase with N.</p>
</blockquote>
<p>äº‹å®ä¸Šï¼Œéšç€æ ·æœ¬é‡çš„å¢åŠ ï¼Œt-testå¾—åˆ°çš„på€¼ä¹Ÿä¼šä¸æ–­é™ä½ã€‚è¿™æ˜¯tç»Ÿè®¡é‡ä¸­æœ‰ä¸€ä¸ªæ ¹å·Nã€‚</p>
<p>ç®—äº†ä¸ªCohenâ€™s dï¼Œä½†æ²¡çœ‹æ‡‚è¿™ä¸ªç»Ÿè®¡é‡åˆ°åº•æœ‰å•¥ç”¨ï¼Ÿ</p>
<blockquote>
<p>This tells us how many standard deviations of the data the mean of the high-fat diet group is from the control group. Under the alternative hypothesis, unlike the t-statistic which is guaranteed to increase, the effect size and Cohenâ€™s d will become more precise.ï¼ˆçœ‹ä¸æ‡‚ï¼‰</p>
</blockquote>
<h3 id="Monte-Carlo-Simulation">Monte Carlo Simulation</h3>
<p>ç”µè„‘å¯ä»¥ç”¨æ¥äº§ç”Ÿä¼ªéšæœºæ•°ï¼ˆpseudo-random numbersï¼‰ï¼Œä¸€ä¸ªæœ€å¤§çš„å¥½å¤„æ˜¯æˆ‘ä»¬å¯ä»¥é€šè¿‡æ¨¡æ‹Ÿæ•°æ®æ¥è¯æ˜æˆ‘ä»¬çš„å‡è®¾æˆ–ç†è®ºã€‚</p>
<blockquote>
<p>ä¼ªéšæœºæ€§ï¼ˆè‹±è¯­ï¼šPseudorandomnessï¼‰æ˜¯ä¸€ä¸ªè¿‡ç¨‹ä¼¼ä¹æ˜¯éšæœºçš„ï¼Œä½†å®é™…ä¸Šå¹¶ä¸æ˜¯ã€‚ä¾‹å¦‚ä¼ªéšæœºæ•°æ˜¯ä½¿ç”¨ä¸€ä¸ªç¡®å®šæ€§çš„ç®—æ³•è®¡ç®—å‡ºæ¥çš„ä¼¼ä¹æ˜¯éšæœºçš„æ•°åºï¼Œå› æ­¤ä¼ªéšæœºæ•°å®é™…ä¸Šå¹¶ä¸éšæœºã€‚åœ¨è®¡ç®—ä¼ªéšæœºæ•°æ—¶å‡å¦‚ä½¿ç”¨çš„å¼€å§‹å€¼ä¸å˜çš„è¯ï¼Œé‚£ä¹ˆä¼ªéšæœºæ•°çš„æ•°åºä¹Ÿä¸å˜ã€‚</p>
</blockquote>
<p>Simulations can also be used to check theoretical or analytical results. Also, many of the theoretical results we use in statistics are based on <strong>asymptotics</strong>: they hold when the sample size goes to <a target="_blank" rel="noopener" href="http://infinity.In">infinity.In</a> practice, we never have an infinite number of samples so we may want to know how well thetheory works with our actual sample size.</p>
<p>ä½œä¸ºä¾‹å­ï¼Œæˆ‘ä»¬è’™ç‰¹å¡æ´›æ¨¡æ‹Ÿï¼ˆ Monte Carlo simulationï¼‰æ¥æ¯”è¾ƒä¸åŒæ ·æœ¬å¤§å°çš„æƒ…å†µä¸‹CLTå’Œtåˆ†å¸ƒçš„è¿‘ä¼¼å€¼ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">library(dplyr)</span><br><span class="line">dat&lt;-read.csv(<span class="string">&quot;mice_pheno.csv&quot;</span>)<span class="comment">#Previously downloaded</span></span><br><span class="line">controlPopulation&lt;-filter(dat,Sex==<span class="string">&quot;F&quot;</span>&amp;Diet==<span class="string">&quot;chow&quot;</span>)%&gt;%select(Bodyweight)%&gt;%unlist</span><br><span class="line"></span><br><span class="line">ttestgenerator&lt;-<span class="keyword">function</span>(n) &#123;</span><br><span class="line">  <span class="comment">#note that here we have a false &quot;high fat&quot; group where we actually</span></span><br><span class="line">  <span class="comment">#sample from the nonsmokers. this is because we are modeling the *null*</span></span><br><span class="line">  cases&lt;-sample(controlPopulation,n)</span><br><span class="line">  controls&lt;-sample(controlPopulation,n)</span><br><span class="line">  tstat&lt;-(mean(cases)-mean(controls))/<span class="built_in">sqrt</span>( var(cases)/n+var(controls)/n ) <span class="comment">#è¿™ä¸ªtç»Ÿè®¡é‡æ˜¯å‡è®¾ä¸¤ä¸ªç¾¤ä½“æ–¹å·®ä¸ç­‰</span></span><br><span class="line">  <span class="built_in">return</span>(tstat)</span><br><span class="line">  &#125;</span><br><span class="line">ttests&lt;-replicate(<span class="number">1000</span>, ttestgenerator(<span class="number">10</span>))</span><br><span class="line">hist(ttests)</span><br></pre></td></tr></table></figure>
<p>ä»ç›´æ–¹å›¾ä¸­ï¼Œæˆ‘ä»¬è¿‘ä¼¼çœ‹å‡ºç¬¦åˆæ­£æ€åˆ†å¸ƒã€‚ä½†æ˜¯ä»ä¸Šæ–‡æåˆ°ï¼Œæˆ‘ä»¬å†çœ‹ä¸€ä¸‹QQå›¾</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">qqnorm(ttests)</span><br><span class="line">qqline(ttests)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä»QQå›¾ä¸­ï¼Œå¯ä»¥çœ‹å‡ºè¿‘ä¼¼ç¬¦åˆæ­£æ€åˆ†å¸ƒã€‚è¿™è¯´æ˜åœ¨æ ·æœ¬æ•°æ˜¯10çš„æƒ…å†µä¸‹ï¼Œå·²ç»ç¬¦åˆæ­£æ€åˆ†å¸ƒï¼Œå¯ä»¥ä½¿ç”¨ä¸­å¿ƒæé™å®šç†äº†ã€‚é‚£ä¹ˆå¦‚æœæ ·æœ¬æ•°ä¸º3å‘¢ï¼Ÿ</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">ttests&lt;-replicate(<span class="number">1000</span>, ttestgenerator(<span class="number">3</span>))</span><br><span class="line">qqnorm(ttests)</span><br><span class="line">qqline(ttests)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<blockquote>
<p>Now we see that the large quantiles, referred to by statisticians as the tails, are larger than expected(below the line on the left side of the plot and above the line on the right side of the plot).</p>
</blockquote>
<p>æˆ‘ä»¬å¯ä»¥çœ‹å‡ºQQå›¾ä¸­å·¦ä¾§ä½äºç›´çº¿ï¼Œå³ä¾§é«˜äºç›´çº¿ã€‚è¿™è¯´æ˜ä¸¤ä¾§çš„å°¾å·´æ¯”æ­£æ€åˆ†å¸ƒè¦è‚¥ã€‚</p>
<p>åœ¨ä¹‹å‰çš„ç« èŠ‚ä¸­ï¼Œæˆ‘ä»¬æåˆ°ï¼Œå¦‚æœ<strong>æ ·æœ¬æ•°ä¸å¤§</strong> å¹¶ä¸” <strong>æ€»ä½“ç¬¦åˆæ­£æ€åˆ†å¸ƒ</strong> ï¼Œé‚£ä¹ˆä½¿ç”¨tåˆ†å¸ƒæ˜¯ä¸€ç§æ›´å¥½çš„è¿‘ä¼¼ã€‚è¯æ˜å¦‚ä¸‹</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">ps&lt;-(seq(<span class="number">0</span>,<span class="number">999</span>)+<span class="number">0.5</span>)/<span class="number">1000</span> <span class="comment">#0åˆ°1çš„å‡åŒ€åˆ†å¸ƒ</span></span><br><span class="line">qqplot(qt(ps,df=<span class="number">2</span>*<span class="number">3</span>-<span class="number">2</span>),ttests,xlim=<span class="built_in">c</span>(-<span class="number">6</span>,<span class="number">6</span>),ylim=<span class="built_in">c</span>(-<span class="number">6</span>,<span class="number">6</span>))</span><br><span class="line"><span class="comment"># qtè·å¾—æŸtåˆ†å¸ƒç´¯ç§¯æ¦‚ç‡çš„æ•°å€¼</span></span><br><span class="line"><span class="comment"># qqplotæ¯”è¾ƒä¸¤ä¸ªå˜é‡åˆ†å¸ƒæ˜¯å¦ç›¸åŒ</span></span><br><span class="line">abline(<span class="number">0</span>,<span class="number">1</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä»ä¸­æˆ‘ä»¬å¯ä»¥çœ‹å‡ºï¼Œtåˆ†å¸ƒåœ¨è¿™ç§æƒ…å†µä¸‹æ›´ä¸ºè¿‘ä¼¼ï¼Œä½†æ˜¯ä»ä¸å®Œç¾ã€‚è¿™æ˜¯å› ä¸º<strong>æ€»ä½“å¹¶ä¸æ˜¯å¾ˆå¥½åœ°ç¬¦åˆæ­£æ€åˆ†å¸ƒ</strong>ã€‚(ä»ä¸‹é¢æ€»ä½“çš„QQå›¾ä¸­çœ‹å‡ºï¼Œæ€»ä½“ç•¥å)</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">qqnorm(controlPopulation)</span><br><span class="line">qqline(controlPopulation)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="Parametric-Simulations-for-the-Observations">Parametric Simulations for the Observations</h3>
<p>æˆ‘ä»¬å†è¿™é‡Œä½¿ç”¨è’™ç‰¹å¡æ´›æ¨¡æ‹Ÿéšæœºå˜é‡å’Œå‡è®¾æ£€éªŒï¼Œè¿™æ˜¯å› ä¸ºæˆ‘ä»¬æœ‰æ€»ä½“æ•°æ®ã€‚è¿™æ˜¯ä¸ºäº†æ•™å­¦çš„ä½œç”¨ï¼Œä½†åœ¨å®é™…è¿ç”¨ä¸­æ˜¯ä¸å¯èƒ½çš„ã€‚ä¸€èˆ¬é‡‡ç”¨è’™ç‰¹å¡æ´›æ¨¡æ‹Ÿï¼Œä¸»è¦ç”¨äºå‚æ•°ä¼°è®¡ï¼ˆå¦‚å‡å€¼å’Œæ ‡å‡†å·®ï¼‰ã€‚</p>
<p>ä»¥å°é¼ ä½“é‡çš„ä¾‹å­æ¥è¯´ï¼Œæˆ‘ä»¬é€šè¿‡å…ˆéªŒçŸ¥è¯†ï¼Œè®¤è¯†åˆ°å°é¼ çš„ä½“é‡å‡å€¼ä¸º24ç›å¸ï¼Œæ ‡å‡†å·®ä¸º3.5ç›å¸ï¼Œè€Œä¸”å°é¼ ä½“é‡è¿‘ä¼¼ç¬¦åˆæ­£æ€åˆ†å¸ƒã€‚é€šè¿‡ä»¥ä¸Šä¿¡æ¯ï¼Œæˆ‘ä»¬å¯ä»¥æ„å»ºä¸€ä¸ªæ€»ä½“æ•°æ®ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">controls&lt;-rnorm(<span class="number">5000</span>, mean=<span class="number">24</span>, sd=<span class="number">3.5</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ç”±äºæˆ‘ä»¬å¯ä»¥é‡å¤ä¸Šè¿°æ“ä½œã€‚å› æ­¤æˆ‘ä»¬ä¸éœ€è¦ç”¨åˆ°sample()å‡½æ•°äº†ã€‚ä»£ç å¦‚ä¸‹</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">ttestgenerator&lt;-function(n, mean&#x3D;24, sd&#x3D;3.5) &#123;</span><br><span class="line">  cases&lt;-rnorm(n,mean,sd)</span><br><span class="line">  controls&lt;-rnorm(n,mean,sd)</span><br><span class="line">  tstat&lt;-(mean(cases)-mean(controls))&#x2F;sqrt( var(cases)&#x2F;n+var(controls)&#x2F;n )</span><br><span class="line">  return(tstat)</span><br><span class="line">&#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="Permutation-Tests">Permutation Tests</h3>
<p>å‡è®¾æˆ‘ä»¬é‡åˆ°è¿™æ ·ä¸€ç§æƒ…å†µï¼Œæ²¡æœ‰ä¸€ç§æ ‡å‡†çš„ç»Ÿè®¡é‡èƒ½è¿‘ä¼¼æ¨¡æ‹Ÿæˆ‘ä»¬çš„æ•°æ®ã€‚</p>
<p>We have computed a summary statistic, such as the difference in mean, but do not have auseful approximation, such as that provided by the CLT.</p>
<p>è¿™ä¸ªæ—¶å€™ Permutation Tests å°±æ´¾ä¸Šç”¨åœºäº†ã€‚</p>
<p>è®©æˆ‘ä»¬å›åˆ°æœ€å¼€å§‹æ‹¿åˆ°çš„æ ·æœ¬</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">dat=read.csv(<span class="string">&quot;femaleMiceWeights.csv&quot;</span>)</span><br><span class="line">library(dplyr)</span><br><span class="line">control&lt;-filter(dat,Diet==<span class="string">&quot;chow&quot;</span>)%&gt;%select(Bodyweight)%&gt;%unlist</span><br><span class="line">treatment&lt;-filter(dat,Diet==<span class="string">&quot;hf&quot;</span>)%&gt;%select(Bodyweight)%&gt;%unlist</span><br><span class="line">obsdiff&lt;-mean(treatment)-mean(control)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>In previous sections, we showed parametric approaches that helped determine if the observeddifference was significant. Permutation tests take advantage of the fact that if we randomly shufflethe cases and control labels, then the null is true. So we shuffle the cases and control labels andassume that the ensuing distribution approximates the null distribution. Here is how we generate anull distribution by shuffling the data 1,000 times:</p>
</blockquote>
<p>è¿™æ®µè¯çš„å¤§æ„æ˜¯ï¼Œä¹‹å‰æˆ‘ä»¬æ˜¯ç”¨å¸¦å‚æ•°çš„æ–¹æ³•ï¼ˆå‡å€¼ï¼Œæ ‡å‡†å·®ï¼‰æ£€éªŒå·®å¼‚æ˜¯å¦æ˜¾è‘—ã€‚è€Œè¿™ä¸ªPermutation tests è®¤ä¸ºï¼Œå¦‚æœæˆ‘ä»¬éšæœºæ´—ç‰Œ cases å’Œ control çš„æ ‡ç­¾ï¼Œé‚£ä¹ˆé›¶å‡è®¾åº”è¯¥æ˜¯æ­£ç¡®çš„ã€‚<br>
ä¸‹é¢æˆ‘ä»¬æ´—ç‰Œäº†1000æ¬¡ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">N&lt;-<span class="number">12</span></span><br><span class="line">avgdiff&lt;-replicate(<span class="number">1000</span>, &#123;</span><br><span class="line">  <span class="built_in">all</span>&lt;-sample(<span class="built_in">c</span>(control,treatment))</span><br><span class="line">  newcontrols&lt;-<span class="built_in">all</span>[<span class="number">1</span>:N]</span><br><span class="line">  newtreatments&lt;-<span class="built_in">all</span>[(N+<span class="number">1</span>):(<span class="number">2</span>*N)]</span><br><span class="line">  <span class="built_in">return</span>(mean(newtreatments)-mean(newcontrols))</span><br><span class="line">  &#125;)</span><br><span class="line"><span class="comment">#è¿™ä¸ª&#123;&#125;å¥½åƒæ˜¯ä¸å¸¦åç§°çš„å‡½æ•°ï¼Œç±»ä¼¼äºPythonä¸­çš„lambda</span></span><br><span class="line"><span class="comment">#sample(x)ä¸åŠ n,å°±æ˜¯é‡æ’ï¼ˆpermutationï¼‰çš„æ„æ€</span></span><br><span class="line">hist(avgdiff)</span><br><span class="line">abline(v=obsdiff, col=<span class="string">&quot;red&quot;</span>, lwd=<span class="number">2</span>)</span><br></pre></td></tr></table></figure>
<blockquote>
<p>How many of the null means are bigger than the observed value? That proportion would be thep-value for the null. We add a 1 to the numerator and denominator to account for misestimation ofthe p-value</p>
</blockquote>
<p>åœ¨è¿™1000æ¬¡ä¸­ï¼Œæœ‰å¤šå°‘æ¯”ä¾‹å¤§äºè¿™ä¸ªobsdiffï¼Œè¿™å°±æ˜¯På€¼ã€‚ä½†æ˜¯ï¼Œåˆ†å­åˆ†æ¯è¦éƒ½åŠ ä¸Šä¸€ä¸ª1ã€‚ï¼ˆä¸ªäººç†è§£å®é™…ä¸Šæ˜¯åšäº†1001æ¬¡ï¼Œæ²¡æœ‰ç®—ä¸ŠçœŸå®æƒ…å†µï¼‰</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">(<span class="built_in">sum</span>(<span class="built_in">abs</span>(avgdiff)&gt;<span class="built_in">abs</span>(obsdiff))+<span class="number">1</span>)/(<span class="built_in">length</span>(avgdiff)+<span class="number">1</span>)</span><br><span class="line"><span class="comment">## [1] 0.05594406</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ç°åœ¨æˆ‘ä»¬æŠŠè¿™ä¸ªç”¨åœ¨æ ·æœ¬é‡æ›´å°çš„æ•°æ®é›†ï¼Œæ¯”å¦‚æ ·æœ¬é‡ä¸º5</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">N&lt;-5</span><br><span class="line">control&lt;-sample(control,N)</span><br><span class="line">treatment&lt;-sample(treatment,N)</span><br><span class="line">obsdiff&lt;-mean(treatment)-mean(control)</span><br><span class="line"></span><br><span class="line">avgdiff&lt;-replicate(1000, &#123;</span><br><span class="line">  all&lt;-sample(c(control,treatment))</span><br><span class="line">  newcontrols&lt;-all[1:N]</span><br><span class="line">  newtreatments&lt;-all[(N+1):(2*N)]</span><br><span class="line">  return(mean(newtreatments)-mean(newcontrols))</span><br><span class="line">&#125;)</span><br><span class="line">hist(avgdiff)</span><br><span class="line">abline(v&#x3D;obsdiff, col&#x3D;&quot;red&quot;, lwd&#x3D;2)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ç°åœ¨å¯ä»¥æ˜æ˜¾çœ‹å‡ºå·®è·ä¸æ˜¾è‘—äº†ã€‚</p>
<blockquote>
<p>Now the observed difference is not significant using this approach. Keep in mind that there is notheoretical guarantee that the null distribution estimated from permutations approximates the actualnull distribution. For example, if there is a real difference between the populations, some of the permutations will be unbalanced and will contain some samples that explain this difference. This implies that the null distribution created with permutations will have larger tails than the actualnull distribution. This is why permutations result in conservative p-values. For this reason, whenwe have few samples, we canâ€™t do permutations.</p>
</blockquote>
<p>æ²¡çœ‹æ‡‚ï¼Œç»“è®ºæ˜¯æ•°æ®é‡å°æ—¶ï¼Œä¸ç”¨è¿™ç§permutation test</p>
<p>è€Œä¸”permutation test ä¹Ÿæœ‰å‡è®¾ï¼š<strong>:samples are assumed to be independentand â€œexchangeableâ€</strong> ã€‚ If there is hidden structure in your data, then permutation tests can result in estimated null distributions that underestimate the size of tails because the permutations may destroy the existing structure in the original data.</p>
<h3 id="Association-Tests">Association Tests</h3>
<p>å‰é¢çš„ç»Ÿè®¡åˆ†æä¸­ï¼Œå¿½ç•¥äº†ä¸€ç±»é‡è¦çš„æ•°æ®ï¼ŒäºŒå…ƒæˆ–è€…è¯´åˆ†ç±»æ•°æ®ã€‚ä¸¾ä¸ªä¾‹å­ï¼Œå½“ä½ æœ‰ä¸¤ç§ç–¾ç—…çš„åŸºå› å‹çš„ç¾¤ä½“(AA/Aa ä¸ aa)ä½œä¸ºcases å’Œ controlã€‚æˆ‘ä»¬è¦æå‡ºçš„é—®é¢˜æ˜¯ï¼Œè¿™ä¸ªåŸºå› å‹ä¸ç–¾ç—…æ˜¯å¦æœ‰å…³ç³»ã€‚<br>
è¿™é‡Œç–¾ç—…çŠ¶å†µè¡¨ç¤ºä¸º0å’Œ1ï¼Œé‚£ä¹ˆæˆ‘ä»¬ä¸ºä»€ä¹ˆä¸èƒ½ç”¨t-testï¼Ÿå› ä¸ºæ•°æ®è¦ä¹ˆæ˜¯0è¦ä¹ˆæ˜¯1ï¼Œå¾ˆæ˜æ˜¾ä¸ä¼šç¬¦åˆæ­£æ€åˆ†å¸ƒï¼Œå› æ­¤ä¸èƒ½ä½¿ç”¨t-testã€‚</p>
<p>å¦‚æœæ ·æœ¬æ•°é‡å¾ˆå¤§çš„è¯ï¼Œæˆ‘ä»¬å¯ä»¥ä½¿ç”¨CLTã€‚ä¸ç„¶çš„è¯ï¼Œæˆ‘ä»¬å°±å¯ä»¥ä½¿ç”¨Association Testsã€‚</p>
<h4 id="äºŒé¡¹åˆ†å¸ƒä¸è¶…å‡ ä½•åˆ†å¸ƒ">äºŒé¡¹åˆ†å¸ƒä¸è¶…å‡ ä½•åˆ†å¸ƒ</h4>
<p>è§£é‡Š: æˆ‘çš„è¿™ç§ç®—æ³•æ˜¯åŸºäºäºŒé¡¹åˆ†å¸ƒï¼Œè€ŒFisherâ€™s exact testæ˜¯åŸºäºè¶…å‡ ä½•åˆ†å¸ƒï¼ˆhypergeometric distributionï¼‰ã€‚<br>
äºŒé¡¹åˆ†å¸ƒ</p>
<ol>
<li>è¯•éªŒæ¬¡æ•°å›ºå®š</li>
<li>æ¯ä¸€æ¬¡äº‹ä»¶éƒ½æœ‰ä¸¤ä¸ªå¯èƒ½çš„ç»“æœ</li>
<li><strong>æ¯æ¬¡è¯•éªŒä¸­çš„äº‹ä»¶ç›¸äº’ç‹¬ç«‹</strong></li>
<li><strong>æ¯ä¸€æ¬¡æˆåŠŸçš„æ¦‚ç‡ç›¸ç­‰</strong></li>
</ol>
<p>è¶…å‡ ä½•åˆ†å¸ƒ å®ƒæè¿°äº†ç”±<strong>æœ‰é™ä¸ªç‰©ä»¶ä¸­</strong>æŠ½å‡ºnä¸ªç‰©ä»¶ï¼ŒæˆåŠŸæŠ½å‡ºæŒ‡å®šç§ç±»çš„ç‰©ä»¶çš„ä¸ªæ•°ï¼ˆä¸å½’è¿˜ ï¼ˆwithout replacementï¼‰ï¼‰ã€‚</p>
<p>ä¾‹å¦‚åœ¨æœ‰Nä¸ªæ ·æœ¬ï¼Œå…¶ä¸­Kä¸ªæ˜¯ä¸åŠæ ¼çš„ã€‚è¶…å‡ ä½•åˆ†å¸ƒæè¿°äº†åœ¨è¯¥Nä¸ªæ ·æœ¬ä¸­æŠ½å‡ºnä¸ªï¼Œå…¶ä¸­kä¸ªæ˜¯ä¸åŠæ ¼çš„æ¦‚ç‡ã€‚</p>
<p>ç‰¹ç‚¹ï¼š</p>
<ol>
<li>æ¯æ¬¡æŠ½å–ä¹Ÿåªæœ‰ä¸¤ä¸ªäº’æ–¥çš„ç»“æœ</li>
<li><strong>æ¯æ¬¡æŠ½å–åæˆåŠŸçš„æ¦‚ç‡å‘ç”Ÿæ”¹å˜</strong>ï¼ˆå› ä¸ºæ˜¯ä¸æ”¾å›æŠ½æ ·ï¼Œè¿™ä¹Ÿå¯¼è‡´äº†æ¯æ¬¡æŠ½å–äº‹ä»¶ä¸ç‹¬ç«‹ï¼‰</li>
</ol>
<h4 id="Lady-Tasting-Tea">Lady Tasting Tea</h4>
<p>R.A.Fisher çš„ä¸€ä½ç†Ÿäººå®£ç§°å¥¹èƒ½åˆ†è¾¨ç‰›å¥¶æ˜¯æ—©äºæˆ–æ™šäºèŒ¶æ”¾å…¥æ¯ä¸­ï¼ˆå°±æ˜¯ç‰›å¥¶å’ŒèŒ¶æ··åˆçš„é¡ºåºï¼‰ã€‚Fisher å°±å’Œå¥¹åšäº†ä¸€ä¸ªå°å®éªŒï¼Œ4æ¬¡ä¸­å¯¹äº†3æ¬¡ã€‚é‚£ä¹ˆï¼Œæˆ‘ä»¬èƒ½å¦è¯´æ˜è¿™ä½å¥³å£«å­˜åœ¨è¿™ç§ç‰¹æ®Šçš„èƒ½åŠ›å‘¢ï¼Ÿè¿™ä¸ªä¾‹å­è¢«ç§°ä¸º&quot;Lady tasting tea&quot;ã€‚<br>
æŒ‰ç…§ä¹‹å‰çš„ç»éªŒã€‚ä¸€ä¸ªåŸºæœ¬çš„é—®é¢˜æ˜¯ï¼Œå¦‚æœè¿™ä½å¥³å£«ä»…ä»…é çŒœï¼Œå¥¹èƒ½å¾—åˆ°3/4æˆ–æ›´é«˜çš„æ­£ç¡®ç‡(å³3/4+4/4)çš„æ¦‚ç‡æ˜¯å¤šå°‘ï¼Ÿ<br>
ä»å’Œä¹‹å‰ä¸€æ ·ï¼Œæˆ‘ä»¬éœ€è¦è®¡ç®—åœ¨é›¶å‡è®¾æˆç«‹çš„å‰æä¸‹äº‹æƒ…å‘ç”Ÿçš„æ¦‚ç‡ã€‚è¿™é‡Œçš„é›¶å‡è®¾å°±æ˜¯è¿™ä½å¥³å£«éƒ½æ˜¯çŒœçš„ã€‚</p>
<blockquote>
<p>he basic question we ask is: if the tester is actually guessing, what are the chances that she gets 3 ormore correct? Just as we have done before, we can compute a probability under the null hypothesisthat she is guessing four of each. If we assume this null hypothesis, we can think of this particularexamples as picking 4 balls out of an urn with 4 green (correct answer) and 4 red (incorrect answer)balls.</p>
</blockquote>
<blockquote>
<p>Under the null hypothesis that she is simply guessing, each ball has the same chance of beingpicked. We can then use combinatorics to figure out each probability. The probability of picking3 is(43)(41)/(84)= 16/70. The probability of picking all 4 correct is(44)(40)/(84)= 1/70. Thus, thechance of observing a 3 or something more extreme, under the null hypothesis, is 0.24. This isthe p-value. The procedure that produced this p-value is calledFisherâ€™s exact testand it uses the hypergeometric distribution.</p>
</blockquote>
<p>å®ƒè¿™ç§ç®—æ³•æ˜¯8ä¸ªçƒé‡Œæ‘¸4ä¸ªã€‚æˆ‘çš„ç®—æ³•æ˜¯æ¯æ¬¡çŒœå¯¹çš„æ¦‚ç‡æ˜¯0.5ï¼Œé‚£ä¹ˆ4æ¬¡çŒœå¯¹3æ¬¡çš„æ¦‚ç‡æ˜¯4*0.5^3^*0.5=0.25ã€‚è¿™ä¸¤ç§ç»“æœä¸ä¸€æ ·ï¼Œæˆ‘ä¹Ÿæä¸å¤ªæ¸…æ¥šã€‚</p>
<p>è¿™é‡Œç®—é”™äº†ï¼Œé™¤äº†ä¸æ¸…æ™°äºŒé¡¹åˆ†å¸ƒä¸è¶…å‡ ä½•åˆ†å¸ƒï¼Œè¿˜æœ‰ä¸€ä¸ªåŸå› ï¼Œä¸æ¸…æ¥šè¿™ä¸ªLady tasting teaè¯•éªŒã€‚åœ¨è¿™ä¸ªè¯•éªŒä¸­ï¼ŒFisherå‡†å¤‡äº†8æ¯èŒ¶ï¼Œæ¯ä¸€ç§ï¼ˆtea first or milk firstï¼‰éƒ½å‡†å¤‡äº†4æ¯ï¼Œå¹¶ä¸”éšæœºæ‹œæ”¾ä½ç½®ã€‚è¯•éªŒè€…éœ€æŒ‘å‡ºæŒ‰å…¶ä¸­ä¸€ç§æ–¹æ³•æ³¡åˆ¶çš„4æ¯ã€‚</p>
<p>æŒ‰Fisherâ€™s exact test åªæœ‰å½“è¿™ä½å¥³å£«æŠŠ4æ¯éƒ½æŒ‘å‡ºæ¥ï¼ˆp=1/70ï¼‰ï¼Œä»–æ‰ä¼šæ‰¿è®¤å¥¹æœ‰è¿™ç§èƒ½åŠ›ï¼ˆä½†æ˜¯ä¸èƒ½é‡åŒ–ï¼‰ã€‚å“ªæ€•å¦‚é¢˜æ‰€ç¤ºï¼Œå¥¹æŒ‘å‡ºæ¥äº†3æ¯ï¼ˆp=16/70+1/70=17/70ï¼‰ï¼ŒPå€¼ä¹Ÿæ˜¯ä¸æ˜¾è‘—ã€‚</p>
<blockquote>
<p>å½“æ ·æœ¬é‡æ›´å¤§æ—¶ï¼Œä¹Ÿå¯ä»¥ä½¿ç”¨å¡æ–¹åˆ†å¸ƒã€‚ä½†æ˜¯ï¼Œå½“æ ·æœ¬æ•°ç›®å¾ˆå°‘æˆ–æ•°æ®åˆ†å¸ƒæä¸å¹³è¡¡ï¼ˆå¦‚æœæŸä¸ªcellçš„æ•°ç›®å°äº5 æˆ–å°äº10ï¼‰æ—¶ï¼Œå¡æ–¹åˆ†å¸ƒçš„ä¼°è®¡æ˜¯ä¸å……åˆ†çš„ã€‚è€Œä¸ä¹‹ç›¸åï¼ŒFisher ç²¾ç¡®æ£€éªŒï¼Œæ­£å¦‚å®ƒçš„åç§°ï¼Œä¸è®ºæ ·æœ¬å¦‚ä½•éƒ½èƒ½ä¿æŒç²¾ç¡®ã€‚ä½†æ˜¯å½“æ ·æœ¬é‡å¾ˆå¤§æˆ–éå¸¸å¹³è¡¡ï¼ŒFisher ç²¾ç¡®æ£€éªŒè®¡ç®—ä¸æ–¹ä¾¿ï¼Œè¿™æ—¶å°±éœ€è¦ç”¨åˆ°å¡æ–¹åˆ†å¸ƒã€‚</p>
</blockquote>
<p>å…¬å¼å¦‚ä¸‹ï¼š<br>
<img src="Fisher's_exact_test.png" alt="Fisher's_exact_test"></p>
<p>å°ç»“ï¼š<br>
Fisher ç²¾ç¡®æ£€éªŒä½¿ç”¨äº<strong>æ£€éªŒä¸¤ä¸ªåˆ†ç±»å˜é‡æ˜¯å¦ç›¸å…³</strong>ã€‚</p>
<ul>
<li>Fisher ç²¾ç¡®æ£€éªŒï¼šé€‚åˆæ ·æœ¬å°æˆ–ä¸å¹³è¡¡æ•°æ®</li>
<li>å¡æ–¹æ£€éªŒï¼šé€‚åˆæ ·æœ¬å¤§ã€å¹³è¡¡æ•°æ®</li>
</ul>
<h4 id="Two-By-Two-Tables">Two By Two Tables</h4>
<p>The data from the experiment above can be summarized by a 2 by 2 table:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line">tab&lt;-matrix(c(3,1,1,3),2,2)</span><br><span class="line">rownames(tab)&lt;-c(&quot;Poured Before&quot;,&quot;Poured After&quot;)</span><br><span class="line">colnames(tab)&lt;-c(&quot;Guessed before&quot;,&quot;Guessed after&quot;)</span><br><span class="line">tab</span><br><span class="line">#The function fisher.test performs the calculations above and can be obtained like this:</span><br><span class="line">fisher.test(tab,alternative&#x3D;&quot;greater&quot;)</span><br><span class="line">##</span><br><span class="line">##Fisher&#39;s Exact Test for Count Data</span><br><span class="line">##</span><br><span class="line">## data:  tab</span><br><span class="line">## p-value &#x3D; 0.2429</span><br><span class="line">## alternative hypothesis: true odds ratio is greater than 1</span><br><span class="line">## 95 percent confidence interval:</span><br><span class="line">##  0.3135693Inf</span><br><span class="line">## sample estimates:</span><br><span class="line">## odds ratio</span><br><span class="line">##   6.408309</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h4 id="Fisher-ç²¾ç¡®æ£€éªŒçš„äº‰è®®">Fisher ç²¾ç¡®æ£€éªŒçš„äº‰è®®</h4>
<p>å¾ˆå¤šäººäº‰è®ºFisher ç²¾ç¡®æ£€éªŒç»“æœä¿å®ˆã€‚å®ƒçš„çœŸå®çš„æ‹’ç»æ¦‚ç‡ä½äºåä¹‰ä¸Šçš„æ˜¾è‘—æ°´å¹³ã€‚The apparent contradiction stems from the combination of a discrete statistic with fixed significance levels.ï¼ˆæœ€æ˜æ˜¾çš„çŸ›ç›¾åœ¨äºå®ƒè®¡ç®—å‡ºçš„ç»Ÿè®¡é‡æ˜¯åˆ†æ•£çš„ï¼Œä½†æ˜¯æ˜¾è‘—æ°´å¹³æ˜¯å›ºå®šçš„ï¼Œå¦‚0.05ï¼‰ï¼ˆå¦‚ä¸Šä¾‹ï¼Œå¯èƒ½å‡ºç°çš„På€¼åªæœ‰5ä¸ªï¼ˆ0ï¼Œ1ï¼Œ2ï¼Œ3ï¼Œ4æ¯ï¼‰ï¼‰ã€‚To avoid the problem, many authors discourage the use of fixed significance levels when dealing with discrete problems.</p>
<h4 id="Chi-square-Test">Chi-square Test</h4>
<p>å‡è®¾æˆ‘ä»¬æœ‰250ä¸ªä¸ªä½“ï¼Œå…¶ä¸­ä¸€éƒ¨åˆ†æ‚£æœ‰ç–¾ç—…ã€‚æˆ‘ä»¬è§‚æµ‹åˆ°aaåŸºå› å‹çš„ä¸ªä½“ä¸­æœ‰20%æ‚£ç—…ï¼›å…¶ä»–åŸºå› å‹çš„ä¸ªä½“ä¸­æ‚£ç—…æ¯”ä¾‹ä¸º10%ã€‚å¦‚æœæˆ‘ä»¬å†æŒ‘250ä¸ªä¸ªä½“è¿™ç§ç°è±¡ä¼šå†æ¬¡å‘ç”Ÿå—ï¼Ÿï¼ˆè¿™ä¸ªæé—®æ–¹å¼ç­‰åŒäºé—®è¿™ä¸ªä½ç‚¹æ˜¯å¦ä¸ç–¾ç—…ç›¸å…³ï¼‰</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">disease=factor(<span class="built_in">c</span>(<span class="built_in">rep</span>(<span class="number">0</span>,<span class="number">180</span>),<span class="built_in">rep</span>(<span class="number">1</span>,<span class="number">20</span>),<span class="built_in">rep</span>(<span class="number">0</span>,<span class="number">40</span>),<span class="built_in">rep</span>(<span class="number">1</span>,<span class="number">10</span>)),labels=<span class="built_in">c</span>(<span class="string">&quot;control&quot;</span>,<span class="string">&quot;cases&quot;</span>))</span><br><span class="line">genotype=factor(<span class="built_in">c</span>(<span class="built_in">rep</span>(<span class="string">&quot;AA/Aa&quot;</span>,<span class="number">200</span>),<span class="built_in">rep</span>(<span class="string">&quot;aa&quot;</span>,<span class="number">50</span>)),levels=<span class="built_in">c</span>(<span class="string">&quot;AA/Aa&quot;</span>,<span class="string">&quot;aa&quot;</span>))</span><br><span class="line">dat&lt;-data.frame(disease, genotype)</span><br><span class="line">dat&lt;-dat[sample(nrow(dat)),]<span class="comment">#shuffle them up</span></span><br><span class="line">head(dat)</span><br><span class="line"></span><br><span class="line"><span class="comment">###ç»Ÿè®¡é¢‘æ•°</span></span><br><span class="line">table(genotype)</span><br><span class="line">table(disease)</span><br><span class="line"></span><br><span class="line"><span class="comment">## åŒå› ç´ ç»Ÿè®¡é¢‘æ•°ï¼Œå»ºç«‹ä¸€ä¸ª2*2è¡¨</span></span><br><span class="line">tab &lt;- table(genotype,disease)</span><br><span class="line">tab</span><br><span class="line"><span class="comment">##disease</span></span><br><span class="line"><span class="comment">## genotype control cases</span></span><br><span class="line"><span class="comment">##    AA/Aa     180    20</span></span><br><span class="line"><span class="comment">##    aa		 40    10</span></span><br></pre></td></tr></table></figure>
<p>è¿™é‡Œæœ€å…¸å‹çš„ç»Ÿè®¡é‡å°±æ˜¯the odds ratio(OR)ã€‚). We compute the oddsof having the disease if you are an â€œaaâ€: 10/40, the odds of having the disease if you are an â€œAA/Aaâ€:20/180, and take the ratio:(10/40)/(20/180)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">(tab[<span class="number">2</span>,<span class="number">2</span>]/tab[<span class="number">2</span>,<span class="number">1</span>])/(tab[<span class="number">1</span>,<span class="number">2</span>]/tab[<span class="number">1</span>,<span class="number">1</span>])</span><br><span class="line"><span class="comment">## [1] 2.25</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä¸ºäº†è®¡ç®—På€¼ï¼Œæˆ‘ä»¬ä¸ç›´æ¥ä½¿ç”¨ORã€‚æˆ‘ä»¬å‡è®¾åŸºå› å‹å’Œç–¾ç—…æ²¡æœ‰å…³ç³»ï¼Œç„¶åè®¡ç®—å‡ºæ¯ä¸ªcellçš„æœŸæœ›å€¼ã€‚åœ¨é›¶å‡è®¾ä¸‹ï¼Œä¸¤ç§åŸºå› å‹çš„200ä¸ªä½“å’Œ50ä¸ªä½“ï¼Œæ‚£ç—…çš„æ¦‚ç‡æ˜¯ç›¸åŒçš„ã€‚</p>
<p>é¦–å…ˆï¼Œæˆ‘ä»¬å¾—åˆ°ç¾¤ä½“ï¼ˆ200+50=250ï¼‰çš„æ‚£ç—…æ¦‚ç‡ä¸º</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">p=mean(disease==<span class="string">&quot;cases&quot;</span>)</span><br><span class="line">p</span><br><span class="line"><span class="comment">## [1] 0.12</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>å› æ­¤ï¼Œå„cell çš„æœŸæœ›å€¼ä¸º</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">expected&lt;-rbind(<span class="built_in">c</span>(<span class="number">1</span>-p,p)*<span class="built_in">sum</span>(genotype==<span class="string">&quot;AA/Aa&quot;</span>),<span class="built_in">c</span>(<span class="number">1</span>-p,p)*<span class="built_in">sum</span>(genotype==<span class="string">&quot;aa&quot;</span>))</span><br><span class="line"><span class="built_in">dimnames</span>(expected)&lt;-<span class="built_in">dimnames</span>(tab)</span><br><span class="line">expected</span><br><span class="line"><span class="comment">##disease</span></span><br><span class="line"><span class="comment">## genotype control cases</span></span><br><span class="line"><span class="comment">##    AA/Aa     176    24</span></span><br><span class="line"><span class="comment">##    aa		 44     6</span></span><br><span class="line"></span><br><span class="line"><span class="comment">##è¿™é‡Œæ²¡è®²æ¸…æ¥šï¼Œç›´æ¥å°±ç®—å‡ºæ¥äº†</span></span><br><span class="line">chisq.test(tab)$p.value</span><br><span class="line"><span class="comment">## [1] 0.08857435</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h4 id="Large-Samples-Small-p-values">Large Samples, Small p-values</h4>
<p>å¦‚ä¸Šæ‰€è¿°ï¼Œä»…æŠ¥é“På€¼æ˜¯ä¸æ­£ç¡®çš„ã€‚å¾ˆå¤šåšGWASçš„äººè¿‡äºå¼ºè°ƒPå€¼ã€‚ä»–ä»¬æ ·æœ¬å¾ˆå¤§ï¼Œå¾—åˆ°çš„På€¼å¾ˆå°ã€‚ä½†æ˜¯å¦‚æœä»”ç»†çœ‹ä»–ä»¬çš„ç»“æœï¼Œæˆ‘ä»¬å°±èƒ½çœ‹åˆ°ä»–ä»¬çš„odds ratioså¹¶ä¸é«˜ï¼šä»…ä»…æ¯”1é«˜ä¸€ç‚¹ã€‚åœ¨è¿™ç§æƒ…å†µä¸‹ï¼ŒåŸºå› å‹æ˜¯AA/Aa æˆ– aa å¯èƒ½ä¸ä¼šæ”¹å˜åŠ¨ç‰©æ‚£ç—…çš„å‡ ç‡ï¼Œä½†æ˜¯è¿™äº›ä½ç‚¹ç¡®å®æ˜¯ç»Ÿè®¡æ„ä¹‰ä¸Šæ˜¾è‘—çš„ã€‚<br>
<strong>odds ratio å’Œ p-value å¹¶æ²¡æœ‰ä¸€å¯¹ä¸€çš„å…³ç³»</strong>ï¼ˆè¿™ä¸€ç‚¹å¾ˆé‡è¦ï¼Œè¿™ä¸ªåœ¨effect size ä¸­ä¹Ÿä¸“é—¨æåˆ°äº†ï¼Œæ•ˆåº”å¤§å°ä¸æ˜¯å¦æ˜¾è‘—æ²¡æœ‰å› æœå…³ç³»ã€‚På€¼åªèƒ½åˆ¤æ–­è¿™ä¸ªæ•ˆåº”æ˜¯ä¸æ˜¯å‡ºäºå¶ç„¶ï¼ˆéšæœºè¯¯å·®ï¼‰ï¼Œå¹¶ä¸èƒ½è¯´æ˜æ•ˆåº”å¤§å°ï¼‰</p>
<h4 id="Confidence-Intervals-For-The-Odd-Ratio">Confidence Intervals For The Odd Ratio</h4>
<p>è®¡ç®—OR çš„ç½®ä¿¡åŒºé—´ä¸èƒ½ç›´æ¥åšåˆ°ã€‚ä¸åƒå…¶ä»–ç»Ÿè®¡é‡ï¼Œå®ƒä»¬éƒ½èƒ½æ¨å¯¼å‡ºè¿‘ä¼¼çš„åˆ†å¸ƒã€‚ä½†æ˜¯ORä»…ä»…æ˜¯ä¸€ä¸ªæ¯”ä¾‹ã€‚å› æ­¤ï¼Œæ²¡æœ‰ç®€å•çš„æ–¹æ³•å¯ä»¥ä½¿ç”¨ï¼Œä¾‹å¦‚CLTã€‚</p>
<p>One approach is to use the theory ofgeneralized linear models which provides estimates of thelog odds ratio, rather than the OR itself, that can be shown to be asymptotically normal.</p>
<p>åé¢çš„æ²¡çœ‹</p>
<h2 id="Exploratory-Data-Analysis">Exploratory Data Analysis</h2>
<p>è¿™ä¸ªæ¢ç´¢æ€§æ•°æ®åˆ†æï¼Œæ²¡å¤ªææ‡‚</p>
<h3 id="Quantile-Quantile-Plots">Quantile Quantile Plots</h3>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">library(UsingR)<span class="comment">##available from CRAN</span></span><br><span class="line">library(rafalib)</span><br><span class="line">x&lt;-father.son$fheight</span><br><span class="line"></span><br><span class="line">ps&lt;-( seq(<span class="number">0</span>,<span class="number">99</span>)+<span class="number">0.5</span>)/<span class="number">100</span> <span class="comment">#0.005 0.015 â€¦â€¦</span></span><br><span class="line">qs&lt;-quantile(x, ps) <span class="comment">#è·å¾—xä¸­çš„å„åˆ†ä½æ•°</span></span><br><span class="line">normalqs&lt;-qnorm(ps, mean(x), popsd(x)) <span class="comment">#å½¢æˆä»¥xå‡å€¼ï¼Œæ€»ä½“æ ‡å‡†å·®çš„æ­£æ€åˆ†å¸ƒçš„åˆ†ä½æ•°</span></span><br><span class="line">plot(normalqs,qs,xlab=<span class="string">&quot;Normal percentiles&quot;</span>,ylab=<span class="string">&quot;Height percentiles&quot;</span>) </span><br><span class="line"><span class="comment"># ä¸¤ä¸ªåˆ†ä½æ•°å‘é‡çš„æ•£ç‚¹å›¾å³ä¸ºQQå›¾</span></span><br><span class="line">abline(<span class="number">0</span>,<span class="number">1</span>)<span class="comment">##identity line</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ç®€åŒ–å‘½ä»¤,æ¯”ä¸Šé¢çš„æ‰‹åŠ¨ç”»å›¾é‡‡äº†æ›´å¤šçš„ç‚¹ï¼Œå› æ­¤æ›´åŠ å‡†ç¡®ã€‚<br>
å…¶ä¸­ç”±äºxè½´ä¸ºæ ‡å‡†æ­£æ€åˆ†å¸ƒï¼Œå› æ­¤æ·»åŠ çš„qqlineçš„æˆªè·ä¸ºmean(x) , æ–œç‡ä¸ºpopsd(x)</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">qqnorm(x)</span><br><span class="line">qqline(x)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>å½“ç„¶ï¼Œæˆ‘ä»¬ä¹Ÿå¯ä»¥çœ‹çœ‹éæ­£æ€åˆ†å¸ƒçš„QQå›¾ï¼Œæ¯”å¦‚tåˆ†å¸ƒï¼Œæ ·æœ¬æ•°è¶Šå°‘ï¼Œå°¾å·´è¶Šè‚¥ã€‚ä½“ç°åœ¨QQå›¾ä¸Šå°±æ˜¯å·¦ä¾§çš„æç«¯å€¼ä½äºidentity lineï¼Œå³ä¾§çš„æç«¯å€¼é«˜äºidentity lineã€‚</p>
<h3 id="Boxplots">Boxplots</h3>
<p>æ•°æ®ä¹Ÿä¸æ˜¯éƒ½æœä»æ­£æ€åˆ†å¸ƒã€‚æ¯”å¦‚å·¥èµ„ï¼Œåªå‘Šè¯‰å‡å€¼å’Œæ ‡å‡†å·®æ ¹æœ¬ä¸è¶³ä»¥è¡¨ç°çœŸå®æƒ…å†µï¼ˆnot neccessarily informativeï¼‰ï¼Œå› ä¸ºæˆ‘ä»¬æ²¡åŠæ³•é€šè¿‡è¿™ä¸¤ä¸ªå‚æ•°è·å–å·¥èµ„çš„åˆ†å¸ƒæƒ…å†µã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">op &lt;- par(mfrow=<span class="built_in">c</span>(<span class="number">1</span>,<span class="number">2</span>))</span><br><span class="line">hist(exec.pay)<span class="comment">##in UsingR package</span></span><br><span class="line">qqnorm(exec.pay)</span><br><span class="line">qqline(exec.pay)</span><br><span class="line"></span><br><span class="line"><span class="comment">#boxplot</span></span><br><span class="line">boxplot(exec.pay, ylab=<span class="string">&quot;10,000s of dollars&quot;</span>, ylim=<span class="built_in">c</span>(<span class="number">0</span>,<span class="number">400</span>))</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="Scatterplots-And-Correlation">Scatterplots And Correlation</h3>
<p>ä¸Šé¢éƒ½æ˜¯å•ä¸€å˜é‡ã€‚æˆ‘ä»¬ä¹Ÿç»å¸¸ç”¨åˆ°åŒå˜é‡æˆ–è€…æ›´å¤šçš„å˜é‡ï¼Œæ¯”å¦‚çˆ¶äº²å’Œå„¿å­çš„èº«é«˜ã€‚å¯ä»¥å…ˆç”»ä¸€ä¸ªæ•£ç‚¹å›¾ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">library(UsingR)</span><br><span class="line">data(<span class="string">&quot;father.son&quot;</span>)</span><br><span class="line">x=father.son$fheight</span><br><span class="line">y=father.son$sheight</span><br><span class="line">plot(x,y,xlab=<span class="string">&quot;Father&#x27;s height in inches&quot;</span>,ylab=<span class="string">&quot;Son&#x27;s height in inches&quot;</span>,main=paste(<span class="string">&quot;correlation =&quot;</span>,<span class="built_in">signif</span>(cor(x,y),<span class="number">2</span>)))</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>ä»å›¾ä¸­çœ‹å‡ºä¸€ä¸ªæ˜æ˜¾çš„è¶‹åŠ¿ï¼šçˆ¶äº²çš„èº«é«˜è¶Šé«˜ï¼Œå„¿å­çš„èº«é«˜ä¹Ÿè¶Šé«˜ã€‚ä¸€ä¸ªå®šé‡çš„ç»Ÿè®¡é‡å°±æ˜¯ç›¸å…³ç³»æ•°ã€‚</p>
<h4 id="Stratification">Stratification</h4>
<p>å¦‚æœæˆ‘ä»¬è¦é¢„æµ‹èº«é«˜ä¸º72 inches çš„çˆ¶äº²çš„å„¿å­çš„èº«é«˜ï¼Œæˆ‘ä»¬å¯ä»¥çœ‹æ‰€æœ‰èº«é«˜ä¸º72 inches çš„çˆ¶äº²çš„å„¿å­çš„æƒ…å†µã€‚å› æ­¤éœ€è¦å¯¹æ•°æ®æŒ‰çˆ¶äº²èº«é«˜è¿›è¡Œåˆ†å±‚ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">groups&lt;-split(y,<span class="built_in">round</span>(x))</span><br><span class="line">boxplot(groups)</span><br><span class="line">print(mean(y[ <span class="built_in">round</span>(x)==<span class="number">72</span>]))</span><br><span class="line"><span class="comment">## [1] 70.67719</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="Bi-variate-Normal-Distribution">Bi-variate Normal Distribution</h3>
<p>å¯¹äºä¸¤ä¸ªå‡æœä»æ­£æ€åˆ†å¸ƒçš„å˜é‡çš„è”åˆåˆ†å¸ƒï¼Œæˆ‘ä»¬ä¸€èˆ¬å¯ä»¥è®¾å®šå½“X = xæ—¶ï¼ŒYçš„æ¡ä»¶åˆ†å¸ƒã€‚ä»¥èº«é«˜æ•°æ®ä¸ºä¾‹ï¼Œæˆ‘ä»¬ç»™å‡ºäº†ä»¥ä¸‹4ä¸ªå±‚æ¬¡ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">groups&lt;-split(y,<span class="built_in">round</span>(x))</span><br><span class="line">op &lt;- par(mfrow=<span class="built_in">c</span>(<span class="number">2</span>,<span class="number">2</span>))</span><br><span class="line"><span class="keyword">for</span>(i <span class="keyword">in</span> <span class="built_in">c</span>(<span class="number">5</span>,<span class="number">8</span>,<span class="number">11</span>,<span class="number">14</span>))&#123;</span><br><span class="line">  qqnorm(groups[[i]],main=paste0(<span class="string">&quot;X=&quot;</span>,<span class="built_in">names</span>(groups)[i],<span class="string">&quot;strata&quot;</span>),</span><br><span class="line">         ylim=<span class="built_in">range</span>(y),xlim=<span class="built_in">c</span>(-<span class="number">2.5</span>,<span class="number">2.5</span>))</span><br><span class="line"><span class="comment"># groups æ˜¯ä¸€ä¸ªlistï¼Œç”¨groups[[i]] è·å¾—çš„æ˜¯å‘é‡</span></span><br><span class="line"><span class="comment"># range() è¿”å›ä¸€ä¸ªåŒ…å«æœ€å°å€¼å’Œæœ€å¤§å€¼çš„å‘é‡</span></span><br><span class="line"><span class="comment"># range(1:10) ## [1]  1 10</span></span><br><span class="line"></span><br><span class="line">  qqline(groups[[i]])</span><br><span class="line">  &#125;</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>åé¢çš„å…¬å¼æ¨å¯¼æ²¡å¤ªæ˜ç™½ã€‚åé¢å¯¹x,yè¿›è¡Œäº†æ ‡å‡†åŒ–ï¼Œç„¶åæŒ‰xåˆ†ç»„æ±‚yå‡å€¼ï¼Œå‘ç°yå‡å€¼ ä¸ x å‘ˆçº¿æ€§å…³ç³»ï¼Œæˆªè·ä¸º0ï¼Œæ–œç‡ä¸ºæ ‡å‡†åŒ–çš„yä¸xçš„ç›¸å…³ç³»æ•°ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">x=( x-mean(x) )/sd(x)</span><br><span class="line">y=( y-mean(y) )/sd(y)</span><br><span class="line">means=tapply(y, <span class="built_in">round</span>(x*<span class="number">4</span>)/<span class="number">4</span>, mean) <span class="comment">#æŒ‰round(x*4)/4 ç»™y åˆ†ç»„æ±‚å‡å€¼</span></span><br><span class="line">fatherheights=<span class="built_in">as.numeric</span>(<span class="built_in">names</span>(means))</span><br><span class="line">mypar(<span class="number">1</span>,<span class="number">1</span>)</span><br><span class="line">plot(fatherheights, means, ylab=<span class="string">&quot;average of strata of son heights&quot;</span>, ylim=<span class="built_in">range</span>(fatherheights))</span><br><span class="line">abline(<span class="number">0</span>, cor(x,y))</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h3 id="Plots-To-Avoid">Plots To Avoid</h3>
<p>ç”»å›¾çš„ç›®çš„æ˜¯å‡†ç¡®è€Œæ¸…æ¥šåœ°å±•ç¤ºæ•°æ®ã€‚æ ¹æ® Karl çš„è¯´æ³•ï¼Œä¸å¥½çš„ç”»å›¾çš„æ ‡å‡†å¦‚ä¸‹ï¼š</p>
<ol>
<li>Display as little information as possible.</li>
<li>Obscure what you do show (with chart junk).</li>
<li>Use pseudo-3D and color gratuitouslyï¼ˆå…è´¹åœ°ï¼‰.</li>
<li>Make a pie chart (preferably in color and 3D).</li>
<li>Use a poorly chosen scale.</li>
<li>Ignore significant figures.</li>
</ol>
<h3 id="Pie-charts">Pie charts</h3>
<blockquote>
<p>â€œPie charts are a very bad way of displaying information. The eye is good at judginglinear measures and bad at judging relative areas. A bar chart or dot chart is a preferableway of displaying this type of data.â€</p>
</blockquote>
<p>é¥¼å›¾æ˜¯ä¸€ç§å¾ˆå·®åŠ²çš„å±•ç¤ºæ•°æ®çš„æ–¹å¼ã€‚äººçœ¼å¾ˆå®¹æ˜“è¾¨åˆ«çº¿æ€§æ–¹å¼ï¼Œä½†æ˜¯åŠ£äºåˆ¤æ–­ç›¸å¯¹åŒºåŸŸã€‚ä¸€ä¸ªæ¡å½¢å›¾æˆ–æ•£ç‚¹å›¾ éƒ½æ˜¯æ›´åˆé€‚çš„æ–¹å¼ã€‚</p>
<p>ä¸¾ä¾‹æ¥è¯´ï¼Œä½ å¯ä»¥çœ‹ç€é¥¼å›¾ï¼Œå»ä¼°è®¡æŸä¸€éƒ¨åˆ†å¯¹åº”çš„ç™¾åˆ†æ¯”ã€‚ä½ ä¼šå‘ç°è¿™å¾ˆå›°éš¾ï¼Œé™¤éè¿™ä¸ªæ¯”ä¾‹å¾ˆæ¥è¿‘25%ï¼Œ50% æˆ– 75%ã€‚</p>
<p>è¿™é‡Œæˆ‘ä»¬å¯ä»¥ç”¨æ¡å½¢å›¾ç”¨ä»£æ›¿ï¼Œè¿™é‡Œæˆ‘ä»¬æ¯éš”10æ·»åŠ äº†ä¸€æ¡æ°´å¹³çº¿</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">pie(browsers,main=<span class="string">&quot;Browser Usage (August 2013)&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="comment">#æ¡å½¢å›¾</span></span><br><span class="line">barplot(browsers, main=<span class="string">&quot;Browser Usage (August 2013)&quot;</span>, ylim=<span class="built_in">c</span>(<span class="number">0</span>,<span class="number">55</span>))</span><br><span class="line">abline(h=<span class="number">1</span>:<span class="number">5</span>*<span class="number">10</span>)</span><br><span class="line">barplot(browsers, add=<span class="literal">TRUE</span>)</span><br><span class="line"></span><br></pre></td></tr></table></figure>
<p>é€šè¿‡æ¡å½¢å›¾å’Œyè½´çš„æ°´å¹³çº¿ï¼Œæˆ‘ä»¬å¯ä»¥å¾ˆæ¸…æ¥šçš„çœ‹åˆ°æ¯ä¸ªå€¼çš„æ¯”ä¾‹ã€‚</p>
<p>å°½é‡ä¸è¦ä½¿ç”¨3Då›¾åƒï¼Œå› ä¸º3Då›¾åƒæ·»åŠ äº†çœ‹å›¾çš„éšœç¢ï¼Œä½¿å¾—æ›´éš¾é€šè¿‡è‚‰çœ¼è¾¨åˆ«å„é¡¹æ¯”ä¾‹ã€‚</p>
<h4 id="Barplots-as-data-summaries">Barplots as data summaries</h4>
<p>å½“æ¡å½¢å›¾ç”¨äºå±•ç¤ºç™¾åˆ†æ¯”æ—¶ï¼Œå¾ˆæœ‰ç”¨ã€‚ä½†æ˜¯ï¼Œæ¡å½¢å›¾ä¹Ÿä¼šè¢«é”™è¯¯åœ°ç”¨äº<strong>åªæœ‰ä¸¤ç»„</strong>çš„æ¯”è¾ƒã€‚ç‰¹åˆ«å½“æ¡å½¢å›¾çš„é«˜åº¦è¡¨ç¤ºå‡å€¼ï¼Œä¸Šé¢çš„å°è§¦è§’è¡¨ç¤ºæ ‡å‡†è¯¯ï¼ˆstandard errorsï¼‰ã€‚</p>
<blockquote>
<p>Much more informative is to summarize with a boxplot. If the number of points is small enough,we might as well add them to the plot. When the number of points is too large for us to see them,just showing a boxplot is preferable. We can even setrange=0inboxplotto avoid drawing manyoutliers when the data is in the range of millions.</p>
</blockquote>
<p>è¿™ä¸ªæ—¶å€™æ›´é€‚åˆç”¨ç®±çº¿å›¾æ¥è¡¨ç¤ºã€‚ç®±çº¿å›¾å±•ç¤ºäº†æ›´å¤šçš„ä¿¡æ¯ï¼Œä¸­å¿ƒã€åˆ†å¸ƒã€ç¦»ç¾¤ç‚¹ã€‚è€Œæ¡å½¢å›¾åªå±•ç¤ºäº†å‡å€¼å’Œæ ‡å‡†è¯¯ï¼Œè€Œæ ‡å‡†è¯¯å¾ˆå¤§ç¨‹åº¦ä¸Šå’Œæ ·æœ¬å¤§å°æœ‰å…³ï¼Œå’Œåˆ†å¸ƒæ— å…³ã€‚</p>
<blockquote>
<p>Notice how much more we see here: the center, spread, range and the points themselves. In thebarplot, we only see the mean and the SE, and the SE has more to do with sample size, than withthe spread of the data.</p>
</blockquote>
<p>é—®é¢˜è¿˜ä¸æ­¢å¦‚æ­¤ã€‚å½“æ•°æ®ä¸­çš„ç¦»ç¾¤ç‚¹å¾ˆå¤§ï¼ˆåˆ†ç»„æ•°æ®çš„å°¾å·´å¾ˆå¤§ï¼‰ã€‚å¦‚ä¸‹æ¡å½¢å›¾ï¼Œçœ‹ä¼¼ä¸¤ç»„åŒºåˆ«å¾ˆå¤§ã€‚</p>
<p><img src="barplot1.png" alt="barplot1"></p>
<p>ä½†ä»”ç»†çœ‹æ•°æ®ï¼Œè¿™ç§åŒºåˆ«åŸæ¥ä¸»è¦æ˜¯å› ä¸ºæœ‰ä¸¤ä¸ªå¾ˆå¤§çš„ç¦»ç¾¤ç‚¹ã€‚å°†æ•°æ®è¿›è¡Œlogä»¥åï¼Œæ›´åŠ æ˜æœ—ã€‚</p>
<p>å¦‚ä¸‹ç®±çº¿å›¾æ‰€ç¤ºï¼Œå·¦è¾¹ä¸ºåŸæ•°æ®ï¼Œå¯ä»¥çœ‹å‡ºäºŒè€…å¯èƒ½åŒºåˆ«ä¸å¤§ï¼Œæœ‰ä¸¤ä¸ªéå¸¸å¤§çš„ç¦»ç¾¤ç‚¹ã€‚logååŸºæœ¬å±•ç¤ºäº†åŸºæœ¬æƒ…å†µã€‚ï¼ˆä¸ªäººæ„Ÿè§‰ï¼Œé‚£ä¸¤ä¸ªç¦»ç¾¤ç‚¹å®åœ¨å¤ªå¤§ï¼Œä¹Ÿå¯ä»¥ç›´æ¥å‰”é™¤ï¼‰</p>
<p><img src="boxplot1.png" alt="barplot1"></p>
<h4 id="Show-the-scatter-plot">Show the scatter plot</h4>
<p>å¾ˆå¤šç»Ÿè®¡åˆ†æéƒ½æ˜¯è¯æ˜ä¸¤ä¸ªå˜é‡ä¹‹é—´å­˜åœ¨å…³ç³»ã€‚ç›¸å…³ç³»æ•°å’Œç›¸åº”çš„å›¾å¯ä»¥å±•ç¤ºè¿™ä¸€ç‚¹ã€‚ç„¶è€Œï¼Œåªå±•ç°å›å½’çº¿æ˜¯ä¸å¤Ÿçš„ï¼Œå› ä¸ºè¿™æ©é¥°äº†æ•£ç‚¹ã€‚</p>
<h4 id="High-correlation-does-not-imply-replication">High correlation does not imply replication</h4>
<p>ç›¸å…³ç³»æ•°å¾ˆé«˜ï¼Œä¸ä»£è¡¨é‡å¤æ€§å¾ˆå¥½ï¼Œè¿™ä¸€éƒ¨åˆ†æ²¡çœ‹æ‡‚ã€‚ä¹‹åè¡¥ä¸Š</p>
<h4 id="Barplots-for-paired-data">Barplots for paired data</h4>
<p>å½“æ ·æœ¬æ•°è¾ƒå°‘ä¸”ä¸ºé…å¯¹å®éªŒæ—¶ï¼Œé€šå¸¸æˆ‘ä»¬ä¼šç”¨ä¸¤ç§é¢œè‰²çš„æ¡å½¢å›¾è¡¨ç¤ºæ¯”å¦‚å®éªŒå‰å’Œå®éªŒåã€‚</p>
<p><img src="pairdata1.png" alt="barplot1"></p>
<p>ä½†æ˜¯æˆ‘ä»¬æœ‰æ›´å¥½çš„åŠæ³•ã€‚æ¯”å¦‚æ•£ç‚¹å›¾ï¼Œæˆ‘ä»¬å¯ä»¥çœ‹åˆ°ç‚¹éƒ½æ˜¯é«˜äºidentity lineï¼ˆè¿™é‡Œåº”è¯¥æ˜¯y=xï¼‰ã€‚å¦ä¸€ç§æ–¹æ³•æ˜¯ç›´æ¥ç”»å·®å€¼çš„æ•£ç‚¹å›¾ã€‚</p>
<p><img src="pairdata2.png" alt="barplot1"></p>
<p><img src="pairdata3.png" alt="barplot1"></p>
<h4 id="Gratuitous-3D-ï¼ˆæ²¡æœ‰å¿…è¦çš„3Dï¼‰">Gratuitous 3D ï¼ˆæ²¡æœ‰å¿…è¦çš„3Dï¼‰</h4>
<p>ä»¥ä¸‹æ˜¯ä¸€ä¸ª3Då›¾ï¼Œä½ ä¼šå‘ç°å¹¶ä¸æ¸…æ¥šã€‚å¾ˆéš¾å»ç¡®å®šå…¶ä¸­æŸä¸€ä¸ªç‚¹å¯¹åº”çš„æ•°å€¼ã€‚</p>
<p>å…¶å®è¿™åªè¦ä¸åŒçš„é¢œè‰²æ¥åŒºåˆ†ä¸‰æ¡çº¿å°±å¯ä»¥äº†ã€‚</p>
<p><img src="3D.png" alt="barplot1"></p>
<p><img src="3D2.png" alt="barplot1"></p>
<h4 id="Too-many-significant-digits">Too many significant digits</h4>
<p>é»˜è®¤æƒ…å†µä¸‹ï¼Œæˆ‘ä»¬ä¼šå¾—åˆ°å¾ˆå¤šé‡è¦çš„æ•°å­—ï¼Œä½†æˆ‘ä»¬ä¸æ˜¯éƒ½è¦æ”¾ä¸Šå»ã€‚ä»Rä¸­ç›´æ¥å¤åˆ¶ç²˜è´´æ˜¯å¾ˆæ„šè ¢çš„ï¼Œè¿™å¯¼è‡´ä½ æœ€åæ˜¾ç¤ºçš„ä¸æ˜¯ä¸€å¼ å›¾ï¼Œè€Œæ›´åƒæ˜¯ä¸€å¼ è¡¨ã€‚</p>
<h4 id="Displaying-data-well">Displaying data well</h4>
<p>å±•ç¤ºæ•°æ®çš„ä¸€èˆ¬å‡†åˆ™ï¼š</p>
<ul>
<li>Be accurate and clear.</li>
<li>Let the data speak.</li>
<li>Show as much information as possible, taking care not to obscure the message.</li>
<li>Science not sales: avoid unnecessary frills (esp. gratuitous 3D).<br>
(ç§‘å­¦ä¸æ˜¯é”€å”®ï¼Œé¿å…ä¸å¿…è¦çš„è™šæ¶å­ï¼Œæ¯”å¦‚3D)</li>
<li>In tables, every digit should be meaningful. Donâ€™t drop ending 0â€™s.<br>
ï¼ˆè¡¨æ ¼ä¸­ï¼Œæ¯ä¸€ä¸ªå°æ•°ä½ç‚¹éƒ½æœ‰æ„ä¹‰ï¼‰</li>
</ul>
<h3 id="Misunderstanding-Correlation-Advanced">Misunderstanding Correlation (Advanced)</h3>
<p>ç°åœ¨ç›¸å…³ç³»æ•°ä¸»è¦ç”¨äºä½“ç°reproducibilityã€‚ä½†æ˜¯å®é™…ä¸Šï¼Œç›¸å…³ç³»æ•°ä¸è¶³ä»¥ä½“ç°reproducibilityã€‚</p>
<p>æœ€åçš„é”™è¯¯æ˜¯ï¼Œè®¡ç®—ç›¸å…³ç³»æ•°çš„æ•°æ®ä¸æ˜¯è¿‘ä¼¼æœä»<strong>äºŒå…ƒæ­£æ€åˆ†å¸ƒ</strong>ã€‚ä¸Šé¢æåˆ°ï¼Œå‡å€¼ï¼Œæ ‡å‡†å·®ï¼Œç›¸å…³ç³»æ•°éƒ½æ˜¯åŸºäºäºŒå…ƒæ­£æ€åˆ†å¸ƒã€‚ä½†æ˜¯ï¼Œæœ‰å¾ˆå¤šæ•°æ®ä¸ç¬¦åˆï¼Œå¦‚åŸºå› è¡¨è¾¾é‡æ•°æ®ï¼Œé€šå¸¸æœ‰å¾ˆè‚¥çš„å³ä¾§å°¾å·´ã€‚</p>
<p>è®¡ç®—reproducibility æœ€ç›´æ¥çš„æ–¹å¼å°±æ˜¯è®¡ç®—ä¸¤ä¸ªå˜é‡çš„è·ç¦»ã€‚</p>
<p>The standard way to quantify reproducibility between two sets of replicated measurements, sayx1; : : : ; xnandy1; : : : ; yn, is simply to compute the distance between themã€‚</p>
<p><img src="reproducibility1.png" alt="barplot1"></p>
<p>åé¢æ²¡çœ‹ï¼Œä¹Ÿçœ‹ä¸æ‡‚ã€‚</p>
<p>æˆ‘è¿è¿™é‡Œçš„reproducibilityæ˜¯ä»€ä¹ˆæ„æ€éƒ½ä¸çŸ¥é“ã€‚</p>
<h3 id="Robust-Summaries">Robust Summaries</h3>
<p>åœ¨åˆ†ææ•°æ®çš„è¿‡ç¨‹ä¸­ï¼Œæˆ‘ä»¬é€šå¸¸å‡è®¾æ•°æ®ç¬¦åˆæ­£æ€åˆ†å¸ƒã€‚ä½†æ˜¯æˆ‘ä»¬æ”¶é›†æ•°æ®çš„è¿‡ç¨‹ä¸­ï¼Œæ€»æ˜¯ä¸å¯é¿å…åœ°åŠ å…¥äº†ä¸€äº›é”™è¯¯çš„æ•°æ®ç‚¹ï¼ˆç¦»ç¾¤ç‚¹ï¼‰ã€‚ç±»ä¼¼è¿™ç§åç¦»å¾ˆå¤§çš„ç¦»ç¾¤ç‚¹ï¼Œä¼šå¯¼è‡´åˆ†æç»“æœå‡ºç°å¾ˆå¤§çš„åå·®ã€‚</p>
<p>å¦‚ä¸‹ä¾‹ï¼Œæˆ‘ä»¬åœ¨0åˆ°1ä¹‹é—´çš„æ•°æ®ä¸­ï¼Œæ’å…¥ä¸€ä¸ªå€¼ä¸º100çš„ç¦»ç¾¤ç‚¹ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">set.seed(<span class="number">1</span>)</span><br><span class="line">x=<span class="built_in">c</span>(rnorm(<span class="number">100</span>,<span class="number">0</span>,<span class="number">1</span>))<span class="comment">##real distribution</span></span><br><span class="line">x[<span class="number">23</span>]&lt;-<span class="number">100</span><span class="comment">##mistake made in 23th measurement</span></span><br><span class="line">boxplot(x)</span><br></pre></td></tr></table></figure>
<p>æˆ‘ä»¬çœ‹ä¸€ä¸‹å‡å€¼å’Œæ ‡å‡†å·®ï¼Œå‘ç°åç¦»å¾ˆå¤§</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">cat(<span class="string">&quot;The average is&quot;</span>,mean(x),<span class="string">&quot;and the SD is&quot;</span>,sd(x))</span><br><span class="line"></span><br><span class="line"><span class="comment">## The average is 1.108142 and the SD is 10.02938</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>
<h4 id="The-median">The median</h4>
<p>è¿™ä¸ªä¸­ä½æ•°ï¼Œå¯¹äºç¦»ç¾¤ç‚¹ï¼Œæ˜¯robustçš„ã€‚</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">median(x)</span><br><span class="line">[<span class="number">1</span>] <span class="number">0.1684483</span></span><br></pre></td></tr></table></figure>
<h4 id="The-median-absolute-deviation">The median absolute deviation</h4>
<p>![](The median absolute deviation.png)</p>
<h4 id="Spearman-correlation">Spearman correlation</h4>
<p>å’Œå‰é¢ç±»ä¼¼ï¼Œä¹Ÿæ˜¯robust çš„ç›¸å…³ç³»æ•°ã€‚å‡ ä¹ä¸å—ç¦»ç¾¤ç‚¹çš„å½±å“ã€‚</p>
<h4 id="Symmetry-of-log-ratios">Symmetry of log ratios</h4>
<p>å½“æ¶‰åŠæ¯”å€¼æ•°æ®æ—¶ï¼Œæ•°æ®ä¸æ˜¯å¯¹ç§°çš„ï¼Œ1/32 æ¯” 32/1 æ›´æ¥è¿‘1ã€‚è¿™æ—¶å€™é‡‡ç”¨log()å‡½æ•°å¤„ç†ä¸€ä¸‹ï¼Œå°±ä¼šå¯¹ç§°ï¼ˆlog(x/y) = log(x) - log(y) ï¼‰</p>
<h3 id="Wilcoxon-Rank-Sum-Test">Wilcoxon Rank Sum Test</h3>
<p>ä»ä¸Šé¢çŸ¥é“ï¼Œæ ·æœ¬å‡å€¼å’Œæ ‡å‡†å·®æ˜“å—ç¦»ç¾¤ç‚¹çš„å½±å“ã€‚tæ£€éªŒéœ€è¦ç”¨åˆ°è¿™ä¸¤ä¸ªæ•°ï¼Œå› æ­¤ä¹Ÿå®¹æ˜“å—åˆ°å½±å“ã€‚è¿™ç§Wilcoxon Rank Sum Test å°±ä¸ä¼šå—åˆ°å½±å“â€¦â€¦</p>
<p>åé¢æ²¡çœ‹</p>
<h2 id="Effect-size-â€“-wikipedia">Effect size â€“ wikipedia</h2>
<p>In statistics, an effect size is a quantitative measure of the magnitude of a phenomenon.(è¿™ä¸ªeffect size ä¸çŸ¥é“å’‹ç¿»è¯‘,æ•ˆåº”æ°´å¹³ï¼Ÿ)ã€‚ä¸¾ä¾‹è¯´æ˜ï¼ŒåŒ…æ‹¬ç›¸å…³ç³»æ•°ï¼Œå›å½’ç³»æ•°ï¼Œå‡å€¼å·®ã€‚å¯¹äºå¤§éƒ¨åˆ†çš„effect sizeï¼Œç»å¯¹å€¼è¶Šå¤§ï¼Œå¾€å¾€è¯´æ˜æ•ˆåº”è¶Šå¼ºã€‚</p>
<h3 id="Relationship-to-test-statistics">Relationship to test statistics</h3>
<p>åŸºäºé‡‡æ ·çš„effect sizes å’Œ å‡è®¾æ£€éªŒçš„ç»Ÿè®¡æ£€éªŒé‡ä¸ä¸€æ ·ã€‚effect sizes æ˜¯æ£€éªŒ strength(magnitude ) ï¼›ç„¶è€Œå‡è®¾æ£€éªŒæ˜¯ç»™å‡ºä¸€ä¸ªæ˜¾è‘—æ€§æ°´å¹³ï¼ˆPå€¼ï¼‰ï¼Œååº”è¿™ä¸ªmagnitude æˆ– relationship æ˜¯ä¸æ˜¯å‡ºäºå¶ç„¶ï¼ˆéšæœºè¯¯å·®ï¼‰ã€‚effect size å¹¶ä¸ç›´æ¥å†³å®šæ˜¾è‘—æ€§æ°´å¹³ï¼Œåä¹‹äº¦ç„¶ã€‚<strong>åªè¦æ ·æœ¬é‡è¶³å¤Ÿå¤§ï¼Œå“ªæ€•effect size å¾ˆå°ï¼Œä¹Ÿèƒ½å¾—åˆ°ç»Ÿè®¡æ˜¾è‘—ç»“æœ</strong>ã€‚å› æ­¤ï¼Œåªçœ‹på€¼å¯èƒ½æœ‰ä¸€å®šçš„è¯¯å¯¼ã€‚</p>
<h3 id="Standardized-and-unstandardized-effect-sizes">Standardized and unstandardized effect sizes</h3>
<p>effect sizes æ ‡å‡†åŒ–çš„åŸå› ï¼š<br>
1. å˜é‡çš„å•ä½æœ¬èº«æ²¡æœ‰æ„ä¹‰<br>
2. ç»“æœæ¥è‡ªå¤šé‡ç»“æœçš„åˆå¹¶<br>
3. ä¸åŒçš„ç»“æœç”¨äº†ä¸åŒçš„å•ä½<br>
4. it is desired to convey the size of an effect relative to the variability in the population</p>
<h3 id="Types">Types</h3>
<h4 id="Correlation-family">Correlation family</h4>
<ol>
<li>
<p>Pearson r</p>
</li>
<li>
<p>Coefficient of determination (r^2^ or R^2^)<br>
Pearson r çš„å¹³æ–¹</p>
<blockquote>
<p>In the case of paired data, this is a measure of the proportion of variance shared by the two variables, and varies from 0 to 1. (åœ¨é…å¯¹æ•°æ®ä¸­ï¼Œä¸€ä¸ªå˜é‡ä¸­çš„æ–¹å·®èƒ½è¢«å¦ä¸€ä¸ªå˜é‡è§£é‡Šçš„æ¯”ä¾‹)</p>
</blockquote>
</li>
</ol>
<h4 id="Difference-family">Difference family</h4>
<p>Effect sizes based on differences between means</p>
<p><img src="Difference_family1.png" alt="Difference_family1"></p>
<p>å…¶ä¸­Î¼1 Î¼2 æ˜¯ä¸¤ä¸ªç¾¤ä½“çš„å‡å€¼ï¼›Ïƒæ˜¯åŸºäºä¸€ä¸ªç¾¤ä½“æˆ–ä¸¤ä¸ªç¾¤ä½“çš„æ ‡å‡†å·®ã€‚</p>
<p>è¿™ç§å½¢å¼å¾ˆåƒtç»Ÿè®¡é‡ï¼Œä½†æ˜¯tç»Ÿè®¡é‡åŒ…å«äº†ä¸€ä¸ªå› ç´ ï¼Œæ ¹å·Nã€‚è¿™æ„å‘³ç€ï¼Œå¯¹äºç»™å®šçš„effect sizeï¼Œæ˜¾è‘—æ°´å¹³ä¼šéšç€æ ·æœ¬é‡çš„å¢åŠ è€Œå¢åŠ ã€‚ Unlike the t-test statistic, the effect size aims to estimate a population parameter and is not affected by the sample size.</p>
<h5 id="Cohenâ€™s-d">Cohenâ€™s d</h5>
<p>![Difference_family1](Cohenâ€™s d.png)</p>
<h2 id="PEVD-kennedy-1993">PEVD -kennedy 1993</h2>
<blockquote>
<p>Unbiased comparisons between animals  can, how- ever, be made if  animals are sampled from a common genetic population, in which case the consequence of â€œdisconnectednessâ€ is reflected   only in the PEV of differences between EBV of animals. The PEV of the individual animal (diagonal element) is unchanged given the same number of animals. The PEV of difference in EBV, however, depends on whether comparisons are within or between management units.</p>
</blockquote>
<p>è¿™æ˜¯å¥½åƒæœ€æ—©æåŠPEVDçš„æ–‡çŒ®ã€‚ä½†æ˜¯å…·ä½“è®¡ç®—çš„éƒ¨åˆ†ï¼Œæ²¡æœ‰æåŠã€‚åªæ˜¯è¯´ç°åœ¨éƒ½æœ€å¤šåªæä¾›ç³»æ•°çŸ©é˜µé€†çŸ©é˜µå¯¹è§’çº¿å…ƒç´ ï¼ˆMisztal å’Œ Wiggans1988,Meyer,1989ï¼‰ï¼Œä½†æ²¡æœ‰æä¾›éå¯¹è§’çº¿å…ƒç´ çš„å€¼ã€‚</p>
<p>å› æ­¤PEVDä¸èƒ½è®¡ç®—ï¼Œç„¶åå°±ç»™å‡ºäº†ä¸‰ç§æ›¿ä»£æ–¹æ¡ˆã€‚å·´æ‹‰å·´æ‹‰ã€‚åé¢å¯¹æˆ‘æ²¡ä½œç”¨ã€‚</p>
<h1>Changes in connectedness over time in alternative sheep sire referencing schemes - Kuehn 2008</h1>
<p>è¿™æ˜¯åœ¨Yuçš„æ–‡ç« ä¸­æåˆ°äº†ä¸€å˜´ï¼Œè¯´PEVD çš„ scale ,ä¹Ÿå°±æ˜¯å»é™¤å•ä½ï¼Œæ˜¯æŒ‰ç…§è¿™ç¯‡æ–‡ç« çš„æ–¹æ³•ã€‚</p>
<p>è¿™ç¯‡æ–‡ç« ç”¨äº†ä¸¤ä¸ªæŒ‡æ ‡ï¼Œç¬¬ä¸€ä¸ªæ˜¯ onnectedness correlation (Lewis etal., 1999, 2005)</p>
<p><img src="r.png" alt="r"></p>
<blockquote>
<p>The connectedness corre-lation was calculated and averaged each year for allpairs of ram lambs born in different flocks across all15 flocks (lamb rij)</p>
</blockquote>
<p>è¿˜æ˜¯è¦ä¸¤ä¸ªç¾¤ä½“ä¸¤ä¸¤é…å¯¹ï¼Œå†æ±‚å‡å€¼ã€‚</p>
<p>æœ‰ä¸¤ç§ç®—æ³•ã€‚ä¸€ç§æ˜¯è¿™ç§ï¼Œè¿˜æœ‰ä¸€ç§å¤§æ¦‚æ„æ€æ˜¯ç›´æ¥ç”¨åœºæ•ˆåº”,å’Œmathurä¸€è‡´ã€‚</p>
<p>scaled PEVD å°±æ˜¯å†é™¤ä»¥åŠ æ€§æ–¹å·®ã€‚</p>
<p>è¿˜æ˜¯ä¸¤ç§è®¡ç®—PEVDçš„ç®—æ³•ï¼Œä¸€ç§æ˜¯ç”¨ä¸ªä½“ï¼Œä½†æ˜¯è¦ä¸¤ä¸¤é…å¯¹æ±‚å‡å€¼ï¼›ä¸€ç§å°±æ˜¯ç›´æ¥ç”¨åœºæ•ˆåº”ã€‚</p>
<h1>Hé˜µ</h1>
<p>mistal 2010 å’Œ 2012 çš„æ–‡çŒ®ä¸­ï¼ŒAä¸‹æ ‡çš„1 å’Œ 2 æ˜¯ç›¸åçš„ã€‚</p>
<h1>æš‘å‡ç»Ÿè®¡é—ä¼ å­¦åŸ¹è®­ç­</h1>
<h2 id="çŸ©é˜µè¿ç®—ï¼ˆmatrix-algebraï¼‰">çŸ©é˜µè¿ç®—ï¼ˆmatrix algebraï¼‰</h2>
<ul>
<li>å¯¹äºå¯¹ç§°é˜µï¼ŒAâ€™ = A ã€‚ï¼ˆå¯¹ç§°é˜µçš„è½¬ç½®çŸ©é˜µä¸ºå…¶æœ¬èº«ï¼‰</li>
<li>ï¼ˆAB)â€™ = Bâ€™Aâ€™</li>
<li>æ–¹é˜µçš„è¿¹ï¼ˆtraceï¼‰<br>
tr(A) = å¯¹è§’çº¿å…ƒç´ ä¹‹å’Œ<br>
trA = sum(diag(A))</li>
<li>çŸ©é˜µçš„ç§©ï¼ˆRankï¼‰ï¼šçŸ©é˜µä¸­çº¿æ€§æ— å…³çš„è¡Œæ•°æˆ–åˆ—æ•°<br>
<strong>ä¸æ»¡ç§©çš„çŸ©é˜µç§°ä¸ºå¥‡å¼‚é˜µï¼ˆsingularï¼‰</strong><br>
r &lt;- qr(D)$rank</li>
<li>æ–¹é˜µçš„<strong>è¡Œåˆ—å¼</strong>ï¼šå¦‚æœ|A|=0ï¼ŒA ä¸æ»¡ç§©ï¼Œå³ A æ˜¯å¥‡å¼‚çš„<br>
detA &lt;- det(A)</li>
</ul>
<h3 id="æ±‚é€†">æ±‚é€†</h3>
<p>å¯¹äºä¸€ä¸ª<strong>æ»¡ç§©</strong>çš„æ–¹é˜µAï¼Œå¿…å‡åœ¨å…¶é€†çŸ©é˜µï¼Œä¸”å”¯ä¸€ã€‚</p>
<blockquote>
<p>ï¼ˆAB)^-1^ = B^-1^ A^-1^<br>
A- = solve(A)</p>
</blockquote>
<p>å¯¹è§’çŸ©é˜µçš„é€†çŸ©é˜µ</p>
<p><img src="diag_inv.png" alt="diag_inv"></p>
<p><strong>å¹¿ä¹‰é€†</strong><br>
AA^-^A = A<br>
å¦‚æœæ»¡ç§©ï¼Œé‚£ä¹ˆå¹¿ä¹‰é€†å’Œé€†ç›¸ç­‰ï¼Œè§£æ˜¯å”¯ä¸€çš„ã€‚<br>
ä¸€ä¸ªå¥‡å¼‚é˜µå¯ä»¥æœ‰æ— ç©·å¤šä¸ªå¹¿ä¹‰é€†çŸ©é˜µã€‚</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">library(MASS)</span><br><span class="line">ginv(D)</span><br></pre></td></tr></table></figure>
<h3 id="æ­£å®šçŸ©é˜µï¼ˆPositive-definiteï¼‰">æ­£å®šçŸ©é˜µï¼ˆPositive definiteï¼‰</h3>
<p>æ²¡çœ‹æ‡‚</p>
<p><img src="positive_definite.png" alt="positive_definite"></p>
<h3 id="Cholesky-åˆ†è§£">Cholesky åˆ†è§£</h3>
<p>ä»»ä½•<strong>å¯¹ç§°æ­£å®šçŸ©é˜µ</strong>éƒ½å¯åˆ†è§£ä¸º</p>
<p>** A = TTâ€™ **<br>
å…¶ä¸­ï¼ŒTæ˜¯ä¸€ä¸ªä¸‹ä¸‰è§’çŸ©é˜µ<br>
T &lt;- chol(A) #æ³¨ï¼šè¯¥å‡½æ•°è¿”å›çš„æ˜¯ä¸€ä¸ªä¸Šä¸‰è§’çŸ©é˜µ</p>
<h3 id="æ­£äº¤çŸ©é˜µï¼ˆOrthogonalï¼‰">æ­£äº¤çŸ©é˜µï¼ˆOrthogonalï¼‰</h3>
<p>æ­£äº¤çŸ©é˜µçš„è½¬ç½®çŸ©é˜µå’Œé€†çŸ©é˜µç›¸ç­‰</p>
<h3 id="ç‰¹å¾å€¼å’Œç‰¹å¾å‘é‡">ç‰¹å¾å€¼å’Œç‰¹å¾å‘é‡</h3>
<p>å¯¹äºæ–¹é˜µ A, å¦‚æœå­˜åœ¨ä¸€ä¸ªå¸¸é‡ d å’Œä¸€ä¸ªéé›¶å‘é‡ u, å®ƒä»¬æ»¡è¶³</p>
<p>ğ€ğ®=ğ‘‘ğ®</p>
<p>ç§° d æ˜¯ A çš„ç‰¹å¾å€¼ï¼ˆeigenvalueï¼‰,  u æ˜¯ä¸ d å¯¹åº”çš„ç‰¹å¾å‘é‡ï¼ˆeigenvectorï¼‰</p>
<p>note: ä¸€ä¸ªçŸ©é˜µå¯ä»¥æœ‰å¤šä¸ªç‰¹å¾å€¼å’Œå¤šä¸ªå¯¹åº”çš„ç‰¹å¾å‘é‡</p>
<p>eigen(A)</p>
<p>ç‰¹å¾å€¼å’Œç‰¹å¾å‘é‡æ€§è´¨ï¼š</p>
<ul>
<li>ä¸€ä¸ªçŸ©é˜µçš„æ‰€æœ‰ç‰¹å¾å€¼ä¹‹å’Œç­‰äºå…¶è¿¹</li>
<li>ä¸€ä¸ªçŸ©é˜µçš„æ‰€æœ‰ç‰¹å¾å€¼ä¹‹ç§¯ç­‰äºå…¶è¡Œåˆ—å¼</li>
<li>ä¸€ä¸ªçŸ©é˜µçš„éé›¶ç‰¹å¾å€¼çš„ä¸ªæ•°ç­‰äºå…¶ç§©</li>
<li>å¦‚æœä¸€ä¸ªçŸ©é˜µçš„æ‰€æœ‰ç‰¹å¾å€¼å‡å¤§äº0ï¼Œåˆ™è¯¥çŸ©é˜µæ˜¯æ­£å®šçš„</li>
<li>å¦‚æœä¸€ä¸ªçŸ©é˜µçš„æ‰€æœ‰ç‰¹å¾å€¼å‡å¤§äºæˆ–ç­‰äº0ï¼ŒåŠæ­£å®š</li>
<li>å¦‚æœQä¸ºå¯¹ç§°çŸ©é˜µï¼Œåˆ™å®ƒå¯åˆ†è§£ä¸º<br>
** Q = UDUâ€™ **<br>
D ä¸ºå¯¹è§’çŸ©é˜µï¼Œå…¶å¯¹è§’çº¿å…ƒç´ ä¸ºQçš„ç‰¹å¾å€¼<br>
U ä¸ºæ­£äº¤çŸ©é˜µï¼Œå…¶å„åˆ—ä¸ºQçš„ç‰¹å¾å‘é‡</li>
</ul>
<h3 id="çŸ©é˜µçš„å¾®åˆ†ï¼ˆDifferentiationï¼‰">çŸ©é˜µçš„å¾®åˆ†ï¼ˆDifferentiationï¼‰</h3>
<p><img src="Differentiation.png" alt="Differentiation"></p>
<p><img src="Differentiation2.png" alt="Differentiation2"></p>
<h3 id="éšæœºå‘é‡">éšæœºå‘é‡</h3>
<p>ç”±éšæœºå˜é‡ç»„æˆçš„å‘é‡</p>
<p><img src="random_vector.png" alt="random_vector"></p>
<p>ä¸‹ä¸€å¼ å›¾ä¸æ‡‚</p>
<p><img src="random_vector2.png" alt="random_vector"></p>
<h3 id="PEV-PEVD-æ˜¯å¦ä¹˜ä»¥æ®‹å·®æ–¹å·®">PEV / PEVD æ˜¯å¦ä¹˜ä»¥æ®‹å·®æ–¹å·®</h3>
<p>å¦‚æœMMEå·¦å³ä¸¤è¾¹æ²¡æœ‰çº¦æ‰R^-1^ï¼Œé‚£ä¹ˆå°±ä¸ç”¨ä¹˜æ®‹å·®æ–¹å·®ï¼›å¦åˆ™å°±è¦ä¹˜ã€‚ï¼ˆæˆ‘è®¡ç®—éƒ½æ˜¯çº¦æ‰çš„ï¼Œæ‰€ä»¥è¦ä¹˜ï¼‰</p>
<p><img src="pev.png" alt="random_vector"></p>
<h3 id="A-1-çš„è®¡ç®—">A^-1^ çš„è®¡ç®—</h3>
<ol>
<li>ç”±äºAæ˜¯å¯¹ç§°æ­£å®šçŸ©é˜µï¼Œå› æ­¤å¯ä»¥ä½¿ç”¨ Cholesky åˆ†è§£ï¼Œå°†å…¶åˆ†è§£ä¸€ä¸ªä¸‹ä¸‰è§’çŸ©é˜µä¹˜ä»¥å…¶è½¬ç½®çŸ©é˜µã€‚</li>
<li>ä¸‹ä¸‰è§’çŸ©é˜µçŸ©é˜µL å¯ä»¥åˆ†è§£ä¸€ä¸ªæ–°çš„ä¸‹ä¸‰è§’çŸ©é˜µï¼ˆTï¼Œå¯¹è§’çº¿å…ƒç´ ä¸º1ï¼‰ ä¹˜ä»¥ ä¸€ä¸ªå¯¹è§’é˜µ ï¼ˆDï¼Œä¸ºLé˜µçš„å¯¹è§’çº¿å…ƒç´ ï¼‰</li>
<li>å› æ­¤ï¼Œğ€=ğ“ğƒğƒğ“â€˜ ã€‚è€Œ <strong>DD</strong> ç›¸åŒçš„å¯¹è§’é˜µç›¸ä¹˜ä»ä¸ºå¯¹è§’é˜µï¼Œå…¶å…ƒç´ ä¸ºåŸæ¥çš„å¹³æ–¹ã€‚</li>
</ol>
<p><img src="A-.png" alt="random_vector"></p>
<p>æ³¨ï¼šä»»ä½•ä¸€ä¸ªçŸ©é˜µï¼ˆTï¼‰å³ä¹˜ä»¥ä¸€ä¸ªå¯¹è§’é˜µï¼ˆDï¼‰ï¼Œç­‰äºåŸçŸ©é˜µçš„ç¬¬ä¸€åˆ—ä¹˜ä»¥d1ï¼Œç¬¬äºŒåˆ—ä¹˜ä»¥d2â€¦â€¦ä»¥æ­¤ç±»æ¨ã€‚</p>
<h4 id="YD-yield-deviation">YD (yield deviation)</h4>
<p>åº”è¯¥æ˜¯æŒ‡æ ¡æ­£è¡¨å‹ï¼ŒæŒ‡æ ¡æ­£äº†æ‰€æœ‰å›ºå®šç¯å¢ƒæ•ˆåº”çš„åŠ¨ç‰©ä¸ªä½“çš„ï¼ˆå¹³å‡ï¼‰è¡¨å‹å€¼</p>
<h1>çŸ©é˜µè¿ç®—å®é™…æ”¶è·</h1>
<p>å¯¹LHS å’Œ LHS çš„ä¸Šä¸‰è§’çŸ©é˜µ æ±‚é€†ï¼Œç»“æœä¸ä¸€è‡´ã€‚</p>
 
      <!-- reward -->
      
    </div>
    

    <!-- copyright -->
    
    <div class="declare">
      <ul class="post-copyright">
        <li>
          <i class="ri-copyright-line"></i>
          <strong>ç‰ˆæƒå£°æ˜ï¼š </strong>
          
          æœ¬åšå®¢æ‰€æœ‰æ–‡ç« é™¤ç‰¹åˆ«å£°æ˜å¤–ï¼Œè‘—ä½œæƒå½’ä½œè€…æ‰€æœ‰ã€‚è½¬è½½è¯·æ³¨æ˜å‡ºå¤„ï¼
          
        </li>
      </ul>
    </div>
    
    <footer class="article-footer">
       
<div class="share-btn">
      <span class="share-sns share-outer">
        <i class="ri-share-forward-line"></i>
        åˆ†äº«
      </span>
      <div class="share-wrap">
        <i class="arrow"></i>
        <div class="share-icons">
          
          <a class="weibo share-sns" href="javascript:;" data-type="weibo">
            <i class="ri-weibo-fill"></i>
          </a>
          <a class="weixin share-sns wxFab" href="javascript:;" data-type="weixin">
            <i class="ri-wechat-fill"></i>
          </a>
          <a class="qq share-sns" href="javascript:;" data-type="qq">
            <i class="ri-qq-fill"></i>
          </a>
          <a class="douban share-sns" href="javascript:;" data-type="douban">
            <i class="ri-douban-line"></i>
          </a>
          <!-- <a class="qzone share-sns" href="javascript:;" data-type="qzone">
            <i class="icon icon-qzone"></i>
          </a> -->
          
          <a class="facebook share-sns" href="javascript:;" data-type="facebook">
            <i class="ri-facebook-circle-fill"></i>
          </a>
          <a class="twitter share-sns" href="javascript:;" data-type="twitter">
            <i class="ri-twitter-fill"></i>
          </a>
          <a class="google share-sns" href="javascript:;" data-type="google">
            <i class="ri-google-fill"></i>
          </a>
        </div>
      </div>
</div>

<div class="wx-share-modal">
    <a class="modal-close" href="javascript:;"><i class="ri-close-circle-line"></i></a>
    <p>æ‰«ä¸€æ‰«ï¼Œåˆ†äº«åˆ°å¾®ä¿¡</p>
    <div class="wx-qrcode">
      <img src="//api.qrserver.com/v1/create-qr-code/?size=150x150&data=http://yoursite.com/posts/9664406d/" alt="å¾®ä¿¡åˆ†äº«äºŒç»´ç ">
    </div>
</div>

<div id="share-mask"></div>  
  <ul class="article-tag-list" itemprop="keywords"><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E6%95%B0%E6%8D%AE%E5%88%86%E6%9E%90/" rel="tag">æ•°æ®åˆ†æ</a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%90%86%E8%AE%BA%E5%AD%A6%E4%B9%A0/" rel="tag">ç†è®ºå­¦ä¹ </a></li><li class="article-tag-list-item"><a class="article-tag-list-link" href="/tags/%E7%BA%BF%E6%80%A7%E6%A8%A1%E5%9E%8B/" rel="tag">çº¿æ€§æ¨¡å‹</a></li></ul>

    </footer>
  </div>

   
  <nav class="article-nav">
    
      <a href="/posts/aec3be7e/" class="article-nav-link">
        <strong class="article-nav-caption">ä¸Šä¸€ç¯‡</strong>
        <div class="article-nav-title">
          
            linear algebra step by step
          
        </div>
      </a>
    
    
      <a href="/posts/59ca09be/" class="article-nav-link">
        <strong class="article-nav-caption">ä¸‹ä¸€ç¯‡</strong>
        <div class="article-nav-title">DMUè½¯ä»¶ä½¿ç”¨ç¬”è®°</div>
      </a>
    
  </nav>

  
   
<div class="gitalk" id="gitalk-container"></div>

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/gitalk@1.6.2/dist/gitalk.css">


<script src="https://cdn.jsdelivr.net/npm/gitalk@1.6.2/dist/gitalk.min.js"></script>


<script src="https://cdn.jsdelivr.net/npm/blueimp-md5@2.10.0/js/md5.min.js"></script>

<script type="text/javascript">
  var gitalk = new Gitalk({
    clientID: '6d2af33404b444be44e5',
    clientSecret: 'f133b587d5da8ae224e4e922f069e3984d8db35c',
    repo: 'gitalk',
    owner: 'VincereZhou',
    admin: ['VincereZhou'],
    // id: location.pathname,      // Ensure uniqueness and length less than 50
    id: md5(location.pathname),
    distractionFreeMode: false,  // Facebook-like distraction free mode
    pagerDirection: 'last'
  })

  gitalk.render('gitalk-container')
</script>

     
</article>

</section>
      <footer class="footer">
  <div class="outer">
    <ul>
      <li>
        Copyrights &copy;
        2019-2021
        <i class="ri-heart-fill heart_icon"></i> Vincere Zhou
      </li>
    </ul>
    <ul>
      <li>
        
        
        <span>
  <span><i class="ri-user-3-fill"></i>è®¿é—®äººæ•°:<span id="busuanzi_value_site_uv"></span></s>
  <span class="division">|</span>
  <span><i class="ri-eye-fill"></i>æµè§ˆæ¬¡æ•°:<span id="busuanzi_value_page_pv"></span></span>
</span>
        
      </li>
    </ul>
    <ul>
      
    </ul>
    <ul>
      
    </ul>
    <ul>
      <li>
        <!-- cnzzç»Ÿè®¡ -->
        
      </li>
    </ul>

    <!-- ä¸åªåªåœ¨ä¸€èµ·å¤©æ•° -->
	<ul>
		<li><span id="lovetime_span"></span></li>
	</ul>
    <script type="text/javascript">			
        function show_runtime() {
            window.setTimeout("show_runtime()", 1000);
            X = new Date("03/04/2021 22:11:00");
            Y = new Date();
            T = (Y.getTime() - X.getTime());
            M = 24 * 60 * 60 * 1000;
            a = T / M;
            A = Math.floor(a);
            b = (a - A) * 24;
            B = Math.floor(b);
            c = (b - B) * 60;
            C = Math.floor((b - B) * 60);
            D = Math.floor((c - C) * 60);
            lovetime_span.innerHTML = "åªåªå’Œç”·æœ‹å‹åœ¨ä¸€èµ·äº† " + A + "å¤©" + B + "å°æ—¶" + C + "åˆ†" + D + "ç§’"
        }
        show_runtime();
    </script>

  </div>
</footer>
      <div class="float_btns">
        <div class="totop" id="totop">
  <i class="ri-arrow-up-line"></i>
</div>

<div class="todark" id="todark">
  <i class="ri-moon-line"></i>
</div>

      </div>
    </main>
    <aside class="sidebar on">
      <button class="navbar-toggle"></button>
<nav class="navbar">
  
  <div class="logo">
    <a href="/"><img src="/images/mojie.jpg" alt="VincereZhou&#39;s blog"></a>
  </div>
  
  <ul class="nav nav-main">
    
    <li class="nav-item">
      <a class="nav-item-link" href="/">ä¸»é¡µ</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/archives">å½’æ¡£</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/categories">åˆ†ç±»</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/tags">æ ‡ç­¾</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/friends">å‹é“¾</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/posts/ac7827ff">åªåª</a>
    </li>
    
    <li class="nav-item">
      <a class="nav-item-link" href="/about">å…³äº</a>
    </li>
    
  </ul>
</nav>
<nav class="navbar navbar-bottom">
  <ul class="nav">
    <li class="nav-item">
      
      <a class="nav-item-link nav-item-search"  title="æœç´¢">
        <i class="ri-search-line"></i>
      </a>
      
      
      <a class="nav-item-link" target="_blank" href="/atom.xml" title="RSS Feed">
        <i class="ri-rss-line"></i>
      </a>
      
    </li>
  </ul>
</nav>
<div class="search-form-wrap">
  <div class="local-search local-search-plugin">
  <input type="search" id="local-search-input" class="local-search-input" placeholder="Search...">
  <div id="local-search-result" class="local-search-result"></div>
</div>
</div>
    </aside>
    <script>
      if (window.matchMedia("(max-width: 768px)").matches) {
        document.querySelector('.content').classList.remove('on');
        document.querySelector('.sidebar').classList.remove('on');
      }
    </script>
    <div id="mask"></div>

<!-- #reward -->
<div id="reward">
  <span class="close"><i class="ri-close-line"></i></span>
  <p class="reward-p"><i class="ri-cup-line"></i>è¯·æˆ‘å–æ¯å’–å•¡å§~</p>
  <div class="reward-box">
    
    <div class="reward-item">
      <img class="reward-img" src="/images/alipay.jpg">
      <span class="reward-type">æ”¯ä»˜å®</span>
    </div>
    
    
    <div class="reward-item">
      <img class="reward-img" src="/images/weixinpay.jpg">
      <span class="reward-type">å¾®ä¿¡</span>
    </div>
    
  </div>
</div>
    
<script src="/js/jquery-2.0.3.min.js"></script>


<script src="/js/lazyload.min.js"></script>

<!-- Tocbot -->


<script src="/js/tocbot.min.js"></script>

<script>
  tocbot.init({
    tocSelector: '.tocbot',
    contentSelector: '.article-entry',
    headingSelector: 'h1, h2, h3, h4, h5, h6',
    hasInnerContainers: true,
    scrollSmooth: true,
    scrollContainer: 'main',
    positionFixedSelector: '.tocbot',
    positionFixedClass: 'is-position-fixed',
    fixedSidebarOffset: 'auto'
  });
</script>

<script src="https://cdn.jsdelivr.net/npm/jquery-modal@0.9.2/jquery.modal.min.js"></script>
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/jquery-modal@0.9.2/jquery.modal.min.css">
<script src="https://cdn.jsdelivr.net/npm/justifiedGallery@3.7.0/dist/js/jquery.justifiedGallery.min.js"></script>

<script src="/dist/main.js"></script>

<!-- ImageViewer -->

<!-- Root element of PhotoSwipe. Must have class pswp. -->
<div class="pswp" tabindex="-1" role="dialog" aria-hidden="true">

    <!-- Background of PhotoSwipe. 
         It's a separate element as animating opacity is faster than rgba(). -->
    <div class="pswp__bg"></div>

    <!-- Slides wrapper with overflow:hidden. -->
    <div class="pswp__scroll-wrap">

        <!-- Container that holds slides. 
            PhotoSwipe keeps only 3 of them in the DOM to save memory.
            Don't modify these 3 pswp__item elements, data is added later on. -->
        <div class="pswp__container">
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
            <div class="pswp__item"></div>
        </div>

        <!-- Default (PhotoSwipeUI_Default) interface on top of sliding area. Can be changed. -->
        <div class="pswp__ui pswp__ui--hidden">

            <div class="pswp__top-bar">

                <!--  Controls are self-explanatory. Order can be changed. -->

                <div class="pswp__counter"></div>

                <button class="pswp__button pswp__button--close" title="Close (Esc)"></button>

                <button class="pswp__button pswp__button--share" style="display:none" title="Share"></button>

                <button class="pswp__button pswp__button--fs" title="Toggle fullscreen"></button>

                <button class="pswp__button pswp__button--zoom" title="Zoom in/out"></button>

                <!-- Preloader demo http://codepen.io/dimsemenov/pen/yyBWoR -->
                <!-- element will get class pswp__preloader--active when preloader is running -->
                <div class="pswp__preloader">
                    <div class="pswp__preloader__icn">
                        <div class="pswp__preloader__cut">
                            <div class="pswp__preloader__donut"></div>
                        </div>
                    </div>
                </div>
            </div>

            <div class="pswp__share-modal pswp__share-modal--hidden pswp__single-tap">
                <div class="pswp__share-tooltip"></div>
            </div>

            <button class="pswp__button pswp__button--arrow--left" title="Previous (arrow left)">
            </button>

            <button class="pswp__button pswp__button--arrow--right" title="Next (arrow right)">
            </button>

            <div class="pswp__caption">
                <div class="pswp__caption__center"></div>
            </div>

        </div>

    </div>

</div>

<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.css">
<link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/default-skin/default-skin.min.css">
<script src="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/photoswipe@4.1.3/dist/photoswipe-ui-default.min.js"></script>

<script>
    function viewer_init() {
        let pswpElement = document.querySelectorAll('.pswp')[0];
        let $imgArr = document.querySelectorAll(('.article-entry img:not(.reward-img)'))

        $imgArr.forEach(($em, i) => {
            $em.onclick = () => {
                // sliderå±•å¼€çŠ¶æ€
                // todo: è¿™æ ·ä¸å¥½ï¼Œåé¢æ”¹æˆçŠ¶æ€
                if (document.querySelector('.left-col.show')) return
                let items = []
                $imgArr.forEach(($em2, i2) => {
                    let img = $em2.getAttribute('data-idx', i2)
                    let src = $em2.getAttribute('data-target') || $em2.getAttribute('src')
                    let title = $em2.getAttribute('alt')
                    // è·å¾—åŸå›¾å°ºå¯¸
                    const image = new Image()
                    image.src = src
                    items.push({
                        src: src,
                        w: image.width || $em2.width,
                        h: image.height || $em2.height,
                        title: title
                    })
                })
                var gallery = new PhotoSwipe(pswpElement, PhotoSwipeUI_Default, items, {
                    index: parseInt(i)
                });
                gallery.init()
            }
        })
    }
    viewer_init()
</script>

<!-- MathJax -->

<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
      tex2jax: {
          inlineMath: [ ['$','$'], ["\\(","\\)"]  ],
          processEscapes: true,
          skipTags: ['script', 'noscript', 'style', 'textarea', 'pre', 'code']
      }
  });

  MathJax.Hub.Queue(function() {
      var all = MathJax.Hub.getAllJax(), i;
      for(i=0; i < all.length; i += 1) {
          all[i].SourceElement().parentNode.className += ' has-jax';
      }
  });
</script>

<script src="https://cdn.jsdelivr.net/npm/mathjax@2.7.6/unpacked/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
<script>
  var ayerConfig = {
    mathjax: true
  }
</script>

<!-- Katex -->

<!-- busuanzi  -->


<script src="/js/busuanzi-2.3.pure.min.js"></script>


<!-- ClickLove -->

<!-- ClickBoom1 -->

<!-- ClickBoom2 -->

<!-- CodeCopy -->


<link rel="stylesheet" href="/css/clipboard.css">

<script src="https://cdn.jsdelivr.net/npm/clipboard@2/dist/clipboard.min.js"></script>
<script>
  function wait(callback, seconds) {
    var timelag = null;
    timelag = window.setTimeout(callback, seconds);
  }
  !function (e, t, a) {
    var initCopyCode = function(){
      var copyHtml = '';
      copyHtml += '<button class="btn-copy" data-clipboard-snippet="">';
      copyHtml += '<i class="ri-file-copy-2-line"></i><span>COPY</span>';
      copyHtml += '</button>';
      $(".highlight .code pre").before(copyHtml);
      $(".article pre code").before(copyHtml);
      var clipboard = new ClipboardJS('.btn-copy', {
        target: function(trigger) {
          return trigger.nextElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        let $btn = $(e.trigger);
        $btn.addClass('copied');
        let $icon = $($btn.find('i'));
        $icon.removeClass('ri-file-copy-2-line');
        $icon.addClass('ri-checkbox-circle-line');
        let $span = $($btn.find('span'));
        $span[0].innerText = 'COPIED';
        
        wait(function () { // ç­‰å¾…ä¸¤ç§’é’Ÿåæ¢å¤
          $icon.removeClass('ri-checkbox-circle-line');
          $icon.addClass('ri-file-copy-2-line');
          $span[0].innerText = 'COPY';
        }, 2000);
      });
      clipboard.on('error', function(e) {
        e.clearSelection();
        let $btn = $(e.trigger);
        $btn.addClass('copy-failed');
        let $icon = $($btn.find('i'));
        $icon.removeClass('ri-file-copy-2-line');
        $icon.addClass('ri-time-line');
        let $span = $($btn.find('span'));
        $span[0].innerText = 'COPY FAILED';
        
        wait(function () { // ç­‰å¾…ä¸¤ç§’é’Ÿåæ¢å¤
          $icon.removeClass('ri-time-line');
          $icon.addClass('ri-file-copy-2-line');
          $span[0].innerText = 'COPY';
        }, 2000);
      });
    }
    initCopyCode();
  }(window, document);
</script>


<!-- CanvasBackground -->


    
  </div>
<script src="/live2dw/lib/L2Dwidget.min.js?094cbace49a39548bed64abff5988b05"></script><script>L2Dwidget.init({"pluginRootPath":"live2dw/","pluginJsPath":"lib/","pluginModelPath":"assets/","tagMode":false,"debug":false,"model":{"jsonPath":"live2d-widget-model-wanko"},"display":{"position":"left","width":150,"height":300,"hOffset":80,"vOffset":-70},"mobile":{"show":false,"scale":0.5},"log":false});</script></body>

</html>